// @scalar/components/dist/components/ScalarMenu/ScalarMenuProducts.vue.js
import { defineComponent as a, createElementBlock as f, openBlock as c, normalizeProps as p, guardReactiveProps as m, unref as s, createVNode as t, withCtx as n, createTextVNode as l } from "vue";
import { ScalarIconHouse as u, ScalarIconBook as h, ScalarIconArrowUpRight as C } from "@scalar/icons";
import { useBindCx as k } from "@scalar/use-hooks/useBindCx";
import d from "./ScalarMenuProduct.vue.js";
const P = /* @__PURE__ */ a({
  inheritAttrs: !1,
  __name: "ScalarMenuProducts",
  props: {
    selected: {},
    hrefs: {}
  },
  emits: ["open"],
  setup(b) {
    const { cx: i } = k();
    return (o, e) => (c(), f("div", p(m(s(i)("flex flex-col"))), [
      t(d, {
        href: o.hrefs?.dashboard ?? "https://dashboard.scalar.com",
        icon: s(u),
        selected: o.selected === "dashboard",
        onClick: e[0] || (e[0] = (r) => o.$emit("open", r, "dashboard"))
      }, {
        default: n(() => e[3] || (e[3] = [
          l(" Dashboard ")
        ])),
        _: 1,
        __: [3]
      }, 8, ["href", "icon", "selected"]),
      t(d, {
        href: o.hrefs?.docs ?? "https://docs.scalar.com",
        icon: s(h),
        selected: o.selected === "docs",
        onClick: e[1] || (e[1] = (r) => o.$emit("open", r, "docs"))
      }, {
        default: n(() => e[4] || (e[4] = [
          l(" Docs ")
        ])),
        _: 1,
        __: [4]
      }, 8, ["href", "icon", "selected"]),
      t(d, {
        href: o.hrefs?.client ?? "https://client.scalar.com",
        icon: s(C),
        selected: o.selected === "client",
        onClick: e[2] || (e[2] = (r) => o.$emit("open", r, "client"))
      }, {
        default: n(() => e[5] || (e[5] = [
          l(" Client ")
        ])),
        _: 1,
        __: [5]
      }, 8, ["href", "icon", "selected"])
    ], 16));
  }
});
export {
  P as default
};
