// @scalar/components/dist/components/ScalarMenu/ScalarMenuLink.vue.js
import { defineComponent as i, createBlock as n, openBlock as o, unref as t, mergeProps as a, withCtx as l, createElementBlock as m, createElementVNode as c, normalizeClass as s, renderSlot as f } from "vue";
import { useBindCx as p } from "@scalar/use-hooks/useBindCx";
import { DropdownMenu as u } from "radix-vue/namespaced";
import d from "../ScalarDropdown/ScalarDropdownButton.vue.js";
import g from "../ScalarIcon/ScalarIconLegacyAdapter.vue.js";
const h = {
  key: 1,
  class: "size-3"
}, S = /* @__PURE__ */ i({
  inheritAttrs: !1,
  __name: "ScalarMenuLink",
  props: {
    is: { default: () => u.Item },
    icon: {},
    strong: { type: Boolean }
  },
  setup(k) {
    const { cx: r } = p();
    return (e, w) => (o(), n(t(d), a(t(r)("flex items-center"), {
      is: e.is,
      as: "a"
    }), {
      default: l(() => [
        e.icon ? (o(), n(t(g), {
          key: 0,
          class: s([
            e.strong ? "text-c-1" : "text-c-2",
            typeof e.icon == "string" ? "size-3" : "size-3.5 -mx-0.25"
          ]),
          icon: e.icon,
          thickness: e.strong ? "2.5" : "2",
          weight: e.strong ? "bold" : "regular"
        }, null, 8, ["class", "icon", "thickness", "weight"])) : (o(), m("div", h)),
        c("div", {
          class: s(["flex items-center flex-1 min-w-0 truncate", e.strong ? "font-medium" : "font-normal"])
        }, [
          f(e.$slots, "default")
        ], 2)
      ]),
      _: 3
    }, 16, ["is"]));
  }
});
export {
  S as default
};
