// @scalar/components/dist/components/ScalarMenu/ScalarMenuTeamProfile.vue.js
import { defineComponent as a, createElementBlock as t, openBlock as r, normalizeProps as n, guardReactiveProps as i, unref as s, createCommentVNode as c, createBlock as d, Fragment as m, createTextVNode as u, toDisplayString as l } from "vue";
import { ScalarIconUsers as p } from "@scalar/icons";
import { useBindCx as f } from "@scalar/use-hooks/useBindCx";
const b = ["src"], h = {
  key: 1,
  "aria-hidden": "true",
  class: "flex items-center justify-center text-3xs font-medium text-c-3 size-5 bg-b-3 rounded"
}, g = {
  key: 2,
  class: "flex-1 truncate"
}, B = /* @__PURE__ */ a({
  inheritAttrs: !1,
  __name: "ScalarMenuTeamProfile",
  props: {
    src: {},
    label: {}
  },
  setup(k) {
    const { cx: o } = f();
    return (e, y) => (r(), t("div", n(i(s(o)("flex h-full items-center gap-1"))), [
      e.src ? (r(), t("img", {
        key: 0,
        class: "size-5 rounded",
        role: "presentation",
        src: e.src
      }, null, 8, b)) : (r(), t("div", h, [
        e.label && e.label.length > 0 ? (r(), t(m, { key: 0 }, [
          u(l(e.label[0]), 1)
        ], 64)) : (r(), d(s(p), {
          key: 1,
          class: "size-3",
          weight: "bold"
        }))
      ])),
      e.label && e.label.length > 0 ? (r(), t("div", g, l(e.label), 1)) : c("", !0)
    ], 16));
  }
});
export {
  B as default
};
