// @scalar/api-client/dist/components/SideNav/DownloadAppButton.vue2.js
import { defineComponent as e, createBlock as a, openBlock as m, withCtx as o, createElementVNode as s, normalizeClass as r, unref as t, createVNode as c } from "vue";
import { ScalarIcon as p } from "@scalar/components";
import l from "./SideNavLink.vue.js";
import { useLayout as i } from "../../hooks/useLayout.js";
const h = /* @__PURE__ */ e({
  __name: "DownloadAppButton",
  setup(u) {
    const { layout: n } = i();
    return (d, _) => (m(), a(l, {
      class: "download-app-button gap-2 !px-3 !py-1.5 sm:px-3",
      href: "https://scalar.com/download?utm_source=web_client&utm_medium=download_button&utm_campaign=topnav",
      icon: "Download",
      target: "_blank"
    }, {
      icon: o(() => [
        c(t(p), {
          icon: "Download",
          size: "sm",
          thickness: "2"
        })
      ]),
      default: o(() => [
        s("span", {
          class: r(["sr-only text-sm font-medium", { "sm:not-sr-only": t(n) === "web" }])
        }, " Download App ", 2)
      ]),
      _: 1
    }));
  }
});
export {
  h as default
};
