// @scalar/api-client/dist/components/SideNav/SideHelp.vue.js
import { defineComponent as i, createBlock as r, openBlock as c, unref as s, withCtx as n, createVNode as l, createTextVNode as d, createElementVNode as e } from "vue";
import { ScalarDropdown as u, ScalarDropdownItem as o, ScalarIcon as a } from "@scalar/components";
import p from "./SideNavLink.vue.js";
const f = {
  class: "block no-underline",
  href: "https://discord.gg/scalar",
  target: "_blank"
}, m = { class: "flex items-center justify-center" }, _ = {
  class: "block no-underline",
  href: "https://github.com/scalar/scalar",
  target: "_blank"
}, x = { class: "flex items-center justify-center" }, g = {
  class: "block no-underline",
  href: "mailto:support@scalar.com",
  target: "_blank"
}, b = { class: "flex items-center justify-center" }, k = {
  class: "block no-underline",
  href: "https://scalar.com/changelog",
  target: "_blank"
}, h = { class: "flex items-center justify-center" }, w = {
  class: "block no-underline",
  href: "https://github.com/scalar/scalar/issues/2669",
  target: "_blank"
}, y = { class: "flex items-center justify-center" }, D = /* @__PURE__ */ i({
  __name: "SideHelp",
  setup(j) {
    return (v, t) => (c(), r(s(u), {
      class: "max-w-[150px]",
      placement: "top-end"
    }, {
      items: n(() => [
        t[6] || (t[6] = e("span", { class: "text-c-2 px-2.5 py-1.5 text-xs font-medium" }, "Support", -1)),
        e("a", f, [
          l(s(o), { class: "flex w-full items-center gap-1.5" }, {
            default: n(() => [
              e("div", m, [
                l(s(a), {
                  icon: "Discord",
                  size: "xs",
                  thickness: "1.75"
                })
              ]),
              t[1] || (t[1] = e("span", null, "Discord", -1))
            ]),
            _: 1,
            __: [1]
          })
        ]),
        e("a", _, [
          l(s(o), { class: "flex w-full items-center gap-1.5" }, {
            default: n(() => [
              e("div", x, [
                l(s(a), {
                  icon: "GitHubLine",
                  size: "xs",
                  thickness: "1.75"
                })
              ]),
              t[2] || (t[2] = e("span", null, "GitHub", -1))
            ]),
            _: 1,
            __: [2]
          })
        ]),
        e("a", g, [
          l(s(o), { class: "flex w-full items-center gap-1.5" }, {
            default: n(() => [
              e("div", b, [
                l(s(a), {
                  icon: "Email",
                  size: "xs"
                })
              ]),
              t[3] || (t[3] = e("span", null, "Email", -1))
            ]),
            _: 1,
            __: [3]
          })
        ]),
        t[7] || (t[7] = e("span", { class: "text-c-2 px-2.5 py-1.5 text-xs font-medium" }, "Product", -1)),
        e("a", k, [
          l(s(o), { class: "flex w-full items-center gap-1.5" }, {
            default: n(() => [
              e("div", h, [
                l(s(a), {
                  icon: "Changelog",
                  size: "xs",
                  thickness: "1.75"
                })
              ]),
              t[4] || (t[4] = e("span", null, "Changelog", -1))
            ]),
            _: 1,
            __: [4]
          })
        ]),
        e("a", w, [
          l(s(o), { class: "flex w-full items-center gap-1.5" }, {
            default: n(() => [
              e("div", y, [
                l(s(a), {
                  icon: "Roadmap",
                  size: "xs",
                  thickness: "1.75"
                })
              ]),
              t[5] || (t[5] = e("span", null, "Roadmap", -1))
            ]),
            _: 1,
            __: [5]
          })
        ])
      ]),
      default: n(() => [
        l(p, {
          is: "button",
          icon: "Help",
          type: "button"
        }, {
          default: n(() => t[0] || (t[0] = [
            d(" About ")
          ])),
          _: 1,
          __: [0]
        })
      ]),
      _: 1
    }));
  }
});
export {
  D as default
};
