// @scalar/api-reference/dist/features/info-object/License.vue.js
import { defineComponent as i, createElementBlock as r, openBlock as t, createVNode as n, createElementVNode as o, unref as l, toDisplayString as s, Fragment as d } from "vue";
import { ScalarIconGavel as a } from "@scalar/icons";
const m = { class: "group flex h-fit items-center last:border-r-0 xl:border-r xl:first:ml-auto" }, c = ["href"], u = { class: "ml-1 empty:hidden" }, p = { class: "ml-1 empty:hidden" }, b = /* @__PURE__ */ i({
  __name: "License",
  props: {
    value: {}
  },
  setup(f) {
    return (e, h) => (t(), r("div", m, [
      e.value?.url ? (t(), r("a", {
        key: 0,
        class: "text-c-1 hover:bg-b-2 mr-2 flex min-h-7 min-w-7 items-center rounded-lg border px-2 py-1 no-underline group-last:mr-0 xl:border-none",
        href: e.value.url,
        target: "_blank",
        rel: "noopener noreferrer"
      }, [
        n(l(a), {
          weight: "bold",
          class: "size-3 text-current"
        }),
        o("span", u, s(e.value?.name || e.value.identifier || e.value.url), 1)
      ], 8, c)) : (t(), r(d, { key: 1 }, [
        n(l(a), {
          weight: "bold",
          class: "size-3 text-current"
        }),
        o("span", p, s(e.value?.name), 1)
      ], 64))
    ]));
  }
});
export {
  b as default
};
