// @scalar/api-reference/dist/features/info-object/TermsOfService.vue.js
import { defineComponent as t, createElementBlock as n, createCommentVNode as l, openBlock as s, createElementVNode as o, createVNode as a, unref as m } from "vue";
import { ScalarIconScroll as c } from "@scalar/icons";
const d = {
  key: 0,
  class: "group flex items-center last:border-r-0 xl:border-r xl:first:ml-auto"
}, i = ["href"], _ = /* @__PURE__ */ t({
  __name: "TermsOfService",
  props: {
    value: {}
  },
  setup(p) {
    return (e, r) => e.value ? (s(), n("div", d, [
      o("a", {
        class: "text-c-1 hover:bg-b-2 mr-2 flex min-h-7 min-w-7 items-center rounded-lg border px-2 py-1 no-underline group-last:mr-0 xl:border-none",
        href: e.value,
        target: "_blank",
        rel: "noopener noreferrer"
      }, [
        a(m(c), {
          weight: "bold",
          class: "size-3 text-current"
        }),
        r[0] || (r[0] = o("span", { class: "ml-1 empty:hidden" }, "Terms of Service", -1))
      ], 8, i)
    ])) : l("", !0);
  }
});
export {
  _ as default
};
