// @scalar/api-reference/dist/components/Content/Tags/ClassicLayout.vue2.js
import { defineComponent as r, createBlock as i, openBlock as n, unref as e, withCtx as t, renderSlot as l, createVNode as o, createTextVNode as s, toDisplayString as c } from "vue";
import { ScalarMarkdown as d } from "@scalar/components";
import m from "../../Section/SectionContainerAccordion.vue.js";
import p from "../../Section/SectionHeader.vue.js";
import f from "../../Anchor/Anchor.vue.js";
import u from "../../Section/SectionHeaderTag.vue.js";
const y = /* @__PURE__ */ r({
  __name: "ClassicLayout",
  props: {
    tag: {}
  },
  setup(g) {
    return (a, _) => (n(), i(e(m), { class: "tag-section" }, {
      title: t(() => [
        o(e(p), { class: "tag-name" }, {
          default: t(() => [
            o(e(f), {
              id: a.tag.id
            }, {
              default: t(() => [
                o(e(u), { level: 2 }, {
                  default: t(() => [
                    s(c(a.tag.title), 1)
                  ]),
                  _: 1
                })
              ]),
              _: 1
            }, 8, ["id"])
          ]),
          _: 1
        }),
        o(e(d), {
          class: "tag-description",
          value: a.tag?.description,
          withImages: ""
        }, null, 8, ["value"])
      ]),
      default: t(() => [
        l(a.$slots, "default", {}, void 0, !0)
      ]),
      _: 3
    }));
  }
});
export {
  y as default
};
