// @scalar/api-reference/dist/components/Content/Tags/ModernLayout.vue2.js
import { defineComponent as p, ref as u, useId as y, computed as S, createBlock as i, openBlock as a, unref as o, withCtx as f, createVNode as C, createElementBlock as T, createCommentVNode as s, renderSlot as k, nextTick as b } from "vue";
import w from "../../ShowMoreButton.vue.js";
import { useNavState as B } from "../../../hooks/useNavState.js";
import L from "./TagSection.vue.js";
import { useSidebar as M } from "../../../v2/blocks/scalar-sidebar-block/hooks/useSidebar.js";
import v from "../../Section/SectionContainer.vue.js";
import I from "../../Lazy/Lazy.vue2.js";
const E = /* @__PURE__ */ p({
  __name: "ModernLayout",
  props: {
    tag: {},
    moreThanOneTag: { type: Boolean },
    isLoading: { type: Boolean },
    onShowMore: { type: Function }
  },
  setup(t) {
    const m = u(), l = u(), d = y(), { collapsedSidebarItems: c } = M(), { hash: n } = B(), g = S(
      () => t.moreThanOneTag || t.tag?.title !== "default" || t.tag?.description !== ""
    );
    async function h() {
      await b(), l.value?.querySelector("button")?.focus();
    }
    const r = (e) => !c[e];
    return (e, O) => (a(), i(o(v), {
      ref_key: "sectionContainerRef",
      ref: m,
      "aria-labelledby": o(d),
      class: "tag-section-container",
      role: "region"
    }, {
      default: f(() => [
        C(o(I), {
          id: `modern-tag-${e.tag.id}`,
          isLazy: !!o(n) && o(n) !== e.tag.id && o(n).startsWith(e.tag.id)
        }, {
          default: f(() => [
            g.value ? (a(), i(L, {
              key: 0,
              headerId: o(d),
              isCollapsed: r(e.tag.id),
              isLoading: e.isLoading,
              tag: e.tag
            }, null, 8, ["headerId", "isCollapsed", "isLoading", "tag"])) : s("", !0),
            r(e.tag.id) && e.moreThanOneTag ? (a(), i(w, {
              key: 1,
              id: e.tag.id,
              "aria-label": `Show all ${e.tag.title} endpoints`,
              onShowMore: e.onShowMore,
              onClick: h
            }, null, 8, ["id", "aria-label", "onShowMore"])) : s("", !0)
          ]),
          _: 1
        }, 8, ["id", "isLazy"]),
        r(e.tag.id) && e.moreThanOneTag ? s("", !0) : (a(), T("div", {
          key: 0,
          ref_key: "contentsRef",
          ref: l,
          class: "contents"
        }, [
          k(e.$slots, "default", {}, void 0, !0)
        ], 512))
      ]),
      _: 3
    }, 8, ["aria-labelledby"]));
  }
});
export {
  E as default
};
