// @scalar/api-reference/dist/components/Content/Tags/TagSection.vue.js
import { defineComponent as p, createBlock as i, createCommentVNode as r, openBlock as n, unref as o, withCtx as t, withDirectives as u, createVNode as a, createTextVNode as d, toDisplayString as f, vShow as s } from "vue";
import { ScalarMarkdown as g } from "@scalar/components";
import C from "../../ScreenReader.vue.js";
import S from "../../Section/Section.vue.js";
import c from "../../Section/SectionHeader.vue.js";
import v from "../../Anchor/Anchor.vue.js";
import _ from "../../Section/SectionHeaderTag.vue.js";
import y from "../../Section/SectionContent.vue.js";
import h from "../../Section/SectionColumns.vue.js";
import m from "../../Section/SectionColumn.vue.js";
import k from "../../OperationsList/OperationsList.vue.js";
import w from "../../../features/specification-extension/SpecificationExtension.vue2.js";
const R = /* @__PURE__ */ p({
  __name: "TagSection",
  props: {
    tag: {},
    headerId: {},
    isCollapsed: { type: Boolean },
    isLoading: { type: Boolean }
  },
  setup(B) {
    return (e, l) => e.tag ? (n(), i(o(S), {
      key: 0,
      id: e.tag.id,
      label: e.tag.title?.toUpperCase(),
      role: "none"
    }, {
      default: t(() => [
        u(a(o(c), null, {
          default: t(() => [
            a(o(v), {
              id: e.tag.id
            }, {
              default: t(() => [
                a(o(_), {
                  id: e.headerId,
                  level: 2
                }, {
                  default: t(() => [
                    d(f(e.tag.title) + " ", 1),
                    e.isCollapsed ? (n(), i(C, { key: 0 }, {
                      default: t(() => l[0] || (l[0] = [
                        d(" (Collapsed)")
                      ])),
                      _: 1,
                      __: [0]
                    })) : r("", !0)
                  ]),
                  _: 1
                }, 8, ["id"])
              ]),
              _: 1
            }, 8, ["id"])
          ]),
          _: 1
        }, 512), [
          [s, !e.isLoading]
        ]),
        a(o(y), { loading: e.isLoading }, {
          default: t(() => [
            a(o(h), null, {
              default: t(() => [
                a(o(m), null, {
                  default: t(() => [
                    a(o(g), {
                      clamp: e.isCollapsed ? 7 : void 0,
                      value: e.tag?.description ?? "",
                      withImages: ""
                    }, null, 8, ["clamp", "value"])
                  ]),
                  _: 1
                }),
                a(o(m), null, {
                  default: t(() => [
                    a(o(k), { tag: e.tag }, null, 8, ["tag"])
                  ]),
                  _: 1
                })
              ]),
              _: 1
            })
          ]),
          _: 1
        }, 8, ["loading"]),
        a(o(w), {
          value: e.tag.xKeys
        }, null, 8, ["value"])
      ]),
      _: 1
    }, 8, ["id", "label"])) : r("", !0);
  }
});
export {
  R as default
};
