// @scalar/components/dist/components/ScalarPopover/ScalarPopover.vue.js
import { defineComponent as m, createBlock as i, openBlock as c, unref as e, withCtx as t, createVNode as r, normalizeProps as u, guardReactiveProps as d, renderSlot as s, mergeProps as _ } from "vue";
import { Popover as P, PopoverButton as v, PopoverPanel as B } from "@headlessui/vue";
import { useBindCx as y } from "@scalar/use-hooks/useBindCx";
import $ from "../ScalarFloating/ScalarFloating.vue.js";
import g from "../ScalarFloating/ScalarFloatingBackdrop.vue.js";
const z = /* @__PURE__ */ m({
  inheritAttrs: !1,
  __name: "ScalarPopover",
  props: {
    focus: { type: Boolean },
    placement: {},
    offset: { type: [Number, Object, Function] },
    resize: { type: Boolean },
    target: {},
    middleware: {},
    teleport: { type: [Boolean, String] }
  },
  setup(b) {
    const { cx: l } = y();
    return (o, h) => (c(), i(e(P), { as: "template" }, {
      default: t(({ open: a }) => [
        r(e($), u(d(o.$props)), {
          floating: t(({ width: p, height: f }) => [
            r(e(B), _({
              focus: o.focus,
              style: { width: p, height: f }
            }, e(l)("relative flex flex-col p-0.75")), {
              default: t(({ close: n }) => [
                s(o.$slots, "popover", {
                  close: () => n(),
                  open: a
                }),
                s(o.$slots, "backdrop", { open: a }, () => [
                  r(e(g))
                ])
              ]),
              _: 2
            }, 1040, ["focus", "style"])
          ]),
          default: t(() => [
            r(e(v), { as: "template" }, {
              default: t(() => [
                s(o.$slots, "default", { open: a })
              ]),
              _: 2
            }, 1024)
          ]),
          _: 2
        }, 1040)
      ]),
      _: 3
    }));
  }
});
export {
  z as default
};
