// @azure/static-web-apps-cli/dist/core/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SWA_CLI_DATA_API_URI = exports.IS_DATA_API_DEV_SERVER = exports.SWA_CLI_API_URI = exports.IS_API_DEV_SERVER = exports.IS_APP_DEV_SERVER = exports.DEFAULT_RUNTIME_LANGUAGE = exports.SUPPORTED_VERSIONS = exports.DEFAULT_VERSION = exports.SWA_CONFIG_SCHEME_FALLBACK_PATH = exports.SWA_CONFIG_SCHEME_URL = exports.OVERRIDABLE_ERROR_CODES = exports.CUSTOM_URL_SCHEME = exports.SWA_CONFIG_FILENAME_LEGACY = exports.SWA_CONFIG_FILENAME = exports.HAS_API = exports.SWA_PUBLIC_DIR = exports.SWA_CLI_APP_PROTOCOL = exports.GIT_IGNORE_FILENAME = exports.ENV_FILENAME = exports.CACHE_CONTROL_MAX_AGE = exports.HEADER_DELETE_KEYWORD = exports.DEFAULT_MIME_TYPE = exports.DEFAULT_DATA_API_BUILDER_SCHEMA_CONTENT = exports.MIME_TYPE_LIST = exports.AUTH_STATUS = exports.ALLOWED_HTTP_METHODS_FOR_STATIC_CONTENT = exports.SWA_AUTH_COOKIE = exports.SWA_AUTH_CONTEXT_COOKIE = exports.SWA_RUNTIME_CONFIG_MAX_SIZE_IN_KB = exports.SWA_COMMANDS = exports.DATA_API_BUILDER_DATABASE_TYPES = exports.DEFAULT_DATA_API_BUILDER_BINARY = exports.DATA_API_BUILDER_VERSION_ID = exports.DATA_API_BUILDER_DEFAULT_REST_PATH = exports.DATA_API_BUILDER_DEFAULT_FOLDER = exports.DATA_API_BUILDER_DEFAULT_SCHEMA_FILE_NAME = exports.DATA_API_BUILDER_DEFAULT_CONFIG_FILE_NAME = exports.DATA_API_BUILDER_LATEST_TAG = exports.DATA_API_BUILDER_RELEASE_TAG = exports.DATA_API_BUILDER_FOLDER = exports.DATA_API_BUILDER_RELEASE_METADATA_URL = exports.DATA_API_BUILDER_COMMAND = exports.DATA_API_BUILDER_BINARY_NAME = exports.STATIC_SITE_CLIENT_RELEASE_METADATA_URL = exports.DEPLOY_FOLDER = exports.DEPLOY_BINARY_STABLE_TAG = exports.DEPLOY_BINARY_NAME = void 0;
const path_1 = __importDefault(require("path"));
const config_1 = require("../config");
const net_1 = require("./utils/net");
const os_1 = __importDefault(require("os"));
// StaticSiteClient related constants
exports.DEPLOY_BINARY_NAME = "StaticSitesClient";
exports.DEPLOY_BINARY_STABLE_TAG = "stable";
exports.DEPLOY_FOLDER = path_1.default.join(os_1.default.homedir(), ".swa", "deploy");
exports.STATIC_SITE_CLIENT_RELEASE_METADATA_URL = "https://swalocaldeploy.azureedge.net/downloads/versions.json";
// Data-api-builder related constants
exports.DATA_API_BUILDER_BINARY_NAME = "DataApiBuilder";
exports.DATA_API_BUILDER_COMMAND = "dab";
exports.DATA_API_BUILDER_RELEASE_METADATA_URL = "https://go.microsoft.com/fwlink/?linkid=2226493";
exports.DATA_API_BUILDER_FOLDER = path_1.default.join(os_1.default.homedir(), ".swa", "dataApiBuilder");
exports.DATA_API_BUILDER_RELEASE_TAG = "released";
exports.DATA_API_BUILDER_LATEST_TAG = "latest";
exports.DATA_API_BUILDER_DEFAULT_CONFIG_FILE_NAME = "staticwebapp.database.config.json";
exports.DATA_API_BUILDER_DEFAULT_SCHEMA_FILE_NAME = "staticwebapp.database.schema.gql";
exports.DATA_API_BUILDER_DEFAULT_FOLDER = "swa-db-connections";
exports.DATA_API_BUILDER_DEFAULT_REST_PATH = "/rest";
exports.DATA_API_BUILDER_VERSION_ID = "0.5.32";
exports.DEFAULT_DATA_API_BUILDER_BINARY = {
    Windows: "Microsoft.DataApiBuilder.exe",
    Linux: "Microsoft.DataApiBuilder",
    MacOs: "Microsoft.DataApiBuilder",
};
exports.DATA_API_BUILDER_DATABASE_TYPES = {
    MsSql: "mssql",
    CosmosDbNoSql: "cosmosdb_nosql",
    CosmosDbPostGreSql: "cosmosdb_postgresql",
    MySql: "mysql",
    PostGreSql: "postgresql",
};
// General SWA_CLI constants
exports.SWA_COMMANDS = ["login", "init", "start", "deploy", "build", "db init"];
exports.SWA_RUNTIME_CONFIG_MAX_SIZE_IN_KB = 20; // 20kb
exports.SWA_AUTH_CONTEXT_COOKIE = `StaticWebAppsAuthContextCookie`;
exports.SWA_AUTH_COOKIE = `StaticWebAppsAuthCookie`;
exports.ALLOWED_HTTP_METHODS_FOR_STATIC_CONTENT = ["GET", "HEAD", "OPTIONS"];
exports.AUTH_STATUS = {
    NoAuth: 0,
    HostNameAuthLogin: 1,
    HostNameAuthComplete: 2,
    HostNameAuthLogout: 3,
    IdentityAuthLogin: 4,
    IdentityRedirect: 5,
    IdentityAuthLoginDone: 6,
    IdentityAuthConsentGranted: 7,
    HostNameAuthAcceptInvitation: 8,
    AuthMe: 9,
    HostNameAuthLogoutComplete: 10,
    IdentityRedirectLogout: 11,
    IdentityAuthLogoutComplete: 12,
    IdentityAuthPurgeBegin: 13,
    IdentityAuthPurgeWarning: 14,
    IdentityAuthPurgeComplete: 15,
    HostNameAuthPurge: 16,
    HostNameAuthLoginCallback: 17,
    IdentityAuthLoginCallback: 18,
    IdentityAuthLogout: 19,
};
// this list must cover this set at least: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types
exports.MIME_TYPE_LIST = {
    ".aac": "audio/aac",
    ".abw": "application/x-abiword",
    ".appinstaller": "application/appinstaller",
    ".appx": "application/appx",
    ".appxbundle": "application/appxbundle",
    ".arc": "application/x-freearc",
    ".asf": "video/x-ms-asf",
    ".asx": "video/x-ms-asf",
    ".avi": "video/x-msvideo",
    ".azw": "application/vnd.amazon.ebook",
    ".bin": "application/octet-stream",
    ".bmp": "image/bmp",
    ".bz": "application/x-bzip",
    ".bz2": "application/x-bzip2",
    ".cco": "application/x-cocoa",
    ".crt": "application/x-x509-ca-cert",
    ".csh": "application/x-csh",
    ".css": "text/css",
    ".csv": "text/csv",
    ".deb": "application/octet-stream",
    ".der": "application/x-x509-ca-cert",
    ".dll": "application/octet-stream",
    ".dmg": "application/octet-stream",
    ".doc": "application/msword",
    ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    ".ear": "application/java-archive",
    ".eot": "application/octet-stream",
    ".epub": "application/epub+zip",
    ".exe": "application/octet-stream",
    ".flv": "video/x-flv",
    ".gif": "image/gif",
    ".gz": "application/gzip",
    ".hqx": "application/mac-binhex40",
    ".htc": "text/x-component",
    ".htm": "text/html",
    ".html": "text/html",
    ".ico": "image/vnd.microsoft.icon",
    ".ics": "text/calendar",
    ".img": "application/octet-stream",
    ".iso": "application/octet-stream",
    ".jar": "application/java-archive",
    ".jardiff": "application/x-java-archive-diff",
    ".jng": "image/x-jng",
    ".jnlp": "application/x-java-jnlp-file",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "text/javascript",
    ".vue": "text/javascript",
    ".json": "application/json",
    ".map": "application/json",
    ".jsonld": "application/ld+json",
    ".mid": "audio/midi",
    ".midi": "audio/x-midi",
    ".mjs": "text/javascript",
    ".mml": "text/mathml",
    ".mng": "video/x-mng",
    ".mov": "video/quicktime",
    ".mp3": "audio/mpeg",
    ".mpeg": "video/mpeg",
    ".mpg": "video/mpeg",
    ".mpkg": "application/vnd.apple.installer+xml",
    ".msi": "application/octet-stream",
    ".msix": "application/msix",
    ".msixbundle": "application/msixbundle",
    ".msm": "application/octet-stream",
    ".msp": "application/octet-stream",
    ".odp": "application/vnd.oasis.opendocument.presentation",
    ".ods": "application/vnd.oasis.opendocument.spreadsheet",
    ".odt": "application/vnd.oasis.opendocument.text",
    ".oga": "audio/ogg",
    ".ogv": "video/ogg",
    ".ogx": "application/ogg",
    ".opus": "audio/opus",
    ".otf": "font/otf",
    ".pdb": "application/x-pilot",
    ".pdf": "application/pdf",
    ".pem": "application/x-x509-ca-cert",
    ".php": "application/php",
    ".pl": "application/x-perl",
    ".pm": "application/x-perl",
    ".png": "image/png",
    ".ppt": "application/vnd.ms-powerpoint",
    ".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
    ".prc": "application/x-pilot",
    ".ra": "audio/x-realaudio",
    ".rar": "application/x-rar-compressed",
    ".rpm": "application/x-redhat-package-manager",
    ".rss": "text/xml",
    ".rtf": "application/rtf",
    ".run": "application/x-makeself",
    ".sea": "application/x-sea",
    ".sh": "application/x-sh",
    ".shtml": "text/html",
    ".sit": "application/x-stuffit",
    ".svg": "image/svg+xml",
    ".swf": "application/x-shockwave-flash",
    ".tar": "application/x-tar",
    ".tcl": "application/x-tcl",
    ".tif": "image/tiff",
    ".tiff": "image/tiff",
    ".tk": "application/x-tcl",
    ".ts": "video/mp2t",
    ".ttf": "font/ttf",
    ".txt": "text/plain",
    ".vsd": "application/vnd.visio",
    ".war": "application/java-archive",
    ".wasm": "application/wasm",
    ".wav": "audio/wav",
    ".weba": "audio/webm",
    ".webm": "video/webm",
    ".webp": "image/webp",
    ".wbmp": "image/vnd.wap.wbmp",
    ".wmv": "video/x-ms-wmv",
    ".woff": "font/woff",
    ".woff2": "font/woff2",
    ".xhtml": "application/xhtml+xml",
    ".xls": "application/vnd.ms-excel",
    ".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    ".xml": "text/xml",
    ".xpi": "application/x-xpinstall",
    ".xul": "application/vnd.mozilla.xul+xml",
    ".zip": "application/zip",
    ".3gp": "video/3gpp",
    ".3g2": "video/3gpp2",
    ".7z": "application/x-7z-compressed",
};
exports.DEFAULT_DATA_API_BUILDER_SCHEMA_CONTENT = `
"""
Add your CosmosDB NoSQL database schema in this file

For example:

type Book @model {
  id: ID
  title: String
}

"""
`;
exports.DEFAULT_MIME_TYPE = "application/octet-stream";
exports.HEADER_DELETE_KEYWORD = "@@HEADER_DELETE_KEYWORD@@";
exports.CACHE_CONTROL_MAX_AGE = 30;
exports.ENV_FILENAME = ".env";
exports.GIT_IGNORE_FILENAME = ".gitignore";
exports.SWA_CLI_APP_PROTOCOL = config_1.DEFAULT_CONFIG.ssl ? `https` : `http`;
exports.SWA_PUBLIC_DIR = path_1.default.resolve(__dirname, "..", "public"); //SWA_PUBLIC_DIR = "../public"
exports.HAS_API = Boolean(config_1.DEFAULT_CONFIG.apiLocation && SWA_CLI_API_URI());
exports.SWA_CONFIG_FILENAME = "staticwebapp.config.json";
exports.SWA_CONFIG_FILENAME_LEGACY = "routes.json";
exports.CUSTOM_URL_SCHEME = "swa://";
exports.OVERRIDABLE_ERROR_CODES = [400, 401, 403, 404];
exports.SWA_CONFIG_SCHEME_URL = "https://json.schemastore.org/staticwebapp.config.json";
exports.SWA_CONFIG_SCHEME_FALLBACK_PATH = path_1.default.join(__dirname, "..", "..", "schema", exports.SWA_CONFIG_FILENAME);
// Constants related to Api runtime
exports.DEFAULT_VERSION = {
    Node: "16",
    Dotnet: "6.0",
    DotnetIsolated: "6.0",
    Python: "3.8",
};
exports.SUPPORTED_VERSIONS = {
    Node: ["12", "14", "16", "18"],
    Dotnet: ["3.1", "6.0"],
    DotnetIsolated: ["6.0", "7.0", "8.0"],
    Python: ["3.8", "3.9", "3.10"],
};
exports.DEFAULT_RUNTIME_LANGUAGE = "node";
// --
// Note: declare these as functions so that their body gets evaluated at runtime!
// The reason for this is that these function depend on values set by environment variables which are set
// during the startup of the CLI (see src/cli/commands/start.ts)
function IS_APP_DEV_SERVER() {
    return (0, net_1.isHttpUrl)(config_1.DEFAULT_CONFIG.outputLocation);
}
exports.IS_APP_DEV_SERVER = IS_APP_DEV_SERVER;
function IS_API_DEV_SERVER() {
    return (0, net_1.isHttpUrl)(config_1.DEFAULT_CONFIG.apiLocation);
}
exports.IS_API_DEV_SERVER = IS_API_DEV_SERVER;
function SWA_CLI_API_URI() {
    return IS_API_DEV_SERVER() ? config_1.DEFAULT_CONFIG.apiLocation : (0, net_1.address)(config_1.DEFAULT_CONFIG.host, config_1.DEFAULT_CONFIG.apiPort);
}
exports.SWA_CLI_API_URI = SWA_CLI_API_URI;
function IS_DATA_API_DEV_SERVER() {
    return (0, net_1.isHttpUrl)(config_1.DEFAULT_CONFIG.dataApiLocation);
}
exports.IS_DATA_API_DEV_SERVER = IS_DATA_API_DEV_SERVER;
function SWA_CLI_DATA_API_URI() {
    return IS_DATA_API_DEV_SERVER() ? config_1.DEFAULT_CONFIG.dataApiLocation : (0, net_1.address)(config_1.DEFAULT_CONFIG.host, config_1.DEFAULT_CONFIG.dataApiPort, "http");
}
exports.SWA_CLI_DATA_API_URI = SWA_CLI_DATA_API_URI;
