// @scalar/components/dist/components/ScalarSidebar/ScalarSidebarIndent.vue2.js
import { defineComponent as o, computed as c, createElementBlock as t, openBlock as n, normalizeProps as u, guardReactiveProps as b, unref as m, Fragment as f, renderList as p, createElementVNode as v, createCommentVNode as g, normalizeClass as y } from "vue";
import { useBindCx as B } from "@scalar/use-hooks/useBindCx";
const _ = {}, C = /* @__PURE__ */ o({
  ..._,
  inheritAttrs: !1,
  __name: "ScalarSidebarIndent",
  props: {
    indent: { default: 0 },
    selected: { type: Boolean, default: !1 },
    disabled: { type: Boolean }
  },
  setup(d) {
    const a = c(() => Array.from({ length: d.indent }, (e, r) => r)), { cx: s } = B();
    return (e, r) => (n(), t("div", u(b(
      m(s)("scalar-sidebar-indent flex justify-center", {
        "mr-[calc(20px-var(--scalar-sidebar-indent))]": e.indent > 0,
        "scalar-sidebar-indent-selected": e.selected
      })
    )), [
      (n(!0), t(f, null, p(a.value, (l, i) => (n(), t("div", {
        key: l,
        class: "relative w-[var(--scalar-sidebar-indent)]"
      }, [
        r[0] || (r[0] = v("div", { class: "scalar-sidebar-indent-border absolute left-2 inset-y-0 w-border bg-sidebar-indent-border" }, null, -1)),
        i === a.value.length - 1 ? (n(), t("div", {
          key: 0,
          class: y([
            "absolute left-2 inset-y-0 w-border",
            e.disabled ? "" : e.selected ? "bg-sidebar-indent-border-active" : "group-hover/button:bg-sidebar-indent-border-hover"
          ])
        }, null, 2)) : g("", !0)
      ]))), 128))
    ], 16));
  }
});
export {
  C as default
};
