// @scalar/components/dist/components/ScalarSidebar/ScalarSidebarButton.vue.js
import { defineComponent as r, createBlock as i, openBlock as t, resolveDynamicComponent as c, mergeProps as u, unref as a, withCtx as f, renderSlot as s, createElementVNode as p, createElementBlock as n, createCommentVNode as o, createVNode as m } from "vue";
import { cva as b, useBindCx as v } from "@scalar/use-hooks/useBindCx";
import y from "./ScalarSidebarIndent.vue.js";
import h from "../ScalarIcon/ScalarIconLegacyAdapter.vue.js";
const B = { class: "flex items-center gap-1 flex-1 py-2 leading-5" }, $ = {
  key: 0,
  class: "size-4"
}, g = {
  key: 0,
  class: "flex items-center"
}, k = {}, w = /* @__PURE__ */ r({
  ...k,
  inheritAttrs: !1,
  __name: "ScalarSidebarButton",
  props: {
    is: { default: "a" },
    icon: {},
    active: { type: Boolean },
    selected: { type: Boolean },
    disabled: { type: Boolean },
    indent: { default: 0 }
  },
  setup(C) {
    const d = b({
      base: [
        "group/button flex items-stretch rounded px-2 font-sidebar text-c-2 no-underline"
      ],
      variants: {
        active: { true: "text-c-1 font-sidebar-active" },
        disabled: { true: "cursor-auto" },
        selected: { true: "cursor-auto bg-b-2 text-c-1 font-sidebar-active" }
      },
      compoundVariants: [
        { selected: !1, disabled: !1, class: "hover:bg-b-2" }
      ],
      defaultVariants: { selected: !1, disabled: !1, active: !1 }
    }), { cx: l } = v();
    return (e, S) => (t(), i(c(e.is), u({
      "aria-selected": e.selected,
      type: e.is === "button" ? "button" : void 0
    }, a(l)(a(d)({ selected: e.selected, disabled: e.disabled, active: e.active }))), {
      default: f(() => [
        s(e.$slots, "indent", {}, () => [
          m(y, {
            indent: e.indent,
            selected: e.selected,
            disabled: e.disabled
          }, null, 8, ["indent", "selected", "disabled"])
        ]),
        p("div", B, [
          e.icon || e.$slots.icon ? (t(), n("div", $, [
            s(e.$slots, "icon", {}, () => [
              e.icon ? (t(), i(a(h), {
                key: 0,
                icon: e.icon
              }, null, 8, ["icon"])) : o("", !0)
            ])
          ])) : o("", !0),
          s(e.$slots, "default")
        ]),
        e.$slots.aside ? (t(), n("div", g, [
          s(e.$slots, "aside")
        ])) : o("", !0)
      ]),
      _: 3
    }, 16, ["aria-selected", "type"]));
  }
});
export {
  w as default
};
