// @scalar/components/dist/components/ScalarSidebar/ScalarSidebarGroup.vue.js
import { defineComponent as p, mergeModels as m, useModel as c, createElementBlock as f, openBlock as r, renderSlot as t, createBlock as v, createCommentVNode as b, unref as l, createVNode as s, withCtx as n, resolveDynamicComponent as B, normalizeProps as _, mergeProps as g } from "vue";
import { useBindCx as $ } from "@scalar/use-hooks/useBindCx";
import y from "./ScalarSidebarButton.vue.js";
import C from "./ScalarSidebarGroupToggle.vue.js";
import S from "./ScalarSidebarIndent.vue.js";
import { useSidebarGroups as k } from "./useSidebarGroups.js";
const V = { class: "group/item contents" }, x = {}, A = /* @__PURE__ */ p({
  ...x,
  inheritAttrs: !1,
  __name: "ScalarSidebarGroup",
  props: /* @__PURE__ */ m({
    is: { default: "ul" },
    icon: {},
    active: { type: Boolean },
    selected: { type: Boolean },
    disabled: { type: Boolean },
    indent: {}
  }, {
    modelValue: { type: Boolean, default: !1 },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(d) {
    const e = c(d, "modelValue"), { level: a } = k({ increment: !0 }), { cx: u } = $();
    return (o, i) => (r(), f("li", V, [
      t(o.$slots, "button", {
        level: l(a),
        open: !!e.value
      }, () => [
        s(y, {
          is: "button",
          active: o.active,
          "aria-expanded": e.value,
          class: "group/group-button",
          disabled: o.disabled,
          indent: l(a),
          selected: o.selected,
          onClick: i[0] || (i[0] = (M) => e.value = !e.value)
        }, {
          indent: n(() => [
            s(S, {
              class: "mr-0",
              indent: l(a)
            }, null, 8, ["indent"])
          ]),
          icon: n(() => [
            t(o.$slots, "icon", { open: e.value }, () => [
              s(C, {
                class: "text-c-3",
                open: e.value
              }, null, 8, ["open"])
            ])
          ]),
          default: n(() => [
            t(o.$slots, "default", {
              open: !!e.value
            })
          ]),
          _: 3
        }, 8, ["active", "aria-expanded", "disabled", "indent", "selected"])
      ]),
      e.value ? (r(), v(B(o.is), _(g({ key: 0 }, l(u)("group/items flex flex-col gap-px"))), {
        default: n(() => [
          t(o.$slots, "items", {
            open: !!e.value
          })
        ]),
        _: 3
      }, 16)) : b("", !0)
    ]));
  }
});
export {
  A as default
};
