// @scalar/components/dist/components/ScalarSidebar/ScalarSidebarItem.vue.js
import { defineComponent as r, useSlots as s, createElementBlock as i, openBlock as a, createVNode as l, mergeProps as p, unref as d, createSlots as c, renderList as _, withCtx as m, renderSlot as f } from "vue";
import u from "./ScalarSidebarButton.vue.js";
import { useSidebarGroups as S } from "./useSidebarGroups.js";
const B = { class: "group/item contents" }, h = {}, C = /* @__PURE__ */ r({
  ...h,
  inheritAttrs: !1,
  __name: "ScalarSidebarItem",
  props: {
    is: {},
    icon: {},
    active: { type: Boolean },
    selected: { type: Boolean },
    disabled: { type: Boolean },
    indent: { default: () => {
    } }
  },
  setup($) {
    const o = s(), { level: n } = S();
    return (e, b) => (a(), i("li", B, [
      l(u, p({ ...e.$attrs, ...e.$props }, {
        indent: e.indent ?? d(n)
      }), c({ _: 2 }, [
        _(o, (v, t) => ({
          name: t,
          fn: m(() => [
            f(e.$slots, t)
          ])
        }))
      ]), 1040, ["indent"])
    ]));
  }
});
export {
  C as default
};
