// @scalar/components/dist/components/ScalarSidebar/ScalarSidebarSection.vue.js
import { defineComponent as c, createElementBlock as d, openBlock as s, createVNode as n, createBlock as p, unref as o, withCtx as i, renderSlot as r, resolveDynamicComponent as f, normalizeProps as m, guardReactiveProps as u } from "vue";
import { useBindCx as _ } from "@scalar/use-hooks/useBindCx";
import b from "./ScalarSidebarButton.vue.js";
import a from "./ScalarSidebarSpacer.vue.js";
import { useSidebarGroups as B } from "./useSidebarGroups.js";
const g = { class: "group/item group/sidebar-section contents" }, h = {}, P = /* @__PURE__ */ c({
  ...h,
  inheritAttrs: !1,
  __name: "ScalarSidebarSection",
  props: {
    is: { default: "ul" },
    icon: {},
    active: { type: Boolean },
    selected: { type: Boolean },
    disabled: { type: Boolean },
    indent: {}
  },
  setup(v) {
    const { level: t } = B({ increment: !1 }), { cx: l } = _();
    return (e, x) => (s(), d("li", g, [
      n(a, {
        class: "group/spacer-before h-3",
        indent: o(t)
      }, null, 8, ["indent"]),
      n(b, {
        is: "div",
        class: "text-sm font-bold",
        indent: o(t),
        icon: e.icon,
        disabled: ""
      }, {
        default: i(() => [
          r(e.$slots, "default")
        ]),
        _: 3
      }, 8, ["indent", "icon"]),
      (s(), p(f(e.is), m(u(o(l)("flex flex-col gap-px"))), {
        default: i(() => [
          r(e.$slots, "items")
        ]),
        _: 3
      }, 16)),
      n(a, {
        class: "group/spacer-after h-3",
        indent: o(t)
      }, null, 8, ["indent"])
    ]));
  }
});
export {
  P as default
};
