// @scalar/components/dist/components/ScalarHeader/ScalarHeaderButton.vue.js
import { defineComponent as r, createBlock as a, openBlock as s, resolveDynamicComponent as i, mergeProps as u, unref as t, withCtx as p, renderSlot as c } from "vue";
import { cva as l, useBindCx as d } from "@scalar/use-hooks/useBindCx";
const f = {}, g = /* @__PURE__ */ r({
  ...f,
  inheritAttrs: !1,
  __name: "ScalarHeaderButton",
  props: {
    is: { default: "button" },
    active: { type: Boolean }
  },
  setup(m) {
    const o = l({
      base: "group/button flex items-center rounded  px-2.5 py-1.5 font-medium no-underline leading-3 ",
      variants: {
        active: {
          true: "bg-b-3 cursor-default",
          false: "bg-transparent hover:bg-b-3 cursor-pointer"
        }
      }
    }), { cx: n } = d();
    return (e, b) => (s(), a(i(e.is), u({
      type: e.is === "button" ? "button" : void 0
    }, t(n)(t(o)({ active: e.active }))), {
      default: p(() => [
        c(e.$slots, "default")
      ]),
      _: 3
    }, 16, ["type"]));
  }
});
export {
  g as default
};
