// @scalar/snippetz/dist/plugins/fsharp/httpclient/httpclient.js
const fsharpHttpclient = {
  target: "fsharp",
  client: "httpclient",
  title: "HttpClient",
  generate: (request, _) => {
    if (!request) {
      return "";
    }
    const finalUrl = buildUrlWithQueryString(request.url, request.queryString);
    let code = "";
    code += generateHttpRequestMessage(request.method, finalUrl);
    if (request.headers && request.headers.length > 0) {
      code += generateHeadersCode(request.headers);
    }
    if (request.postData) {
      code += generatePostDataCode(request.postData);
    }
    if (request.cookies && request.cookies.length > 0 && request.url) {
      code += generateCookiesCode(request.cookies, request.url);
      code += "let client = new HttpClient(handler)\n";
    } else {
      code += "let client = new HttpClient()\n";
    }
    code += "let! result = client.SendAsync(httpRequestMessage)\n";
    return code;
  }
};
function buildQueryString(queryParams) {
  if (!queryParams || queryParams.length === 0) {
    return "";
  }
  const params = queryParams.map((param) => `${param.name}=${param.value}`);
  return "?" + params.join("&");
}
function buildUrlWithQueryString(baseUrl, queryParams) {
  if (!baseUrl) {
    return "";
  }
  if (!queryParams || queryParams.length === 0) {
    return baseUrl;
  }
  return baseUrl + buildQueryString(queryParams);
}
function generateHttpRequestMessage(method, url) {
  return `let httpRequestMessage = new HttpRequestMessage(
  HttpMethod("${method}"),
  new Uri("${escapeString(url)}")
)

`;
}
function generateHeadersCode(headers) {
  let code = "";
  for (const header of headers) {
    code += `httpRequestMessage.Headers.Add("${escapeString(header.name ?? "")}", "${escapeString(header.value ?? "")}")
`;
  }
  code += "\n";
  return code;
}
function generateCookiesCode(cookies, url) {
  let code = "let cookieContainer = CookieContainer()\n";
  for (const cookie of cookies) {
    code += `cookieContainer.Add(Uri("${escapeString(url)}"), Cookie("${escapeString(cookie.name ?? "")}", "${escapeString(cookie.value ?? "")}"))
`;
  }
  code += "use handler = new HttpClientHandler()\n";
  code += "handler.CookieContainer <- cookieContainer\n\n";
  return code;
}
function generatePostDataCode(postData) {
  if (!postData) {
    return "";
  }
  let code = "";
  switch (postData.mimeType) {
    case "multipart/form-data":
      code += generateMultipartFormDataCode(postData);
      break;
    case "application/x-www-form-urlencoded":
      code += generateUrlEncodedFormDataCode(postData);
      break;
    case "application/json":
      code += generateJsonContentCode(postData);
      break;
    default:
      code += generateGenericContentCode(postData, postData.mimeType);
      break;
  }
  code += "httpRequestMessage.Content <- content\n\n";
  return code;
}
function generateGenericContentCode(postData, contentType) {
  let code = `let content = new StringContent("${escapeString(postData.text ?? "")}", Encoding.UTF8, "${escapeString(contentType ?? "")}")
`;
  code += `content.Headers.ContentType <- MediaTypeHeaderValue("${escapeString(contentType ?? "")}")
`;
  return code;
}
function generateMultipartFormDataCode(postData) {
  let code = "let content = new MultipartFormDataContent()\n";
  let fileIndex = 0;
  for (const param of postData.params) {
    if (param.value === "BINARY") {
      const escapedFileName = escapeString(param.fileName ?? "");
      code += `let fileStreamContent_${fileIndex} = new StreamContent(File.OpenRead("${escapedFileName}"))
`;
      code += `fileStreamContent_${fileIndex}.Headers.ContentType <- MediaTypeHeaderValue("${escapeString(param.contentType ?? "")}")
`;
      code += `content.Add(fileStreamContent_${fileIndex}, "${escapedFileName}", "${escapedFileName}")
`;
      fileIndex++;
    } else {
      code += `content.Add(new StringContent("${escapeString(param.value ?? "")}"), "${escapeString(param.name ?? "")}")
`;
    }
  }
  return code;
}
function generateJsonContentCode(postData) {
  let prettyJson;
  try {
    prettyJson = JSON.stringify(JSON.parse(postData.text ?? "{}"), null, 2);
  } catch {
    prettyJson = postData.text ?? "{}";
  }
  return `let content = new StringContent("""${prettyJson}""", Encoding.UTF8, "application/json")
`;
}
function generateUrlEncodedFormDataCode(postData) {
  let code = "let formUrlEncodedContentDictionary = new Dictionary<string, string>()\n";
  for (const param of postData.params) {
    code += `formUrlEncodedContentDictionary.Add("${escapeString(param.name ?? "")}", "${escapeString(param.value ?? "")}")
`;
  }
  code += "let content = new FormUrlEncodedContent(formUrlEncodedContentDictionary)\n";
  return code;
}
function escapeString(str) {
  if (str == null) {
    return "";
  }
  return str.replace(/\\/g, "\\\\").replace(/"/g, '\\"').replace(/\n/g, "\\n").replace(/\r/g, "\\r").replace(/\t/g, "\\t");
}
export {
  fsharpHttpclient
};
