// @scalar/api-reference/dist/features/specification-extension/SpecificationExtension.vue2.js
import { defineComponent as v, computed as u, createElementBlock as m, createCommentVNode as g, openBlock as r, Fragment as E, renderList as _, createBlock as o, unref as k, withCtx as y, resolveDynamicComponent as c, mergeProps as l } from "vue";
import { ScalarErrorBoundary as C } from "@scalar/components";
import { usePluginManager as h } from "../../plugins/hooks/usePluginManager.js";
const B = {
  key: 0,
  class: "text-base"
}, j = /* @__PURE__ */ v({
  __name: "SpecificationExtension",
  props: {
    value: {}
  },
  setup(s) {
    const { getSpecificationExtensions: p } = h();
    function i(t) {
      return Object.keys(t ?? {}).filter(
        (n) => n.startsWith("x-")
      );
    }
    function f(t) {
      return t.flatMap((n) => p(n)).filter((n) => n.component);
    }
    const d = u(() => i(s.value)), a = u(
      () => f(d.value)
    );
    return (t, n) => typeof t.value == "object" && a.value.length ? (r(), m("div", B, [
      (r(!0), m(E, null, _(a.value, (e) => (r(), o(k(C), null, {
        default: y(() => [
          e.renderer ? (r(), o(c(e.renderer), l({
            key: 0,
            ref_for: !0
          }, {
            [e.name]: t.value?.[e.name],
            component: e.component
          }), null, 16)) : (r(), o(c(e.component), l({
            key: 1,
            ref_for: !0
          }, { [e.name]: t.value?.[e.name] }), null, 16))
        ]),
        _: 2
      }, 1024))), 256))
    ])) : g("", !0);
  }
});
export {
  j as default
};
