// vue3-snackbar/dist/vue3-snackbar.js
import { h as y, computed as c, ref as O, onMounted as D, watch as E, openBlock as v, createElementBlock as k, normalizeClass as T, unref as $, normalizeStyle as G, renderSlot as S, createElementVNode as C, createVNode as I, mergeProps as W, createCommentVNode as A, toDisplayString as V, withModifiers as H, toRaw as Y, inject as ee, nextTick as te, getCurrentScope as se, onScopeDispose as ae, getCurrentInstance as U, toValue as J, onUnmounted as ne, createBlock as N, Teleport as oe, TransitionGroup as re, withCtx as j, Fragment as ie, renderList as q, createSlots as le } from "vue";
var ue = "M8.27,3L3,8.27V15.73L8.27,21H15.73C17.5,19.24 21,15.73 21,15.73V8.27L15.73,3M9.1,5H14.9L19,9.1V14.9L14.9,19H9.1L5,14.9V9.1M11,15H13V17H11V15M11,7H13V13H11V7", ce = "M12,2L1,21H23M12,6L19.53,19H4.47M11,10V14H13V10M11,16V18H13V16", de = "M12 2C6.5 2 2 6.5 2 12S6.5 22 12 22 22 17.5 22 12 17.5 2 12 2M10 17L5 12L6.41 10.59L10 14.17L17.59 6.58L19 8L10 17Z", fe = "M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z", me = "M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20,12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M11,17H13V11H11V17Z";
const pe = {
  transform: "rotate(var(--r, 0deg)) scale(var(--sx, 1), var(--sy, 1))"
}, z = {
  mdi: {
    size: 24,
    viewbox: "0 0 24 24"
  },
  "simple-icons": {
    size: 24,
    viewbox: "0 0 24 24"
  },
  default: {
    size: 0,
    viewbox: "0 0 0 0"
  }
}, ge = {
  props: {
    type: {
      type: String,
      required: !0
    },
    path: {
      type: [String, Object, Array],
      required: !1
    },
    pathString: {
      type: [Array, String, Object],
      required: !1
    }
  },
  setup(t) {
    return () => t.type === "fad" ? (console.warn("vue3-icon does not currently support Duotone FontAwesome icons"), y("path")) : Array.isArray(t.path) ? y(
      "g",
      {
        style: {
          fill: "currentcolor"
        }
      },
      t.path.map((s) => typeof s == "string" ? y("path", { d: s }) : y("path", { ...s }))
    ) : y("path", {
      d: t.pathString,
      style: {
        fill: "currentcolor"
      }
    });
  }
}, P = {
  name: "icon",
  props: {
    /**
     * The icon type, e.g. mdi or simple-icons
     */
    type: {
      type: String,
      default: "mdi"
    },
    /**
     * The FontAwesome icon object
     */
    faIcon: {
      type: Object,
      default: null
    },
    /**
     * The svg path for the icon
     */
    path: {
      type: [String, Object, Array]
    },
    /**
     * The size of the icon when rendered in the browser
     */
    size: {
      type: [Number, String],
      default: 24
    },
    /**
     * The SVG viewbox, affects path position, but not render size
     */
    viewbox: {
      type: String,
      required: !1
    },
    /**
     * Flip the icon either horizontally, vertically, or both
     */
    flip: {
      type: String,
      default: "",
      validator: (t) => ["", "horizontal", "vertical", "both"].includes(t)
    },
    /**
     * Rotate the icon
     */
    rotate: {
      type: [Number, String],
      default: 0
    },
    /**
     * Don't add the XML namespace attribute
     */
    noNamespace: {
      type: Boolean,
      default: !1
    },
    /**
     * Don't add any styles to SVG, disables flip and rotate
     */
    noStyles: {
      type: Boolean,
      default: !1
    },
    /**
     * Don't add the width/height attributes to the SVG element
     */
    noDimensions: {
      type: Boolean,
      default: !1
    }
  },
  setup(t) {
    if (!t.path && !t.faIcon)
      return console.warn("vue3-icon requires either a 'path' or an 'fa-icon' property"), () => y("div");
    const s = c(() => {
      var f;
      return ((f = t.faIcon) == null ? void 0 : f.prefix) || t.type;
    }), n = c(() => parseInt(t.rotate, 10)), e = c(() => z[s.value] || z.default), a = c(() => parseInt(t.size, 10) || e.value.size), r = c(() => t.faIcon ? `0 0 ${t.faIcon.icon[0]} ${t.faIcon.icon[1]}` : !1), l = c(() => r.value || t.viewbox || e.value.viewbox), m = c(() => {
      if (t.noStyles !== !0)
        return {
          ...pe,
          "--sx": ["both", "horizontal"].includes(t.flip) ? "-1" : "1",
          "--sy": ["both", "vertical"].includes(t.flip) ? "-1" : "1",
          "--r": isNaN(n.value) ? n.value : n.value + "deg"
        };
    }), d = c(() => {
      var f;
      return t.faIcon ? (f = t.faIcon) == null ? void 0 : f.icon[4] : t.type === "simple-icons" && typeof t.path == "object" ? t.path.path : t.path;
    });
    return () => y(
      "svg",
      {
        xmlns: t.noNamespace !== !0 ? "http://www.w3.org/2000/svg" : void 0,
        style: m.value,
        class: ["vue3-icon"],
        width: t.noDimensions !== !0 ? a.value : void 0,
        height: t.noDimensions !== !0 ? a.value : void 0,
        viewBox: l.value
      },
      [y(ge, { path: t.path, type: s.value, pathString: d.value })]
    );
  }
}, ve = ["role"], ye = { class: "vue3-snackbar-message-wrapper" }, be = {
  key: 0,
  class: "vue3-snackbar-message-icon"
}, he = { class: "vue3-snackbar-message-content" }, ke = {
  key: 0,
  class: "vue3-snackbar-message-badge"
}, Se = { class: "vue3-snackbar-message-title" }, we = {
  key: 0,
  class: "vue3-snackbar-message-additional"
}, Ce = { class: "vue3-snackbar-message-close" }, Me = {
  __name: "Vue3SnackbarMessage",
  props: {
    borderClass: {
      type: String,
      default: ""
    },
    message: {
      type: Object,
      default: () => ({})
    },
    messageClass: {
      type: String,
      default: ""
    },
    messageActionClass: {
      type: String,
      default: "vue3-snackbar-message-action"
    },
    dense: {
      type: Boolean,
      default: !1
    },
    iconPresets: {
      type: Object,
      required: !0
    }
  },
  emits: ["dismiss", "click:action"],
  setup(t, { emit: s }) {
    const n = s, e = t;
    let a = null, r = null, l = O(!1);
    const m = () => {
      if ((e.message.duration === 0 || e.message.duration === "0") && e.message.dismissible !== !1)
        return;
      const i = !e.message.duration && !e.message.dismissible ? 4e3 : e.message.duration;
      a = setTimeout(f, i);
    };
    D(m), E(
      () => e.message.count,
      (o) => {
        if (o === 1) return !1;
        clearTimeout(a), clearTimeout(r), r = setTimeout(() => {
          l.value = !1;
        }, 1e3), l.value = !0, m();
      }
    );
    const d = () => {
      a && clearTimeout(a), f();
    }, f = () => {
      n("dismiss", e.message);
    }, w = c(
      () => Object.assign(
        {},
        {
          success: {
            path: de
          },
          info: {
            path: me
          },
          warning: {
            path: ce
          },
          error: {
            path: ue
          }
        },
        e.iconPresets
      )
    ), b = c(() => {
      const o = w.value[e.message.type];
      return e.message.type && !o && console.warn("[Vue3 Snackbar] Couldn't find icon preset for type " + e.message.type), o ? (o.type = "mdi", o) : e.message.icon && typeof e.message.icon == "object" ? e.message.icon : {
        path: "",
        type: "default"
      };
    }), u = c(() => {
      var o;
      return e.message.type === "error" ? "alert" : ((o = e.message) == null ? void 0 : o.role) ?? "status";
    }), g = {
      name: "RenderedAction",
      render() {
        const o = c(() => Y(e.message.action));
        return o.value ? y(o.value, {
          message: e.message,
          isDismissible: e.message.isDismissible,
          dismiss: d
        }) : null;
      }
    };
    return (o, i) => (v(), k("article", {
      class: T(["vue3-snackbar-message", [
        e.message.type || "custom",
        e.messageClass,
        e.borderClass,
        {
          "has-background": e.message.background,
          "has-border": e.borderClass,
          "is-dense": e.dense,
          "shake-baby-shake": $(l)
        }
      ]]),
      role: u.value,
      style: G({
        "--message-background": e.message.background,
        "--message-text-color": e.message.textColor,
        "--message-icon-color": e.message.iconColor
      })
    }, [
      S(o.$slots, "message-inner", {
        message: e.message
      }, () => [
        C("div", ye, [
          b.value ? (v(), k("div", be, [
            S(o.$slots, "message-icon", {
              message: e.message,
              icon: b.value
            }, () => [
              I($(P), W(b.value, { role: "img" }), null, 16)
            ])
          ])) : A("", !0),
          C("div", he, [
            S(o.$slots, "message-badge", {
              message: e.message,
              count: e.message.count
            }, () => [
              e.message.count > 1 ? (v(), k("div", ke, V(e.message.count), 1)) : A("", !0)
            ]),
            S(o.$slots, "message-content", {
              message: e.message,
              title: e.message.title,
              text: e.message.text
            }, () => [
              C("div", Se, V(e.message.title || e.message.text), 1),
              e.message.title && e.message.text ? (v(), k("div", we, V(e.message.text), 1)) : A("", !0)
            ]),
            o.$slots["message-action"] ? (v(), k("div", {
              key: 0,
              class: T(e.messageActionClass),
              onClick: i[0] || (i[0] = H((p) => n("click:action", { message: e.message }), ["stop"]))
            }, [
              S(o.$slots, "message-action", {
                message: e.message,
                isDismissible: e.message.dismissible,
                dismiss: d
              }, () => [
                I(g)
              ])
            ], 2)) : A("", !0)
          ]),
          i[1] || (i[1] = C("div", { class: "spacer" }, null, -1)),
          C("div", Ce, [
            S(o.$slots, "message-close-icon", {
              message: e.message,
              isDimissible: e.message.dismissible,
              isDismissible: e.message.dismissible,
              dismiss: d
            }, () => [
              e.message.dismissible !== !1 ? (v(), k("button", {
                key: 0,
                onClick: H(d, ["stop"])
              }, [
                I($(P), {
                  type: "mdi",
                  path: $(fe)
                }, null, 8, ["path"])
              ])) : A("", !0)
            ])
          ])
        ])
      ])
    ], 14, ve));
  }
}, Ae = typeof window < "u" ? HTMLElement : Object, Le = {
  /* ******************************************
   * LOCATION PROPS
   ****************************************** */
  top: {
    type: Boolean,
    default: !1
  },
  bottom: {
    type: Boolean,
    default: !1
  },
  left: {
    type: Boolean,
    default: !1
  },
  right: {
    type: Boolean,
    default: !1
  },
  /* ******************************************
   * COLOUR PROPS
   ****************************************** */
  success: {
    type: String,
    default: "#4caf50"
  },
  error: {
    type: String,
    default: "#ff5252"
  },
  warning: {
    type: String,
    default: "#fb8c00"
  },
  info: {
    type: String,
    default: "#2196f3"
  },
  iconPresets: {
    type: Object,
    default: () => ({})
  },
  messageTextColor: {
    type: String,
    default: "#fff"
  },
  messageIconColor: {
    type: String,
    default: "currentColor"
  },
  /* ******************************************
   * OTHER PROPS
   ****************************************** */
  attach: {
    type: [String, Ae],
    default: "body"
  },
  border: {
    type: String,
    default: "",
    validator: (t) => ["top", "bottom", "left", "right", ""].includes(t)
  },
  backgroundOpacity: {
    type: [String, Number],
    default: 0.12,
    validator: (t) => !isNaN(parseFloat(t)) && isFinite(t)
  },
  backgroundColor: {
    type: String,
    default: "currentColor"
  },
  baseBackgroundColor: {
    type: String,
    default: "#fff"
  },
  duration: {
    type: [Number, String],
    default: null
  },
  messageClass: {
    type: String
  },
  messageActionClass: {
    type: String,
    default: "vue3-snackbar-message-action"
  },
  zIndex: {
    type: Number,
    default: 1e4
  },
  dense: {
    type: Boolean,
    default: !1
  },
  reverse: {
    type: Boolean,
    default: !1
  },
  limit: {
    type: Number,
    default: null
  },
  groups: {
    type: Boolean,
    default: !1
  },
  shadow: {
    type: Boolean,
    default: !1
  },
  dismissOnActionClick: {
    type: Boolean,
    default: !0
  }
};
function $e(t) {
  return t && t.__esModule && Object.prototype.hasOwnProperty.call(t, "default") ? t.default : t;
}
var x = { exports: {} }, R;
function xe() {
  if (R) return x.exports;
  R = 1;
  function t() {
  }
  return t.prototype = {
    on: function(s, n, e) {
      var a = this.e || (this.e = {});
      return (a[s] || (a[s] = [])).push({
        fn: n,
        ctx: e
      }), this;
    },
    once: function(s, n, e) {
      var a = this;
      function r() {
        a.off(s, r), n.apply(e, arguments);
      }
      return r._ = n, this.on(s, r, e);
    },
    emit: function(s) {
      var n = [].slice.call(arguments, 1), e = ((this.e || (this.e = {}))[s] || []).slice(), a = 0, r = e.length;
      for (a; a < r; a++)
        e[a].fn.apply(e[a].ctx, n);
      return this;
    },
    off: function(s, n) {
      var e = this.e || (this.e = {}), a = e[s], r = [];
      if (a && n)
        for (var l = 0, m = a.length; l < m; l++)
          a[l].fn !== n && a[l].fn._ !== n && r.push(a[l]);
      return r.length ? e[s] = r : delete e[s], this;
    }
  }, x.exports = t, x.exports.TinyEmitter = t, x.exports;
}
var B, F;
function _e() {
  if (F) return B;
  F = 1;
  var t = xe();
  return B = new t(), B;
}
var Ie = _e();
const _ = /* @__PURE__ */ $e(Ie), M = {
  $on: (...t) => _.on(...t),
  $once: (...t) => _.once(...t),
  $off: (...t) => _.off(...t),
  $emit: (...t) => _.emit(...t)
}, L = O([]), K = Symbol(), Oe = (t) => Math.abs(t.split("").reduce((s, n) => (s << 5) - s + n.charCodeAt(0) | 0, 0));
let Z = 1;
M.$on("add", (t) => {
  t = { ...t }, t.group || (t.group = Oe(`${t.type}${t.title}${t.text}`).toString(16)), t.id = Z, Z++, L.value.push(t);
});
function Ze() {
  const t = ee(K);
  if (!t)
    throw new Error("No Snackbar provided!");
  return t;
}
const Ge = {
  install: (t, s = {}) => {
    const { disableGlobals: n = !1 } = s, e = {
      add: (a) => {
        M.$emit("add", a);
      },
      clear: () => {
        M.$emit("clear");
      }
    };
    n !== !0 && (t.config.globalProperties.$snackbar = e, typeof window < "u" && (window.$snackbar = e)), t.provide(K, e);
  }
};
function Ve(t) {
  return se() ? (ae(t), !0) : !1;
}
const Q = typeof window < "u" && typeof document < "u";
typeof WorkerGlobalScope < "u" && globalThis instanceof WorkerGlobalScope;
const Be = (t) => t != null;
function Te(t) {
  return U();
}
function De(t) {
  return Array.isArray(t) ? t : [t];
}
function Ee(t, s = !0, n) {
  Te() ? D(t, n) : s ? t() : te(t);
}
const He = Q ? window : void 0, Ne = Q ? window.document : void 0;
function je(t) {
  var s;
  const n = J(t);
  return (s = n == null ? void 0 : n.$el) != null ? s : n;
}
function qe() {
  const t = O(!1), s = U();
  return s && D(() => {
    t.value = !0;
  }, s), t;
}
function ze(t) {
  const s = qe();
  return c(() => (s.value, !!t()));
}
function Pe(t, s, n = {}) {
  const { window: e = He, ...a } = n;
  let r;
  const l = ze(() => e && "MutationObserver" in e), m = () => {
    r && (r.disconnect(), r = void 0);
  }, d = c(() => {
    const u = J(t), g = De(u).map(je).filter(Be);
    return new Set(g);
  }), f = E(
    () => d.value,
    (u) => {
      m(), l.value && u.size && (r = new MutationObserver(s), u.forEach((g) => r.observe(g, a)));
    },
    { immediate: !0, flush: "post" }
  ), w = () => r == null ? void 0 : r.takeRecords(), b = () => {
    f(), m();
  };
  return Ve(b), {
    isSupported: l,
    stop: b,
    takeRecords: w
  };
}
function Re(t = {}) {
  const {
    document: s = Ne,
    selector: n = "html",
    observe: e = !1,
    initialValue: a = "ltr"
  } = t;
  function r() {
    var m, d;
    return (d = (m = s == null ? void 0 : s.querySelector(n)) == null ? void 0 : m.getAttribute("dir")) != null ? d : a;
  }
  const l = O(r());
  return Ee(() => l.value = r()), e && s && Pe(
    s.querySelector(n),
    () => l.value = r(),
    { attributes: !0 }
  ), c({
    get() {
      return l.value;
    },
    set(m) {
      var d, f;
      l.value = m, s && (l.value ? (d = s.querySelector(n)) == null || d.setAttribute("dir", l.value) : (f = s.querySelector(n)) == null || f.removeAttribute("dir"));
    }
  });
}
const We = {
  __name: "Vue3Snackbar",
  props: { ...Le },
  emits: ["added", "group-added", "dismissed", "removed", "cleared", "click:action"],
  setup(t, { emit: s }) {
    const n = Re(), e = t, a = s, r = c(() => ({
      "is-top": e.top,
      "is-bottom": e.top === !1 && e.bottom,
      "is-left": e.left,
      "is-right": e.left === !1 && e.right,
      "is-middle": e.top === !1 && e.bottom === !1,
      "is-centre": e.left === !1 && e.right === !1,
      "has-shadow": e.shadow,
      "is-rtl": n.value === "rtl"
    })), l = c(() => ({
      "--success-colour": e.success,
      "--error-colour": e.error,
      "--warning-colour": e.warning,
      "--info-colour": e.info,
      "--snackbar-zindex": e.zIndex,
      "--background-opacity": e.backgroundOpacity,
      "--background-color": e.backgroundColor,
      "--base-background-color": e.baseBackgroundColor,
      "--message-text-color": e.messageTextColor,
      "--message-icon-color": e.messageIconColor
    })), m = c(() => e.border ? `border-${e.border}` : ""), d = (u, g = !1) => {
      a(g ? "dismissed" : "removed", u), L.value = L.value.filter((o) => e.groups ? o.group !== u.group : o.id !== u.id);
    }, f = c(() => {
      const g = [...L.value].reduce((o, i) => {
        e.duration && !i.duration && i.duration !== 0 && (i.duration = +e.duration);
        const p = e.groups ? i.group : i.id;
        if (o.has(p)) {
          const h = o.get(p);
          i.count = h.count + 1, i.id = h.id, o.set(p, i), a("group-added", i);
        } else
          i.count = 1, a("added", i), o.set(p, i);
        return o;
      }, /* @__PURE__ */ new Map());
      return Array.from(g.values());
    }), w = c(() => {
      const u = e.limit ? f.value.slice(e.limit * -1) : f.value;
      return e.reverse ? u.reverse() : u;
    });
    E(
      w,
      (u) => {
        const g = u.map((p) => p.id), i = f.value.map((p) => p.id).filter((p) => !g.includes(p));
        for (const p of i) {
          const h = f.value.find((X) => X.id === p);
          h && d(h, !1);
        }
      },
      {
        immediate: !0
      }
    ), M.$on("clear", () => {
      a("cleared"), L.value = [];
    }), ne(() => {
      M.$off("add"), M.$off("clear");
    });
    const b = (u) => {
      a("click:action", u), e.dismissOnActionClick && d(u.message, !0);
    };
    return (u, g) => (v(), N(oe, {
      to: e.attach
    }, [
      C("section", {
        id: "vue3-snackbar--container",
        class: T([[r.value], "vue3-snackbar"]),
        style: G(l.value),
        "aria-live": "polite"
      }, [
        I(re, {
          name: "vue3-snackbar-message",
          tag: "div",
          class: "vue3-snackbar--messages"
        }, {
          default: j(() => [
            (v(!0), k(ie, null, q(w.value, (o) => (v(), N(Me, {
              key: o.id,
              message: o,
              "message-class": e.messageClass,
              "message-action-class": e.messageActionClass,
              dense: e.dense,
              "border-class": m.value,
              "icon-presets": e.iconPresets,
              onDismiss: g[0] || (g[0] = (i) => d(i, !0)),
              "onClick:action": b
            }, le({ _: 2 }, [
              q(u.$slots, (i, p) => ({
                name: p,
                fn: j((h) => [
                  S(u.$slots, p, W({ ref_for: !0 }, h))
                ])
              }))
            ]), 1032, ["message", "message-class", "message-action-class", "dense", "border-class", "icon-presets"]))), 128))
          ]),
          _: 3
        })
      ], 6)
    ], 8, ["to"]));
  }
};
export {
  L as SnackbarMessages,
  Ge as SnackbarService,
  We as Vue3Snackbar,
  Ze as useSnackbar
};
