// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/helpers/oauth.js
import { shouldUseProxy as U } from "@scalar/oas-utils/helpers";
import { encode as A, fromUint8Array as S } from "js-base64";
const C = () => {
  const s = new Uint8Array(32);
  return crypto.getRandomValues(s), S(s, !0);
}, E = async (s, r) => {
  if (r === "plain")
    return s;
  const f = new TextEncoder().encode(s), h = await crypto.subtle.digest("SHA-256", f);
  return S(new Uint8Array(h), !0);
}, z = async (s, r, m, f, h) => {
  const i = s[r];
  try {
    if (!i)
      return [new Error("Flow not found"), null];
    const e = m.join(" ");
    if (r === "clientCredentials" || r === "password")
      return k(s, r, e, {
        proxyUrl: h
      });
    const t = (Math.random() + 1).toString(36).substring(2, 10), c = new URL(s[r].authorizationUrl);
    let o = null;
    if (r === "implicit")
      c.searchParams.set("response_type", "token");
    else if (r === "authorizationCode") {
      const a = s[r];
      if (c.searchParams.set("response_type", "code"), a["x-usePkce"] !== "no") {
        const n = C(), l = await E(n, a["x-usePkce"]);
        o = {
          codeVerifier: n,
          codeChallenge: l,
          codeChallengeMethod: a["x-usePkce"] === "SHA-256" ? "S256" : "plain"
        }, c.searchParams.set("code_challenge", l), c.searchParams.set("code_challenge_method", o.codeChallengeMethod);
      }
    }
    const w = s[r];
    if (w["x-scalar-secret-redirect-uri"].startsWith("/")) {
      const a = f.url || window.location.origin + window.location.pathname, n = new URL(w["x-scalar-secret-redirect-uri"], a).toString();
      c.searchParams.set("redirect_uri", n);
    } else
      c.searchParams.set("redirect_uri", w["x-scalar-secret-redirect-uri"]);
    i["x-scalar-security-query"] && Object.keys(i["x-scalar-security-query"]).forEach((a) => {
      const n = i["x-scalar-security-query"]?.[a];
      n && c.searchParams.set(a, n);
    }), c.searchParams.set("client_id", i["x-scalar-secret-client-id"]), c.searchParams.set("state", t), e && c.searchParams.set("scope", e);
    const d = window.open(c, "openAuth2Window", "left=100,top=100,width=800,height=600");
    return d ? new Promise((a) => {
      const n = setInterval(() => {
        let l = null, g = null, x = null, _ = null;
        try {
          const u = new URL(d.location.href).searchParams, y = i["x-tokenName"] || "access_token";
          l = u.get(y), g = u.get("code"), x = u.get("error"), _ = u.get("error_description");
          const p = new URLSearchParams(d.location.href.split("#")[1]);
          l ||= p.get(y), g ||= p.get("code"), x ||= p.get("error"), _ ||= p.get("error_description");
        } catch {
        }
        if (d.closed || l || g || x)
          if (clearInterval(n), d.close(), x)
            a([new Error(`OAuth error: ${x}${_ ? ` (${_})` : ""}`), null]);
          else if (l) {
            const u = d.location.href.match(/state=([^&]*)/)?.[1];
            a(u === t ? [null, l] : [new Error("State mismatch"), null]);
          } else g && r === "authorizationCode" ? new URL(d.location.href).searchParams.get("state") === t ? k(s, r, e, {
            code: g,
            pkce: o,
            proxyUrl: h
          }).then(a) : a([new Error("State mismatch"), null]) : (clearInterval(n), a([new Error("Window was closed without granting authorization"), null]));
      }, 200);
    }) : [new Error("Failed to open auth window"), null];
  } catch {
    return [new Error("Failed to authorize oauth2 flow"), null];
  }
}, k = async (s, r, m, {
  code: f,
  pkce: h,
  proxyUrl: i
} = {}) => {
  const e = s[r];
  if (!e)
    return [new Error("OAuth2 flow was not defined"), null];
  const t = new URLSearchParams();
  if (t.set("client_id", e["x-scalar-secret-client-id"]), m && (r === "clientCredentials" || r === "password") && t.set("scope", m), e["x-scalar-secret-client-secret"] && (!e["x-scalar-credentials-location"] || e["x-scalar-credentials-location"] === "body") && t.set("client_secret", e["x-scalar-secret-client-secret"]), "x-scalar-secret-redirect-uri" in e && e["x-scalar-secret-redirect-uri"] && t.set("redirect_uri", e["x-scalar-secret-redirect-uri"]), f)
    t.set("code", f), t.set("grant_type", "authorization_code"), h && t.set("code_verifier", h.codeVerifier);
  else if (r === "password") {
    const o = s[r];
    t.set("grant_type", "password"), t.set("username", o["x-scalar-secret-username"]), t.set("password", o["x-scalar-secret-password"]);
  } else
    t.set("grant_type", "client_credentials");
  e["x-scalar-security-body"] && Object.entries(e["x-scalar-security-body"]).forEach(([o, w]) => {
    w && t.set(o, w);
  });
  try {
    const o = {
      "Content-Type": "application/x-www-form-urlencoded"
    };
    e["x-scalar-secret-client-secret"] && (!e["x-scalar-credentials-location"] || e["x-scalar-credentials-location"] === "header") && (o.Authorization = `Basic ${A(`${e["x-scalar-secret-client-id"]}:${e["x-scalar-secret-client-secret"]}`)}`);
    const P = U(i, e.tokenUrl) ? `${i}?${new URLSearchParams([["scalar_url", e.tokenUrl]]).toString()}` : e.tokenUrl, a = await (await fetch(P, {
      method: "POST",
      headers: o,
      body: t
    })).json(), n = e["x-tokenName"] || "access_token";
    return [null, a[n]];
  } catch {
    return [new Error("Failed to get an access token. Please check your credentials."), null];
  }
};
export {
  z as authorizeOauth2,
  k as authorizeServers,
  E as generateCodeChallenge
};
