// @scalar/nuxt/dist/module.mjs
import { defineNuxtModule, createResolver, addComponent, extendPages } from '@nuxt/kit';

const module = defineNuxtModule({
  meta: {
    name: "@scalar/nuxt",
    configKey: "scalar"
  },
  // Default configuration options of the Nuxt module
  defaults: {
    darkMode: true,
    metaData: {
      title: "API Documentation by Scalar"
    },
    pathRouting: {
      basePath: "/docs"
    },
    showSidebar: true,
    devtools: true,
    configurations: [],
    layout: false
  },
  setup(_options, _nuxt) {
    const resolver = createResolver(import.meta.url);
    let isOpenApiEnabled = false;
    _nuxt.options.build.transpile.push("yaml");
    _nuxt.options.imports.transform ||= {};
    _nuxt.options.imports.transform.exclude ||= [];
    _nuxt.options.imports.transform.exclude.push(/scalar/);
    _nuxt.options.vite ||= {};
    _nuxt.options.vite.optimizeDeps ||= {};
    _nuxt.options.vite.optimizeDeps.include ||= [];
    _nuxt.options.vite.optimizeDeps.include.push(
      "@scalar/nuxt > @scalar/api-reference",
      "@scalar/nuxt > jsonpointer",
      "@scalar/nuxt > ajv-draft-04",
      "@scalar/nuxt > ajv-formats",
      "@scalar/nuxt > ajv",
      "@scalar/nuxt > ajv-draft-04 > ajv",
      "@scalar/nuxt > ajv-formats > ajv",
      "@scalar/nuxt > whatwg-mimetype",
      "@scalar/nuxt > @scalar/openapi-parser",
      "@scalar/nuxt > debug",
      "@scalar/nuxt > extend",
      "@scalar/nuxt > highlightjs-curl",
      "@scalar/nuxt > highlight.js/lib/core"
    );
    _nuxt.options.vite.ssr ||= {};
    if (Array.isArray(_nuxt.options.vite.ssr.noExternal)) {
      _nuxt.options.vite.ssr.noExternal.push("ajv-draft-04", "ajv-formats", "ajv", "jsonpointer", "whatwg-mimetype");
    } else {
      _nuxt.options.vite.ssr.noExternal = [
        ...Array.isArray(_nuxt.options.vite.ssr.noExternal) ? _nuxt.options.vite.ssr.noExternal : [],
        "ajv-draft-04",
        "ajv-formats",
        "ajv",
        "jsonpointer",
        "whatwg-mimetype"
      ];
    }
    _nuxt.hook("nitro:config", (config) => {
      if (config.experimental?.openAPI) {
        isOpenApiEnabled = true;
        config.openAPI ||= {};
        config.openAPI.production ||= "prerender";
      }
    });
    addComponent({
      name: "ScalarApiReference",
      export: "default",
      filePath: resolver.resolve("./runtime/components/ScalarApiReference.vue")
    });
    extendPages((pages) => {
      if (_options.configurations.length) {
        const { configurations, ...baseConfig } = _options;
        configurations.forEach((_config, index) => {
          const configuration = { ...baseConfig, ..._config };
          pages.push({
            name: "scalar-" + index,
            path: configuration.pathRouting?.basePath + ":pathMatch(.*)*",
            meta: {
              layout: _options.layout,
              configuration,
              isOpenApiEnabled
            },
            file: resolver.resolve("./runtime/pages/ScalarPage.vue")
          });
        });
      } else {
        pages.push({
          name: "scalar",
          path: _options.pathRouting?.basePath + ":pathMatch(.*)*",
          meta: {
            layout: _options.layout,
            configuration: _options,
            isOpenApiEnabled
          },
          file: resolver.resolve("./runtime/pages/ScalarPage.vue")
        });
      }
    });
    if (_nuxt.options.dev && _options.devtools) {
      _nuxt.hook("devtools:customTabs", (tabs) => {
        tabs.push({
          name: "scalar",
          title: "Scalar",
          icon: "https://gist.githubusercontent.com/cameronrohani/0fa020f6dcf957266bff49e7b6b7c05e/raw/17fce1ef37bbb036dca36b778c8b422056ad6fdf/scalar-logo-nuxt-color.svg",
          category: "server",
          view: {
            type: "iframe",
            src: _options.pathRouting?.basePath ?? "/docs"
          }
        });
      });
    }
  }
});

export { module as default };
