// @scalar/components/dist/components/ScalarCheckboxInput/ScalarCheckboxRadioGroup.vue.js
import { defineComponent as s, mergeModels as c, useModel as i, useId as _, createBlock as o, openBlock as l, unref as t, withCtx as u, createElementBlock as f, Fragment as p, renderList as v, createTextVNode as V, toDisplayString as k } from "vue";
import x from "./ScalarCheckboxInput.vue.js";
import g from "../ScalarForm/ScalarFormInputGroup.vue.js";
const h = {}, b = /* @__PURE__ */ s({
  ...h,
  __name: "ScalarCheckboxRadioGroup",
  props: /* @__PURE__ */ c({
    options: { default: () => [] }
  }, {
    modelValue: {},
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(r) {
    const a = i(r, "modelValue"), d = _();
    return (m, y) => (l(), o(t(g), null, {
      default: u(() => [
        (l(!0), f(p, null, v(m.options, (e) => (l(), o(x, {
          type: "radio",
          value: e.value,
          name: t(d),
          modelValue: a.value?.value === e.value,
          "onUpdate:modelValue": (n) => a.value = n ? e : void 0,
          key: e.value
        }, {
          default: u(() => [
            V(k(e.label), 1)
          ]),
          _: 2
        }, 1032, ["value", "name", "modelValue", "onUpdate:modelValue"]))), 128))
      ]),
      _: 1
    }));
  }
});
export {
  b as default
};
