// @scalar/components/dist/components/ScalarCheckboxInput/ScalarCheckboxInput.vue.js
import { defineComponent as u, mergeModels as c, useModel as m, createBlock as p, openBlock as f, unref as o, mergeProps as s, withCtx as _, createVNode as y, createElementVNode as r, withDirectives as h, renderSlot as v, vModelDynamic as x } from "vue";
import { useBindCx as V } from "@scalar/use-hooks/useBindCx";
import b from "./ScalarCheckbox.vue.js";
import k from "../ScalarForm/ScalarFormInput.vue.js";
const C = { class: "flex-1 text-left min-w-0 truncate" }, g = ["type"], B = {}, D = /* @__PURE__ */ u({
  ...B,
  inheritAttrs: !1,
  __name: "ScalarCheckboxInput",
  props: /* @__PURE__ */ c({
    type: { default: "checkbox" }
  }, {
    modelValue: { type: Boolean },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(a) {
    const e = m(a, "modelValue"), { classCx: n, otherAttrs: d } = V();
    return (t, l) => (f(), p(o(k), s(
      { is: "label" },
      o(n)("cursor-pointer gap-2 hover:bg-b-2", { "text-c-1": e.value })
    ), {
      default: _(() => [
        y(b, {
          selected: e.value,
          type: t.type,
          class: "shrink-0"
        }, null, 8, ["selected", "type"]),
        r("div", C, [
          v(t.$slots, "default")
        ]),
        h(r("input", s({
          type: t.type,
          "onUpdate:modelValue": l[0] || (l[0] = (i) => e.value = i),
          class: "sr-only"
        }, o(d)), null, 16, g), [
          [x, e.value]
        ])
      ]),
      _: 3
    }, 16));
  }
});
export {
  D as default
};
