// @scalar/components/dist/components/ScalarCheckboxInput/ScalarCheckboxGroup.vue.js
import { defineComponent as c, mergeModels as f, useModel as _, useId as i, createBlock as t, openBlock as u, unref as o, withCtx as m, createElementBlock as p, Fragment as v, renderList as V, createTextVNode as k, toDisplayString as x } from "vue";
import g from "./ScalarCheckboxInput.vue.js";
import h from "../ScalarForm/ScalarFormInputGroup.vue.js";
const B = {}, S = /* @__PURE__ */ c({
  ...B,
  __name: "ScalarCheckboxGroup",
  props: /* @__PURE__ */ f({
    options: { default: () => [] }
  }, {
    modelValue: { default: [] },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(r) {
    const l = _(r, "modelValue"), d = i();
    return (n, C) => (u(), t(o(h), null, {
      default: m(() => [
        (u(!0), p(v, null, V(n.options, (e) => (u(), t(g, {
          value: e.value,
          name: o(d),
          modelValue: l.value?.some(({ value: a }) => a === e.value),
          "onUpdate:modelValue": (a) => l.value = a ? [...l.value, e] : l.value.filter(({ value: s }) => s !== e.value),
          key: e.value
        }, {
          default: m(() => [
            k(x(e.label), 1)
          ]),
          _: 2
        }, 1032, ["value", "name", "modelValue", "onUpdate:modelValue"]))), 128))
      ]),
      _: 1
    }));
  }
});
export {
  S as default
};
