// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarShareRegister.vue.js
import { defineComponent as m, createElementBlock as t, openBlock as l, Fragment as n, createElementVNode as r, createVNode as c, renderList as d, createBlock as u, createTextVNode as o, resolveDynamicComponent as S, toDisplayString as g, withCtx as i } from "vue";
import { ScalarIconGlobeSimple as b, ScalarIconGitBranch as f, ScalarIconFileMd as _, ScalarIconLockSimple as k, ScalarIconWarningOctagon as w, ScalarIconBracketsCurly as h } from "@scalar/icons";
import y from "./ApiReferenceToolbarBlurb.vue.js";
import x from "./ApiReferenceToolbarRegisterButton.vue.js";
const B = { class: "text-c-2 grid grid-cols-2 gap-2.5 font-medium" }, N = /* @__PURE__ */ m({
  __name: "ApiReferenceToolbarShareRegister",
  props: {
    workspace: {}
  },
  setup(I) {
    const s = [
      { icon: b, label: "Custom Domains" },
      { icon: f, label: "GitHub Sync" },
      { icon: _, label: "Markdown/MDX" },
      { icon: k, label: "Password Protection" },
      { icon: w, label: "Spectral Linting" },
      { icon: h, label: "JSON Schema Support" }
    ];
    return (p, e) => (l(), t(n, null, [
      r("ul", B, [
        (l(), t(n, null, d(s, (a) => r("li", {
          key: a.label,
          class: "flex items-center gap-2"
        }, [
          (l(), u(S(a.icon), {
            class: "text-c-3 size-3.5",
            weight: "bold"
          })),
          o(" " + g(a.label), 1)
        ])), 64))
      ]),
      c(x, { workspace: p.workspace }, {
        default: i(() => e[0] || (e[0] = [
          o(" Deploy on Scalar ")
        ])),
        _: 1,
        __: [0]
      }, 8, ["workspace"]),
      c(y, null, {
        default: i(() => e[1] || (e[1] = [
          o(" Uploading documents to the Scalar Registry is a Premium feature. See what else is included in our "),
          r("a", {
            href: "https://guides.scalar.com/",
            target: "_blank"
          }, " guides", -1),
          o(". ")
        ])),
        _: 1,
        __: [1]
      })
    ], 64));
  }
});
export {
  N as default
};
