// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarConfig.vue.js
import { defineComponent as p, mergeModels as g, useModel as V, computed as m, createBlock as b, openBlock as c, withCtx as l, createVNode as t, createElementVNode as x, unref as u, createTextVNode as r } from "vue";
import { ScalarFormSection as y, ScalarCodeBlock as C, ScalarFormField as d } from "@scalar/components";
import { prettyPrintJson as S } from "@scalar/oas-utils/helpers";
import $ from "./ApiReferenceToolbarConfigLayout.vue.js";
import k from "./ApiReferenceToolbarConfigLayoutOptions.vue.js";
import B from "./ApiReferenceToolbarConfigTheme.vue.js";
import F from "./ApiReferenceToolbarPopover.vue.js";
const M = { class: "flex flex-col gap-4" }, E = /* @__PURE__ */ p({
  __name: "ApiReferenceToolbarConfig",
  props: /* @__PURE__ */ g({
    configuration: {}
  }, {
    overrides: {},
    overridesModifiers: {}
  }),
  emits: ["update:overrides"],
  setup(i) {
    const o = V(i, "overrides"), v = m(() => S({
      ...o.value,
      // Make sure the overrides are first
      ...i.configuration,
      ...o.value
      // But also that they override the configuration
    })), f = m({
      get: () => o.value?.theme ?? i.configuration?.theme ?? "default",
      set: (a) => o.value = { ...o.value, theme: a }
    }), s = m({
      get: () => o.value?.layout ?? i.configuration?.layout ?? "modern",
      set: (a) => o.value = { ...o.value, layout: a }
    });
    return (a, e) => (c(), b(F, { class: "w-120" }, {
      label: l(() => e[3] || (e[3] = [
        r("Configure")
      ])),
      default: l(() => [
        t(u(y), null, {
          label: l(() => e[4] || (e[4] = [
            r("Scalar Configuration")
          ])),
          default: l(() => [
            t(u(C), {
              class: "bg-b-1.5 max-h-40 rounded border text-sm",
              content: v.value,
              lang: "json"
            }, null, 8, ["content"])
          ]),
          _: 1
        }),
        x("div", M, [
          t(u(d), null, {
            label: l(() => e[5] || (e[5] = [
              r("Theme")
            ])),
            default: l(() => [
              t(B, {
                modelValue: f.value,
                "onUpdate:modelValue": e[0] || (e[0] = (n) => f.value = n)
              }, null, 8, ["modelValue"])
            ]),
            _: 1
          }),
          t(u(d), null, {
            label: l(() => e[6] || (e[6] = [
              r("Layout")
            ])),
            default: l(() => [
              t($, {
                modelValue: s.value,
                "onUpdate:modelValue": e[1] || (e[1] = (n) => s.value = n)
              }, null, 8, ["modelValue"])
            ]),
            _: 1
          }),
          t(u(d), { is: "div" }, {
            label: l(() => e[7] || (e[7] = [
              r("Layout Options")
            ])),
            default: l(() => [
              t(k, {
                configuration: a.configuration,
                modelValue: o.value,
                "onUpdate:modelValue": e[2] || (e[2] = (n) => o.value = n)
              }, null, 8, ["configuration", "modelValue"])
            ]),
            _: 1
          })
        ])
      ]),
      _: 1
    }));
  }
});
export {
  E as default
};
