// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarConfigTheme.vue.js
import { defineComponent as s, useModel as i, computed as m, createBlock as p, openBlock as c, unref as n, withCtx as r, createVNode as f, createTextVNode as v, toDisplayString as V } from "vue";
import { ScalarCombobox as b, ScalarListboxInput as x } from "@scalar/components";
import { themeIds as _, themeLabels as h } from "@scalar/themes";
const T = /* @__PURE__ */ s({
  __name: "ApiReferenceToolbarConfigTheme",
  props: {
    modelValue: {},
    modelModifiers: {}
  },
  emits: ["update:modelValue"],
  setup(d) {
    const u = i(d, "modelValue"), t = m(
      () => _.map((e) => ({ id: e, label: h[e] }))
    ), l = m({
      get: () => {
        const e = u.value ?? "default";
        return t.value.find((o) => o.id === e) ?? t.value[0];
      },
      set: (e) => u.value = e.id
    });
    return (e, o) => (c(), p(n(b), {
      modelValue: l.value,
      "onUpdate:modelValue": o[0] || (o[0] = (a) => l.value = a),
      resize: "",
      options: t.value
    }, {
      default: r(({ open: a }) => [
        f(n(x), { open: a }, {
          default: r(() => [
            v(V(l.value.label), 1)
          ]),
          _: 2
        }, 1032, ["open"])
      ]),
      _: 1
    }, 8, ["modelValue", "options"]));
  }
});
export {
  T as default
};
