// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarShareTemporary.vue.js
import { defineComponent as p, mergeModels as f, useModel as v, createElementBlock as k, openBlock as b, Fragment as g, createVNode as t, unref as r, createSlots as y, withCtx as s, createTextVNode as m } from "vue";
import { useLoadingState as S, ScalarTextInput as T, ScalarIconButton as C, ScalarButton as w } from "@scalar/components";
import { ScalarIconCopy as x } from "@scalar/icons";
import { useClipboard as R } from "@scalar/use-hooks/useClipboard";
import { useToasts as A } from "@scalar/use-toasts";
import { REGISTRY_SHARE_URL as B } from "../../consts/urls.js";
import L from "./ApiReferenceToolbarBlurb.vue.js";
import { uploadTempDocument as E } from "./uploadTempDocument.js";
const N = /* @__PURE__ */ p({
  __name: "ApiReferenceToolbarShareTemporary",
  props: /* @__PURE__ */ f({
    workspace: {}
  }, {
    url: {},
    urlModifiers: {}
  }),
  emits: ["update:url"],
  setup(l) {
    const { toast: d } = A(), a = S(), o = v(l, "url");
    async function c() {
      if (a.isLoading || !l.workspace || o.value)
        return;
      a.startLoading();
      const i = l.workspace.exportActiveDocument("json");
      if (!i) {
        d("Unable to export active document", "error"), a.invalidate();
        return;
      }
      try {
        o.value = await E(i), n(o.value), a.validate();
      } catch (e) {
        const u = e instanceof Error ? e.message : "An unknown error occurred";
        d(u, "error"), a.invalidate();
      }
    }
    const { copyToClipboard: n } = R();
    return (i, e) => (b(), k(g, null, [
      t(r(T), {
        readonly: "",
        placeholder: `${r(B)}/apis/...`,
        modelValue: o.value,
        onClick: e[1] || (e[1] = (u) => o.value && r(n)(o.value))
      }, y({ _: 2 }, [
        o.value ? {
          name: "aside",
          fn: s(() => [
            t(r(C), {
              icon: r(x),
              label: "Copy link to clipboard",
              class: "-m-1.5 -ml-1",
              size: "sm",
              onClick: e[0] || (e[0] = (u) => r(n)(o.value))
            }, null, 8, ["icon"])
          ]),
          key: "0"
        } : void 0
      ]), 1032, ["placeholder", "modelValue"]),
      t(r(w), {
        class: "h-auto p-2.5",
        loading: r(a),
        variant: "outlined",
        disabled: !!o.value,
        onClick: c
      }, {
        default: s(() => e[2] || (e[2] = [
          m(" Generate ")
        ])),
        _: 1,
        __: [2]
      }, 8, ["loading", "disabled"]),
      t(L, { class: "-mt-1" }, {
        default: s(() => e[3] || (e[3] = [
          m(" Shared documents will automatically be deleted after 7 days. ")
        ])),
        _: 1,
        __: [3]
      })
    ], 64));
  }
});
export {
  N as default
};
