// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarSdks.vue.js
import { defineComponent as f, ref as b, createBlock as k, openBlock as s, withCtx as e, createVNode as r, unref as t, createElementBlock as _, Fragment as y, renderList as S, createElementVNode as p, createTextVNode as n, toDisplayString as v } from "vue";
import { ScalarFormSection as V, ScalarFormInputGroup as h, ScalarToggleInput as w, ScalarIcon as G } from "@scalar/components";
import x from "./ApiReferenceToolbarBlurb.vue.js";
import T from "./ApiReferenceToolbarPopover.vue.js";
import A from "./ApiReferenceToolbarRegisterButton.vue.js";
const B = { class: "inline-flex items-center gap-2" }, I = /* @__PURE__ */ f({
  __name: "ApiReferenceToolbarSdks",
  props: {
    workspace: {}
  },
  setup(L) {
    const i = [
      {
        key: "typescript",
        label: "Typescript",
        icon: "programming-language-typescript"
      },
      { key: "python", label: "Python", icon: "programming-language-python" },
      { key: "csharp", label: "C#", icon: "programming-language-csharp" },
      { key: "java", label: "Java", icon: "programming-language-java" },
      { key: "ruby", label: "Ruby", icon: "programming-language-ruby" },
      { key: "php", label: "PHP", icon: "programming-language-php" },
      { key: "go", label: "Go", icon: "programming-language-go" }
    ], a = b([]);
    return (u, l) => (s(), k(T, { class: "w-110" }, {
      label: e(() => l[0] || (l[0] = [
        n("Generate SDKs")
      ])),
      default: e(() => [
        r(t(V), null, {
          label: e(() => l[1] || (l[1] = [
            n("Select Languages")
          ])),
          default: e(() => [
            r(t(h), null, {
              default: e(() => [
                (s(), _(y, null, S(i, ({ key: o, label: g, icon: c }) => r(t(w), {
                  key: o,
                  modelValue: a.value.includes(o),
                  "onUpdate:modelValue": (m) => m ? a.value = [...a.value, o] : a.value = a.value.filter(
                    (d) => d !== o
                  )
                }, {
                  default: e(() => [
                    p("span", B, [
                      r(t(G), {
                        icon: c,
                        class: "text-c-2 size-3.5"
                      }, null, 8, ["icon"]),
                      n(" " + v(g), 1)
                    ])
                  ]),
                  _: 2
                }, 1032, ["modelValue", "onUpdate:modelValue"])), 64))
              ]),
              _: 1
            }),
            r(A, {
              sdks: a.value,
              workspace: u.workspace
            }, null, 8, ["sdks", "workspace"]),
            r(x, null, {
              default: e(() => l[2] || (l[2] = [
                n(" Generating SDKs is a paid feature starting at $100/mo, learn more in our "),
                p("a", {
                  href: "https://guides.scalar.com/scalar/scalar-sdks/getting-started",
                  target: "_blank"
                }, " guides", -1),
                n(". ")
              ])),
              _: 1,
              __: [2]
            })
          ]),
          _: 1
        })
      ]),
      _: 1
    }));
  }
});
export {
  I as default
};
