// @scalar/api-reference/dist/features/toolbar/ApiReferenceToolbarConfigLayoutOptions.vue.js
import { defineComponent as p, mergeModels as f, useModel as g, createBlock as M, openBlock as S, unref as o, withCtx as d, createVNode as n, createTextVNode as a } from "vue";
import { ScalarFormInputGroup as T, ScalarToggleInput as i } from "@scalar/components";
const U = /* @__PURE__ */ p({
  __name: "ApiReferenceToolbarConfigLayoutOptions",
  props: /* @__PURE__ */ f({
    configuration: {}
  }, {
    modelValue: {
      default: {}
    },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(r) {
    const s = g(r, "modelValue");
    function t(m, e = !1) {
      return s.value[m] ?? r.configuration?.[m] ?? e;
    }
    function u(m, e, l = !1) {
      e !== l ? s.value = { ...s.value, [m]: e } : s.value = Object.fromEntries(
        Object.entries(s.value).filter(([V]) => m !== V)
      );
    }
    return (m, e) => (S(), M(o(T), null, {
      default: d(() => [
        n(o(i), {
          modelValue: t("showSidebar", !0),
          "onUpdate:modelValue": e[0] || (e[0] = (l) => u("showSidebar", !!l, !0))
        }, {
          default: d(() => e[9] || (e[9] = [
            a(" Show Sidebar ")
          ])),
          _: 1,
          __: [9]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("defaultOpenAllTags"),
          "onUpdate:modelValue": e[1] || (e[1] = (l) => u("defaultOpenAllTags", !!l))
        }, {
          default: d(() => e[10] || (e[10] = [
            a(" Default Open All Tags ")
          ])),
          _: 1,
          __: [10]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("expandAllModelSections"),
          "onUpdate:modelValue": e[2] || (e[2] = (l) => u("expandAllModelSections", !!l))
        }, {
          default: d(() => e[11] || (e[11] = [
            a(" Expand All Model Sections ")
          ])),
          _: 1,
          __: [11]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("expandAllResponses"),
          "onUpdate:modelValue": e[3] || (e[3] = (l) => u("expandAllResponses", !!l))
        }, {
          default: d(() => e[12] || (e[12] = [
            a(" Expand All Responses ")
          ])),
          _: 1,
          __: [12]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("hideClientButton"),
          "onUpdate:modelValue": e[4] || (e[4] = (l) => u("hideClientButton", !!l))
        }, {
          default: d(() => e[13] || (e[13] = [
            a(" Hide Client Button ")
          ])),
          _: 1,
          __: [13]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("hideDarkModeToggle"),
          "onUpdate:modelValue": e[5] || (e[5] = (l) => u("hideDarkModeToggle", !!l))
        }, {
          default: d(() => e[14] || (e[14] = [
            a(" Hide Dark Mode Toggle ")
          ])),
          _: 1,
          __: [14]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("hideModels"),
          "onUpdate:modelValue": e[6] || (e[6] = (l) => u("hideModels", !!l))
        }, {
          default: d(() => e[15] || (e[15] = [
            a(" Hide Models ")
          ])),
          _: 1,
          __: [15]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("hideSearch"),
          "onUpdate:modelValue": e[7] || (e[7] = (l) => u("hideSearch", !!l))
        }, {
          default: d(() => e[16] || (e[16] = [
            a(" Hide Search ")
          ])),
          _: 1,
          __: [16]
        }, 8, ["modelValue"]),
        n(o(i), {
          modelValue: t("hideTestRequestButton"),
          "onUpdate:modelValue": e[8] || (e[8] = (l) => u("hideTestRequestButton", !!l))
        }, {
          default: d(() => e[17] || (e[17] = [
            a(" Hide Test Request Button ")
          ])),
          _: 1,
          __: [17]
        }, 8, ["modelValue"])
      ]),
      _: 1
    }));
  }
});
export {
  U as default
};
