// @scalar/api-client/dist/views/Collection/CollectionOverview.vue2.js
import { defineComponent as m, createBlock as r, openBlock as a, withCtx as p, createCommentVNode as u, unref as e } from "vue";
import v from "../../components/ViewLayout/ViewLayoutSection.vue.js";
import { useActiveEntities as d } from "../../store/active-entities.js";
import f from "./components/MarkdownInput.vue.js";
import { useWorkspace as _ } from "../../store/store.js";
const h = /* @__PURE__ */ m({
  __name: "CollectionOverview",
  setup(k) {
    const {
      activeCollection: o,
      activeEnvironment: t,
      activeEnvVariables: c,
      activeWorkspace: n
    } = d(), { collectionMutators: l } = _(), s = (i) => {
      o.value && l.edit(o.value.uid, "info.description", i);
    };
    return (i, V) => (a(), r(v, null, {
      default: p(() => [
        e(t) && e(n) ? (a(), r(f, {
          key: 0,
          envVariables: e(c),
          environment: e(t),
          modelValue: e(o)?.info?.description ?? "",
          workspace: e(n),
          "onUpdate:modelValue": s
        }, null, 8, ["envVariables", "environment", "modelValue", "workspace"])) : u("", !0)
      ]),
      _: 1
    }));
  }
});
export {
  h as default
};
