// @scalar/api-client/dist/views/Collection/CollectionServers.vue2.js
import { defineComponent as y, ref as w, computed as C, useTemplateRef as A, createElementBlock as u, openBlock as o, createElementVNode as n, createVNode as a, createStaticVNode as M, Fragment as N, renderList as D, createBlock as f, createCommentVNode as I, unref as e, toDisplayString as $, withCtx as v } from "vue";
import { useModal as j, ScalarMarkdown as z, ScalarButton as m, ScalarIcon as B, ScalarModal as E } from "@scalar/components";
import { ScalarIconTrash as U } from "@scalar/icons";
import { emitCustomEvent as V } from "@scalar/workspace-store/events";
import T from "../../components/Sidebar/Actions/DeleteSidebarListElement.vue.js";
import { useActiveEntities as L } from "../../store/active-entities.js";
import P from "./CollectionServerForm.vue.js";
import { useWorkspace as R } from "../../store/store.js";
const F = {
  ref: "wrapper-ref",
  class: "flex h-full w-full flex-col gap-12 px-1.5 pt-8"
}, O = { class: "flex flex-col gap-4" }, W = { class: "rounded-lg border" }, Y = { class: "bg-b-2 flex items-start justify-between rounded-t-lg py-1 pr-1 pl-3 text-sm" }, q = {
  key: 1,
  class: "self-center"
}, G = { class: "text-c-3 flex h-full items-center justify-center rounded-lg border p-4" }, le = /* @__PURE__ */ y({
  __name: "CollectionServers",
  setup(H) {
    const { activeCollection: t } = L(), { servers: i, events: p, serverMutators: x } = R(), c = j(), r = w(null), h = C(() => !i || !t.value?.servers ? [] : Object.values(i).filter(
      (d) => t.value?.servers.includes(d.uid)
    )), _ = () => p.commandPalette.emit({
      commandName: "Add Server"
    }), S = () => {
      !t.value?.uid || !r.value || (x.delete(r.value, t.value.uid), V(b.value, "scalar-delete-server", {
        url: i[r.value]?.url ?? ""
      }), c.hide());
    }, g = (d) => {
      r.value = d, c.show();
    }, b = A("wrapper-ref");
    return (d, s) => (o(), u("div", F, [
      n("div", O, [
        s[2] || (s[2] = M('<div class="flex items-start justify-between gap-2"><div class="flex flex-col gap-2"><div class="flex h-8 items-center"><h3 class="font-bold">Servers</h3></div><p class="text-sm"> Add different base URLs for your API. You can use <code class="font-code text-c-2">{variables}</code> for dynamic parts. </p></div></div>', 1)),
        (o(!0), u(N, null, D(h.value, (l, k) => (o(), u("div", {
          key: l.uid
        }, [
          n("div", W, [
            n("div", Y, [
              l.description ? (o(), f(e(z), {
                key: 0,
                class: "self-center",
                value: l.description
              }, null, 8, ["value"])) : (o(), u("span", q, "Server " + $(k + 1), 1)),
              a(e(m), {
                class: "hover:bg-b-3 hover:text-c-1 h-fit p-1.25",
                variant: "ghost",
                onClick: (J) => g(l.uid)
              }, {
                default: v(() => [
                  a(e(U), { class: "size-3.5" })
                ]),
                _: 2
              }, 1032, ["onClick"])
            ]),
            e(t) ? (o(), f(P, {
              key: 0,
              collectionId: e(t).uid,
              serverUid: l.uid
            }, null, 8, ["collectionId", "serverUid"])) : I("", !0)
          ])
        ]))), 128)),
        n("div", G, [
          a(e(m), {
            class: "hover:bg-b-2 hover:text-c-1 flex items-center gap-2",
            size: "sm",
            variant: "ghost",
            onClick: _
          }, {
            default: v(() => [
              a(e(B), {
                class: "inline-flex",
                icon: "Add",
                size: "sm",
                thickness: "1.5"
              }),
              s[1] || (s[1] = n("span", null, "Add Server", -1))
            ]),
            _: 1,
            __: [1]
          })
        ])
      ]),
      a(e(E), {
        size: "xxs",
        state: e(c),
        title: `Delete ${r.value ? e(i)[r.value]?.url : "Server"}`
      }, {
        default: v(() => [
          a(T, {
            variableName: "Server",
            warningMessage: "Are you sure you want to delete this server? This action cannot be undone.",
            onClose: s[0] || (s[0] = (l) => e(c).hide()),
            onDelete: S
          })
        ]),
        _: 1
      }, 8, ["state", "title"])
    ], 512));
  }
});
export {
  le as default
};
