// @azure/static-web-apps-cli/dist/core/utils/command.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execFileCommand = exports.runCommand = void 0;
const process_1 = __importDefault(require("process"));
const child_process_1 = require("child_process");
function runCommand(command, cwd) {
    (0, child_process_1.execSync)(command, {
        stdio: "inherit",
        cwd: cwd !== null && cwd !== void 0 ? cwd : process_1.default.cwd(),
        // Set CI to avoid extra NPM logs and potentially unwanted interactive modes
        env: {
            ...process_1.default.env,
            // Internal flag to avoid duplicating user messages
            SWA_CLI_INTERNAL_COMMAND: "1",
        },
    });
}
exports.runCommand = runCommand;
function execFileCommand(command, cwd, args) {
    const child = (0, child_process_1.execFileSync)(command, args, {
        stdio: "inherit",
        cwd: cwd !== null && cwd !== void 0 ? cwd : process_1.default.cwd(),
        // Set CI to avoid extra NPM logs and potentially unwanted interactive modes
        env: {
            ...process_1.default.env,
            // Internal flag to avoid duplicating user messages
            SWA_CLI_INTERNAL_COMMAND: "1",
        },
    });
    return child;
}
exports.execFileCommand = execFileCommand;
