// @azure/static-web-apps-cli/dist/core/utils/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUserOrConfigOption = exports.isConfigFileOption = exports.isUserOption = exports.getUserOptions = exports.configureOptions = void 0;
const commander_1 = require("commander");
const config_1 = require("../../config");
const constants_1 = require("../constants");
const cli_config_1 = require("./cli-config");
const logger_1 = require("./logger");
let userDefinedOptions = {};
let configFileDefinedOptions = {};
async function configureOptions(configName, options, command, commandName, loadConfigFile = true) {
    const verbose = options.verbose;
    setLogLevel(verbose);
    userDefinedOptions = getUserOptions(command);
    const configFileOptions = loadConfigFile ? await (0, cli_config_1.getConfigFileOptions)(options.configName || configName, options.config) : {};
    const configFileCommandSpecificOptions = commandName ? configFileOptions[commandName] || {} : {};
    // Clean up subcommands overrides before merging
    // to avoid confusing the user when printing options
    constants_1.SWA_COMMANDS.forEach((command) => {
        delete configFileOptions[command];
    });
    configFileDefinedOptions = { ...configFileOptions, ...configFileCommandSpecificOptions };
    options = {
        ...options,
        ...configFileDefinedOptions,
        ...userDefinedOptions,
    };
    // Re-set log level again after config file has been read,
    // as it may have changed the log level
    setLogLevel(options.verbose);
    if (options.printConfig) {
        logger_1.logger.log("\nOptions: ");
        logger_1.logger.log({ ...config_1.DEFAULT_CONFIG, ...options });
    }
    return options;
}
exports.configureOptions = configureOptions;
function setLogLevel(verbosity) {
    process.env.SWA_CLI_DEBUG = verbosity;
    if (verbosity === null || verbosity === void 0 ? void 0 : verbosity.includes("silly")) {
        // When silly level is set,
        // propagate debugging level to other tools using the DEBUG environment variable
        process.env.DEBUG = "*";
    }
}
function getUserOptions(command) {
    const userOptions = {};
    const options = command.optsWithGlobals();
    for (const option in options) {
        // If the option is not found in the command context, it returns undefined
        // meaning that we have to find its source in the global context.
        const source = command.getOptionValueSource(option) || commander_1.program.getOptionValueSource(option);
        if (source === "cli") {
            userOptions[option] = options[option];
        }
    }
    return userOptions;
}
exports.getUserOptions = getUserOptions;
function isUserOption(option) {
    return userDefinedOptions[option] !== undefined;
}
exports.isUserOption = isUserOption;
function isConfigFileOption(option) {
    return configFileDefinedOptions[option] !== undefined;
}
exports.isConfigFileOption = isConfigFileOption;
function isUserOrConfigOption(option) {
    return isUserOption(option) || isConfigFileOption(option);
}
exports.isUserOrConfigOption = isUserOrConfigOption;
