// @scalar/api-reference/dist/features/Operation/components/Header.vue.js
import { defineComponent as r, createBlock as o, createCommentVNode as a, openBlock as d, unref as m } from "vue";
import { getResolvedRef as i } from "@scalar/workspace-store/helpers/get-resolved-ref";
import s from "../../../components/Content/Schema/SchemaProperty.vue.js";
const u = /* @__PURE__ */ r({
  __name: "Header",
  props: {
    header: {},
    name: {},
    breadcrumb: {},
    orderSchemaPropertiesBy: {},
    orderRequiredPropertiesFirst: { type: Boolean }
  },
  setup(p) {
    return (e, n) => "schema" in e.header && e.header.schema ? (d(), o(s, {
      key: 0,
      breadcrumb: e.breadcrumb ? [...e.breadcrumb, "headers"] : void 0,
      description: e.header.description,
      name: e.name,
      options: {
        orderRequiredPropertiesFirst: e.orderRequiredPropertiesFirst,
        orderSchemaPropertiesBy: e.orderSchemaPropertiesBy
      },
      schema: m(i)(e.header.schema)
    }, null, 8, ["breadcrumb", "description", "name", "options", "schema"])) : a("", !0);
  }
});
export {
  u as default
};
