// @scalar/api-reference/dist/features/Operation/components/Headers.vue2.js
import { defineComponent as h, createBlock as m, openBlock as r, unref as a, withCtx as o, createElementVNode as n, normalizeClass as d, createVNode as t, createElementBlock as i, Fragment as c, createTextVNode as p, renderList as f } from "vue";
import { Disclosure as b, DisclosureButton as B, DisclosurePanel as P } from "@headlessui/vue";
import { ScalarIcon as _ } from "@scalar/components";
import { getResolvedRef as y } from "@scalar/workspace-store/helpers/get-resolved-ref";
import v from "./Header.vue.js";
const C = /* @__PURE__ */ h({
  __name: "Headers",
  props: {
    headers: {},
    breadcrumb: {},
    orderRequiredPropertiesFirst: { type: Boolean },
    orderSchemaPropertiesBy: {}
  },
  setup(R) {
    return (e, S) => (r(), m(a(b), null, {
      default: o(({ open: s }) => [
        n("div", {
          class: d(["headers-card headers-card--compact", [{ "headers-card--open": s }]])
        }, [
          n("div", {
            class: d(["headers-properties", { "headers-properties-open": s }])
          }, [
            t(a(B), {
              class: "headers-card-title headers-card-title--compact",
              style: {
                top: "calc(var(--refs-viewport-offset)))"
              }
            }, {
              default: o(() => [
                t(a(_), {
                  class: d(["headers-card-title-icon", { "headers-card-title-icon--open": s }]),
                  icon: "Add",
                  size: "sm"
                }, null, 8, ["class"]),
                s ? (r(), i(c, { key: 0 }, [
                  p(" Hide Headers ")
                ], 64)) : (r(), i(c, { key: 1 }, [
                  p(" Show Headers ")
                ], 64))
              ]),
              _: 2
            }, 1024),
            t(a(P), null, {
              default: o(() => [
                (r(!0), i(c, null, f(e.headers, (u, l) => (r(), m(v, {
                  key: l,
                  breadcrumb: e.breadcrumb ? [...e.breadcrumb, "headers"] : void 0,
                  header: a(y)(u),
                  name: l,
                  orderRequiredPropertiesFirst: e.orderRequiredPropertiesFirst,
                  orderSchemaPropertiesBy: e.orderSchemaPropertiesBy
                }, null, 8, ["breadcrumb", "header", "name", "orderRequiredPropertiesFirst", "orderSchemaPropertiesBy"]))), 128))
              ]),
              _: 1
            })
          ], 2)
        ], 2)
      ]),
      _: 1
    }));
  }
});
export {
  C as default
};
