// @scalar/api-reference/dist/features/Operation/components/OperationParameters.vue.js
import { defineComponent as i, createElementBlock as p, openBlock as b, Fragment as u, createVNode as m, createBlock as n, createCommentVNode as l, withCtx as o, createTextVNode as t } from "vue";
import s from "./ParameterList.vue.js";
import y from "./RequestBody.vue.js";
const k = /* @__PURE__ */ i({
  __name: "OperationParameters",
  props: {
    breadcrumb: {},
    parameters: { default: () => [] },
    requestBody: {},
    options: {}
  },
  setup(d) {
    const a = (e) => d.parameters?.filter((r) => r.in === e) ?? [];
    return (e, r) => (b(), p(u, null, [
      m(s, {
        breadcrumb: e.breadcrumb ? [...e.breadcrumb, "path"] : void 0,
        options: e.options,
        parameters: a("path")
      }, {
        title: o(() => r[0] || (r[0] = [
          t("Path Parameters")
        ])),
        _: 1
      }, 8, ["breadcrumb", "options", "parameters"]),
      m(s, {
        breadcrumb: e.breadcrumb ? [...e.breadcrumb, "query"] : void 0,
        options: e.options,
        parameters: a("query")
      }, {
        title: o(() => r[1] || (r[1] = [
          t("Query Parameters")
        ])),
        _: 1
      }, 8, ["breadcrumb", "options", "parameters"]),
      m(s, {
        breadcrumb: e.breadcrumb ? [...e.breadcrumb, "headers"] : void 0,
        options: e.options,
        parameters: a("header")
      }, {
        title: o(() => r[2] || (r[2] = [
          t("Headers")
        ])),
        _: 1
      }, 8, ["breadcrumb", "options", "parameters"]),
      m(s, {
        breadcrumb: e.breadcrumb ? [...e.breadcrumb, "cookies"] : void 0,
        options: e.options,
        parameters: a("cookie")
      }, {
        title: o(() => r[3] || (r[3] = [
          t("Cookies")
        ])),
        _: 1
      }, 8, ["breadcrumb", "options", "parameters"]),
      e.requestBody ? (b(), n(y, {
        key: 0,
        breadcrumb: e.breadcrumb ? [...e.breadcrumb, "body"] : void 0,
        options: e.options,
        requestBody: e.requestBody
      }, {
        title: o(() => r[4] || (r[4] = [
          t("Body")
        ])),
        _: 1
      }, 8, ["breadcrumb", "options", "requestBody"])) : l("", !0)
    ], 64));
  }
});
export {
  k as default
};
