// @scalar/api-reference/dist/features/Operation/components/OperationResponses.vue.js
import { defineComponent as a, createElementBlock as s, createCommentVNode as p, openBlock as r, createElementVNode as n, Fragment as i, renderList as l, createBlock as d, unref as c } from "vue";
import { getResolvedRef as u } from "@scalar/workspace-store/helpers/get-resolved-ref";
import b from "./ParameterListItem.vue.js";
const f = {
  key: 0,
  class: "mt-6"
}, k = { class: "mb-3 list-none p-0 text-sm" }, y = /* @__PURE__ */ a({
  __name: "OperationResponses",
  props: {
    responses: {},
    breadcrumb: {},
    options: {}
  },
  setup(_) {
    return (e, t) => Object.keys(e.responses ?? {}).length ? (r(), s("div", f, [
      t[0] || (t[0] = n("div", { class: "text-c-1 mt-3 mb-3 leading-[1.45] font-medium" }, "Responses", -1)),
      n("ul", k, [
        (r(!0), s(i, null, l(e.responses, (m, o) => (r(), d(b, {
          key: o,
          breadcrumb: e.breadcrumb,
          name: o,
          options: e.options,
          parameter: c(u)(m)
        }, null, 8, ["breadcrumb", "name", "options", "parameter"]))), 128))
      ])
    ])) : p("", !0);
  }
});
export {
  y as default
};
