// @scalar/api-reference/dist/features/Operation/components/ParameterList.vue.js
import { defineComponent as n, createElementBlock as a, createCommentVNode as m, openBlock as r, createElementVNode as o, renderSlot as s, Fragment as l, renderList as p, createBlock as c } from "vue";
import i from "./ParameterListItem.vue.js";
const d = {
  key: 0,
  class: "mt-6"
}, u = { class: "text-c-1 mt-3 mb-3 text-lg leading-[1.45] font-medium" }, b = { class: "mb-3 list-none p-0 text-sm" }, g = /* @__PURE__ */ n({
  __name: "ParameterList",
  props: {
    parameters: {},
    breadcrumb: {},
    options: {}
  },
  setup(_) {
    return (e, f) => e.parameters?.length ? (r(), a("div", d, [
      o("div", u, [
        s(e.$slots, "title")
      ]),
      o("ul", b, [
        (r(!0), a(l, null, p(e.parameters, (t) => (r(), c(i, {
          key: t.name,
          breadcrumb: e.breadcrumb,
          name: t.name,
          options: e.options,
          parameter: t
        }, null, 8, ["breadcrumb", "name", "options", "parameter"]))), 128))
      ])
    ])) : m("", !0);
  }
});
export {
  g as default
};
