// @scalar/api-reference/dist/features/Operation/components/ParameterListItem.vue2.js
import { defineComponent as w, computed as m, ref as B, createElementBlock as E, openBlock as i, createVNode as u, createElementVNode as c, unref as s, withCtx as f, createBlock as d, createCommentVNode as h, normalizeClass as b, toDisplayString as S } from "vue";
import { Disclosure as q, DisclosureButton as R, DisclosurePanel as C } from "@headlessui/vue";
import { ScalarMarkdown as V } from "@scalar/components";
import { isDefined as x } from "@scalar/helpers/array/is-defined";
import { ScalarIconCaretRight as D } from "@scalar/icons";
import { getResolvedRef as y } from "@scalar/workspace-store/helpers/get-resolved-ref";
import F from "../../../components/Content/Schema/SchemaProperty.vue.js";
import O from "./ContentTypeSelect.vue.js";
import I from "./Headers.vue.js";
const N = { class: "parameter-item group/parameter-item relative" }, j = { class: "parameter-item-name" }, z = { class: "parameter-item-type" }, W = { class: "absolute top-3 right-0 opacity-0 group-focus-within/parameter-item:opacity-100 group-hover/parameter-item:opacity-100" }, Q = /* @__PURE__ */ w({
  __name: "ParameterListItem",
  props: {
    parameter: {},
    name: {},
    breadcrumb: {},
    options: {}
  },
  setup(e) {
    const l = m(
      () => "schema" in e.parameter && e.parameter.schema ? y(e.parameter.schema) : null
    ), a = m(
      () => "content" in e.parameter && e.parameter.content ? e.parameter.content : null
    ), n = B(Object.keys(a.value || {})[0]), v = m(
      () => "headers" in e.parameter && e.parameter.headers ? e.parameter.headers : null
    ), P = m(() => {
      const r = a.value ? a.value?.[n.value]?.schema : l.value, p = "deprecated" in e.parameter ? e.parameter.deprecated : l.value?.deprecated, t = "examples" in e.parameter ? e.parameter.examples : {}, g = l.value?.examples ?? [], k = [...Object.values({
        ...t,
        ...a.value?.[n.value]?.examples
      }), ...g];
      return {
        ...y(r),
        deprecated: p,
        ..."example" in e.parameter && x(e.parameter.example) && { example: e.parameter.example },
        examples: k
      };
    }), o = m(
      () => !!(e.options.collapsableItems && (a.value || v.value || l.value))
    );
    return (r, p) => (i(), E("li", N, [
      u(s(q), null, {
        default: f(({ open: t }) => [
          o.value ? (i(), d(s(R), {
            key: 0,
            class: b(["parameter-item-trigger", { "parameter-item-trigger-open": t }])
          }, {
            default: f(() => [
              c("span", j, [
                u(s(D), {
                  class: b(["parameter-item-icon size-3 transition-transform duration-100", { "rotate-90": t }]),
                  weight: "bold"
                }, null, 8, ["class"]),
                c("span", null, S(r.name), 1)
              ]),
              c("span", z, [
                r.parameter.description ? (i(), d(s(V), {
                  key: 0,
                  class: "markdown",
                  value: r.parameter.description
                }, null, 8, ["value"])) : h("", !0)
              ])
            ]),
            _: 2
          }, 1032, ["class"])) : h("", !0),
          u(s(C), {
            class: "parameter-item-container parameter-item-container-markdown",
            static: !o.value
          }, {
            default: f(() => [
              v.value ? (i(), d(I, {
                key: 0,
                breadcrumb: r.breadcrumb,
                headers: v.value,
                orderRequiredPropertiesFirst: r.options.orderRequiredPropertiesFirst,
                orderSchemaPropertiesBy: r.options.orderSchemaPropertiesBy
              }, null, 8, ["breadcrumb", "headers", "orderRequiredPropertiesFirst", "orderSchemaPropertiesBy"])) : h("", !0),
              u(F, {
                is: "div",
                breadcrumb: r.breadcrumb,
                compact: "",
                description: o.value ? "" : r.parameter.description,
                hideWriteOnly: !0,
                name: o.value ? "" : r.name,
                noncollapsible: !0,
                options: {
                  hideWriteOnly: !0,
                  orderRequiredPropertiesFirst: r.options.orderRequiredPropertiesFirst,
                  orderSchemaPropertiesBy: r.options.orderSchemaPropertiesBy
                },
                required: "required" in r.parameter && r.parameter.required,
                schema: P.value,
                withExamples: typeof r.options.withExamples == "boolean" ? r.options.withExamples : !0
              }, null, 8, ["breadcrumb", "description", "name", "options", "required", "schema", "withExamples"])
            ]),
            _: 1
          }, 8, ["static"])
        ]),
        _: 1
      }),
      c("div", W, [
        o.value && a.value ? (i(), d(O, {
          key: 0,
          modelValue: n.value,
          "onUpdate:modelValue": p[0] || (p[0] = (t) => n.value = t),
          class: "parameter-item-content-type",
          content: a.value
        }, null, 8, ["modelValue", "content"])) : h("", !0)
      ])
    ]));
  }
});
export {
  Q as default
};
