// @scalar/components/dist/components/ScalarForm/ScalarFormError.vue.js
import { defineComponent as c, computed as l, createElementBlock as s, openBlock as a, mergeProps as d, unref as n, createBlock as m, renderSlot as u, resolveDynamicComponent as f } from "vue";
import { ScalarIconWarningCircle as v, ScalarIconWarning as g } from "@scalar/icons";
import { cva as p, useBindCx as b } from "@scalar/use-hooks/useBindCx";
const C = /* @__PURE__ */ c({
  inheritAttrs: !1,
  __name: "ScalarFormError",
  props: {
    variant: { default: "error" },
    icon: {}
  },
  setup(r) {
    const t = l(() => r.icon ? r.icon : r.variant === "error" ? v : g), o = p({
      base: "flex items-baseline gap-1.5 rounded border p-2",
      variants: {
        variant: {
          error: "border-c-danger bg-b-danger text-c-danger",
          warning: "border-c-alert bg-b-alert text-c-alert"
        }
      }
    }), { cx: i } = b();
    return (e, x) => (a(), s("div", d({ role: "alert" }, n(i)(n(o)({ variant: e.variant }))), [
      (a(), m(f(t.value), {
        class: "relative top-0.5 shrink-0",
        weight: "bold"
      })),
      u(e.$slots, "default")
    ], 16));
  }
});
export {
  C as default
};
