// @scalar/api-client/dist/v2/blocks/scalar-response-block/components/Headers.vue.js
import { defineComponent as x, createBlock as f, openBlock as t, withCtx as e, createElementBlock as s, createVNode as n, unref as d, createTextVNode as l, Fragment as u, renderList as y, toDisplayString as i, renderSlot as h } from "vue";
import b from "../../../../components/DataTable/DataTable.vue.js";
import c from "../../../../components/DataTable/DataTableRow.vue.js";
import _ from "../../../../components/DataTable/DataTableText.vue.js";
import g from "../../../../components/HelpfulLink.vue.js";
import k from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import { httpHeaders as w } from "../../../../data/httpHeaders.js";
import p from "../../../../components/DataTable/DataTableHeader.vue.js";
const C = {
  key: 0,
  class: "max-h-[calc(100%-32px)] overflow-y-auto"
}, H = {
  key: 1,
  class: "text-c-3 bg-b-1 flex min-h-12 items-center justify-center rounded border px-4 text-base"
}, T = /* @__PURE__ */ x({
  __name: "Headers",
  props: {
    headers: {}
  },
  setup($) {
    const m = (r) => w.find(
      (o) => o.name.toLowerCase() === r.toLowerCase()
    );
    return (r, o) => (t(), f(k, {
      class: "overflow-auto",
      defaultOpen: !1,
      itemCount: r.headers.length
    }, {
      title: e(() => [
        h(r.$slots, "title")
      ]),
      default: e(() => [
        r.headers.length ? (t(), s("div", C, [
          n(b, {
            columns: ["minmax(auto, min-content)", "minmax(50%, 1fr)"],
            scroll: ""
          }, {
            default: e(() => [
              n(c, { class: "sr-only !block" }, {
                default: e(() => [
                  n(d(p), null, {
                    default: e(() => o[0] || (o[0] = [
                      l("Header Key")
                    ])),
                    _: 1,
                    __: [0]
                  }),
                  n(d(p), null, {
                    default: e(() => o[1] || (o[1] = [
                      l("Header Value")
                    ])),
                    _: 1,
                    __: [1]
                  })
                ]),
                _: 1
              }),
              (t(!0), s(u, null, y(r.headers, (a) => (t(), f(c, {
                key: a.name,
                class: "group/row text-c-1"
              }, {
                default: e(() => [
                  n(_, { class: "bg-b-1 sticky left-0 z-1 max-w-full group-first/row:border-t-0" }, {
                    default: e(() => [
                      typeof m(a.name)?.url == "string" ? (t(), f(g, {
                        key: 0,
                        class: "decoration-c-3",
                        href: m(a.name).url
                      }, {
                        default: e(() => [
                          l(i(a.name), 1)
                        ]),
                        _: 2
                      }, 1032, ["href"])) : (t(), s(u, { key: 1 }, [
                        l(i(a.name), 1)
                      ], 64))
                    ]),
                    _: 2
                  }, 1024),
                  n(_, {
                    class: "z-0 group-first/row:border-t-0",
                    text: a.value
                  }, null, 8, ["text"])
                ]),
                _: 2
              }, 1024))), 128))
            ]),
            _: 1
          })
        ])) : (t(), s("div", H, " No Headers "))
      ]),
      _: 3
    }, 8, ["itemCount"]));
  }
});
export {
  T as default
};
