// @scalar/api-client/dist/v2/blocks/scalar-response-block/components/ResponseBody.vue2.js
import { defineComponent as w, ref as g, computed as l, createBlock as n, openBlock as t, createSlots as _, withCtx as i, createElementBlock as x, createCommentVNode as r, createElementVNode as p, toDisplayString as v, createTextVNode as c, createVNode as B } from "vue";
import k from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import { processResponseBody as R } from "../helpers/process-response-body.js";
import { getMediaTypeConfig as T } from "../helpers/media-types.js";
import V from "./ResponseBodyDownload.vue.js";
import U from "./ResponseBodyInfo.vue.js";
import b from "./ResponseBodyPreview.vue.js";
import C from "./ResponseBodyRaw.vue.js";
import N from "./ResponseBodyToggle.vue.js";
const $ = {
  key: 0,
  class: "bg-b-1 flex max-h-[calc(100%-32px)] flex-col overflow-hidden"
}, E = { class: "box-content flex min-h-8 items-center justify-between border-y px-3" }, P = { class: "text-xxs font-code leading-3" }, G = /* @__PURE__ */ w({
  __name: "ResponseBody",
  props: {
    title: {},
    layout: {},
    data: {},
    headers: {}
  },
  setup(d) {
    const s = g(!0), m = l(
      () => !!(a.value?.raw && a.value.preview)
    ), f = l(() => s.value || !m.value), y = l(() => !s.value || !m.value), e = l(
      () => R({
        data: d.data,
        headers: d.headers
      })
    ), a = l(
      () => T(e.value.mimeType?.essence ?? "")
    );
    return (o, u) => (t(), n(k, {
      class: "max-h-content overflow-y-hidden",
      layout: o.layout
    }, _({
      title: i(() => [
        c(v(o.title), 1)
      ]),
      default: i(() => [
        o.data ? (t(), x("div", $, [
          p("div", E, [
            p("span", P, v(e.value.mimeType?.essence), 1),
            m.value ? (t(), n(N, {
              key: 0,
              modelValue: s.value,
              "onUpdate:modelValue": u[0] || (u[0] = (h) => s.value = h)
            }, null, 8, ["modelValue"])) : r("", !0)
          ]),
          a.value?.raw && y.value ? (t(), n(C, {
            key: e.value.dataUrl,
            content: o.data,
            language: a.value.language
          }, null, 8, ["content", "language"])) : r("", !0),
          a.value?.preview && f.value ? (t(), n(b, {
            key: e.value.dataUrl,
            alpha: a.value.alpha,
            mode: a.value.preview,
            src: e.value.dataUrl,
            type: e.value.mimeType?.essence ?? ""
          }, null, 8, ["alpha", "mode", "src", "type"])) : r("", !0),
          !a.value?.raw && !a.value?.preview ? (t(), n(U, { key: 2 }, {
            default: i(() => u[1] || (u[1] = [
              c(" Binary file ")
            ])),
            _: 1,
            __: [1]
          })) : r("", !0)
        ])) : r("", !0)
      ]),
      _: 2
    }, [
      o.data && e.value.dataUrl ? {
        name: "actions",
        fn: i(() => [
          B(V, {
            filename: e.value.attachmentFilename,
            href: e.value.dataUrl,
            type: e.value.mimeType?.essence
          }, null, 8, ["filename", "href", "type"])
        ]),
        key: "0"
      } : void 0
    ]), 1032, ["layout"]));
  }
});
export {
  G as default
};
