// @scalar/api-client/dist/v2/blocks/scalar-response-block/components/ResponseBodyVirtual.vue.js
import { defineComponent as l, computed as r, createBlock as i, openBlock as d, createSlots as m, withCtx as a, createElementVNode as c, createVNode as n, unref as f, createTextVNode as p } from "vue";
import { ScalarVirtualText as u } from "@scalar/components";
import { formatJsonOrYamlString as x } from "@scalar/oas-utils/helpers";
import h from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import { processResponseBody as y } from "../helpers/process-response-body.js";
import v from "./ResponseBodyDownload.vue.js";
const S = /* @__PURE__ */ l({
  __name: "ResponseBodyVirtual",
  props: {
    content: {},
    data: {},
    headers: {}
  },
  setup(o) {
    const s = r(() => x(o.content)), e = r(
      () => y({
        data: o.data,
        headers: o.headers
      })
    );
    return (b, t) => (d(), i(h, { class: "!max-h-100% response-body-virtual overflow-x-auto" }, m({
      title: a(() => [
        t[0] || (t[0] = p("Body"))
      ]),
      default: a(() => [
        t[1] || (t[1] = c("div", { class: "font-code text-xxs rounded-t border border-b-0 px-2.5 py-1.5" }, " This response body is massive! Syntax highlighting won't work here. ", -1)),
        n(f(u), {
          containerClass: "custom-scroll scalar-code-block border rounded-b flex flex-1 max-h-screen",
          contentClass: "language-plaintext whitespace-pre font-code text-base",
          lineHeight: 20,
          text: s.value
        }, null, 8, ["text"])
      ]),
      _: 2
    }, [
      e.value.dataUrl ? {
        name: "actions",
        fn: a(() => [
          n(v, {
            filename: e.value.attachmentFilename,
            href: e.value.dataUrl,
            type: e.value.mimeType?.essence
          }, null, 8, ["filename", "href", "type"])
        ]),
        key: "0"
      } : void 0
    ]), 1024));
  }
});
export {
  S as default
};
