// @scalar/api-client/dist/v2/blocks/scalar-response-block/components/ResponseCookies.vue.js
import { defineComponent as c, createBlock as r, openBlock as e, withCtx as t, createElementBlock as s, Fragment as i, renderList as f, createVNode as a, createTextVNode as u } from "vue";
import p from "../../../../components/DataTable/DataTable.vue.js";
import k from "../../../../components/DataTable/DataTableRow.vue.js";
import m from "../../../../components/DataTable/DataTableText.vue.js";
import _ from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
const d = {
  key: 1,
  class: "text-c-3 bg-b-1 flex min-h-[65px] items-center justify-center border-t px-4 text-sm"
}, B = /* @__PURE__ */ c({
  __name: "ResponseCookies",
  props: {
    cookies: {}
  },
  setup(x) {
    return (o, l) => (e(), r(_, {
      defaultOpen: !1,
      itemCount: o.cookies.length
    }, {
      title: t(() => l[0] || (l[0] = [
        u("Cookies")
      ])),
      default: t(() => [
        (e(), s(i, { key: 0 }, [
          o.cookies.length ? (e(), r(p, {
            key: 0,
            class: "flex-1",
            columns: ["", ""]
          }, {
            default: t(() => [
              (e(!0), s(i, null, f(o.cookies, (n) => (e(), r(k, {
                key: n.name
              }, {
                default: t(() => [
                  a(m, {
                    text: n.name
                  }, null, 8, ["text"]),
                  a(m, {
                    text: n.value
                  }, null, 8, ["text"])
                ]),
                _: 2
              }, 1024))), 128))
            ]),
            _: 1
          })) : (e(), s("div", d, " No cookies "))
        ], 64))
      ]),
      _: 1
    }, 8, ["itemCount"]));
  }
});
export {
  B as default
};
