// @scalar/api-client/dist/v2/blocks/scalar-response-block/components/ResponseBlock.vue2.js
import { defineComponent as S, computed as i, ref as I, useId as E, createBlock as a, openBlock as o, withCtx as d, createElementVNode as m, normalizeClass as V, createElementBlock as y, createVNode as A, Fragment as k, createCommentVNode as u, createTextVNode as g, renderList as O, unref as B, withDirectives as z, resolveDynamicComponent as M, mergeProps as N, vShow as L } from "vue";
import { ScalarErrorBoundary as j } from "@scalar/components";
import D from "../../../../components/SectionFilter.vue.js";
import F from "../../../../components/ViewLayout/ViewLayoutSection.vue.js";
import C from "./Headers.vue.js";
import { textMediaTypes as w } from "../helpers/media-types.js";
import U from "./ResponseBody.vue.js";
import K from "./ResponseBodyStreaming.vue.js";
import Z from "./ResponseBodyVirtual.vue.js";
/* empty css                         */
import G from "./ResponseCookies.vue.js";
import J from "./ResponseEmpty.vue.js";
import Q from "./ResponseLoadingOverlay.vue.js";
import W from "./ResponseMetaInformation.vue.js";
import { usePluginManager as X } from "../../../../plugins/hooks/usePluginManager.js";
const Y = { class: "flex h-8 flex-1 items-center" }, x = ["id", "role"], b = 2e5, ce = /* @__PURE__ */ S({
  __name: "ResponseBlock",
  props: {
    response: {},
    request: {},
    layout: {},
    totalPerformedRequests: {},
    appVersion: {},
    events: {}
  },
  emits: ["addRequest", "sendRequest", "openCommandPalette"],
  setup(t, { expose: H, emit: P }) {
    const f = P, T = X().getViewComponents("response.section"), p = i(() => {
      const e = t.response?.headers;
      return e ? Object.keys(e).map((s) => ({
        name: s,
        value: e[s] ?? ""
      })) : [];
    }), R = i(
      () => t.response?.cookieHeaderKeys.flatMap((e) => {
        const s = t.response?.headers?.[e];
        return s ? {
          name: e,
          value: s
        } : [];
      }) ?? []
    ), $ = ["Cookies", "Headers", "Body"], n = I("All"), v = i(() => ["All", ...$]), l = i(
      () => Object.fromEntries(
        v.value.map((e) => [e, E()])
      )
    ), h = i(() => {
      if (!t.response || !("size" in t.response))
        return !1;
      const e = t.response.headers?.["content-type"] || t.response.headers?.["Content-Type"];
      return !e || (t.response.size ?? 0) <= b || e.includes("text/html") ? !1 : w.some((r) => e.includes(r)) && (t.response.size ?? 0) > b;
    }), q = i(
      () => t.request?.headers ? [...t.request.headers].map((e) => ({
        name: e[0],
        value: e[1],
        required: !1
      })) : []
    ), c = (e) => e === "All" || n.value === e;
    return H({
      responseHeaders: p,
      responseCookies: R,
      requestHeaders: q,
      shouldVirtualize: h,
      activeFilter: n,
      filters: v
    }), (e, s) => (o(), a(F, { "aria-label": "Response" }, {
      title: d(() => [
        m("div", Y, [
          m("div", {
            "aria-live": "polite",
            class: V(["flex items-center", { "animate-response-heading": e.response }])
          }, [
            s[4] || (s[4] = m("span", { class: "response-heading pointer-events-none absolute" }, " Response ", -1)),
            e.response ? (o(), a(W, {
              key: 0,
              class: "animate-response-children",
              events: e.events,
              response: e.response
            }, null, 8, ["events", "response"])) : u("", !0)
          ], 2),
          A(D, {
            modelValue: n.value,
            "onUpdate:modelValue": s[0] || (s[0] = (r) => n.value = r),
            filterIds: l.value,
            filters: v.value
          }, null, 8, ["modelValue", "filterIds", "filters"])
        ])
      ]),
      default: d(() => [
        m("div", {
          id: l.value.All,
          class: V(["custom-scroll response-section-content relative grid h-full justify-stretch", {
            "content-start": e.response
          }]),
          role: n.value === "All" && e.response ? "tabpanel" : "none"
        }, [
          e.response ? (o(), y(k, { key: 1 }, [
            c("Cookies") ? (o(), a(G, {
              key: 0,
              id: l.value.Cookies,
              class: "response-section-content-cookies",
              cookies: R.value,
              role: n.value === "All" ? "none" : "tabpanel"
            }, null, 8, ["id", "cookies", "role"])) : u("", !0),
            c("Headers") ? (o(), a(C, {
              key: 1,
              id: l.value.Headers,
              class: "response-section-content-headers",
              headers: q.value,
              role: n.value === "All" ? "none" : "tabpanel"
            }, {
              title: d(() => s[5] || (s[5] = [
                g("Request Headers")
              ])),
              _: 1
            }, 8, ["id", "headers", "role"])) : u("", !0),
            c("Headers") ? (o(), a(C, {
              key: 2,
              id: l.value.Headers,
              class: "response-section-content-headers",
              headers: p.value,
              role: n.value === "All" ? "none" : "tabpanel"
            }, {
              title: d(() => s[6] || (s[6] = [
                g("Response Headers")
              ])),
              _: 1
            }, 8, ["id", "headers", "role"])) : u("", !0),
            (o(!0), y(k, null, O(B(T), (r) => (o(), a(B(j), {
              key: r.component
            }, {
              default: d(() => [
                z((o(), a(M(r.component), N({ ref_for: !0 }, r.props ?? {}), null, 16)), [
                  [L, n.value === "All" || n.value === r.title]
                ])
              ]),
              _: 2
            }, 1024))), 128)),
            n.value === "All" || n.value === "Body" ? (o(), y(k, { key: 3 }, [
              "reader" in e.response ? (o(), a(K, {
                key: 0,
                id: l.value.Body,
                class: "response-section-content-body",
                reader: e.response.reader
              }, null, 8, ["id", "reader"])) : h.value && typeof e.response?.data == "string" ? (o(), a(Z, {
                key: 1,
                id: l.value.Body,
                content: e.response.data,
                data: e.response?.data,
                headers: p.value,
                role: n.value === "All" ? "none" : "tabpanel"
              }, null, 8, ["id", "content", "data", "headers", "role"])) : (o(), a(U, {
                key: 2,
                id: l.value.Body,
                active: !0,
                class: "response-section-content-body",
                data: e.response?.data,
                headers: p.value,
                layout: "client",
                role: n.value === "All" ? "none" : "tabpanel",
                title: "Body"
              }, null, 8, ["id", "data", "headers", "role"]))
            ], 64)) : u("", !0)
          ], 64)) : (o(), a(J, {
            key: 0,
            appVersion: e.appVersion,
            events: e.events,
            layout: e.layout,
            totalPerformedRequests: e.totalPerformedRequests,
            onAddRequest: s[1] || (s[1] = (r) => f("addRequest")),
            onOpenCommandPalette: s[2] || (s[2] = (r) => f("openCommandPalette")),
            onSendRequest: s[3] || (s[3] = (r) => f("sendRequest"))
          }, null, 8, ["appVersion", "events", "layout", "totalPerformedRequests"])),
          A(Q, { events: e.events }, null, 8, ["events"])
        ], 10, x)
      ]),
      _: 1
    }));
  }
});
export {
  ce as default
};
