// @scalar/components/dist/components/ScalarSearchInput/ScalarSearchInput.vue.js
import { defineComponent as p, mergeModels as f, useModel as m, ref as g, createElementBlock as h, openBlock as t, normalizeProps as v, guardReactiveProps as b, unref as e, createVNode as k, withDirectives as _, createBlock as n, createCommentVNode as C, createElementVNode as S, mergeProps as y, vModelText as V, withModifiers as M } from "vue";
import { ScalarIconMagnifyingGlass as x } from "@scalar/icons";
import { useBindCx as z } from "@scalar/use-hooks/useBindCx";
import B from "../ScalarLoading/ScalarLoading.vue.js";
import E from "../ScalarIconButton/ScalarIconButton.vue.js";
const N = ["aria-label"], P = {}, q = /* @__PURE__ */ p({
  ...P,
  inheritAttrs: !1,
  __name: "ScalarSearchInput",
  props: /* @__PURE__ */ f({
    loading: {},
    label: {}
  }, {
    modelValue: {},
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(s) {
    const a = m(s, "modelValue"), o = g(null);
    function i() {
      a.value = "", o.value && o.value.focus();
    }
    const { classCx: c, otherAttrs: d } = z();
    return (l, r) => (t(), h("label", v(b(
      e(c)(
        "flex items-center text-lg font-medium h-10 pl-3 pr-1 py-2 gap-2.25"
      )
    )), [
      k(e(x), { class: "text-sidebar-c-search size-4" }),
      _(S("input", y({
        ref_key: "inputRef",
        ref: o,
        "aria-label": l.label ?? "Enter search query",
        autocapitalize: "off",
        autocomplete: "off",
        autocorrect: "off",
        class: "flex-1 appearance-none rounded border-none bg-transparent outline-none",
        placeholder: "Search...",
        spellcheck: "false",
        type: "search"
      }, e(d), {
        "onUpdate:modelValue": r[0] || (r[0] = (u) => a.value = u)
      }), null, 16, N), [
        [V, a.value]
      ]),
      l.loading && l.loading.isLoading ? (t(), n(e(B), {
        key: 0,
        class: "self-center",
        loadingState: l.loading,
        size: "md"
      }, null, 8, ["loadingState"])) : a.value ? (t(), n(e(E), {
        key: 1,
        class: "p-0 size-5",
        icon: "Close",
        label: "Clear Search",
        thickness: "1.5",
        onClick: M(i, ["stop", "prevent"])
      })) : C("", !0)
    ], 16));
  }
});
export {
  q as default
};
