// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/components/AuthSelector.vue2.js
import { defineComponent as M, useId as N, ref as b, computed as S, createBlock as x, openBlock as r, unref as o, withCtx as u, createVNode as m, createCommentVNode as g, createElementVNode as i, createElementBlock as v, Fragment as k, createTextVNode as C, toDisplayString as h, withModifiers as U, normalizeClass as j } from "vue";
import { useModal as z, ScalarComboboxMultiselect as E, ScalarButton as F, ScalarListboxCheckbox as P, ScalarIconButton as G } from "@scalar/components";
import { ScalarIconCaretDown as H, ScalarIconTrash as J } from "@scalar/icons";
import { isDefined as B } from "@scalar/oas-utils/helpers";
import { getResolvedRef as K } from "@scalar/workspace-store/helpers/get-resolved-ref";
import Q from "./DeleteRequestAuthModal.vue.js";
import { formatComplexScheme as W, formatScheme as X, getSecuritySchemeOptions as Y } from "../helpers/security-scheme.js";
import Z from "./RequestAuthDataTable.vue.js";
import _ from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import { useLayout as ee } from "../../../../hooks/useLayout.js";
const te = ["id"], le = { class: "flex flex-1" }, oe = { class: "min-w-0 flex-1 truncate" }, ye = /* @__PURE__ */ M({
  __name: "AuthSelector",
  props: {
    environment: {},
    envVariables: {},
    layout: {},
    security: {},
    selectedSecurity: {},
    securitySchemes: {},
    server: {},
    title: {}
  },
  emits: ["deleteOperationAuth", "update:securityScheme", "update:selectedScopes", "update:selectedSecurity"],
  setup(n, { expose: I, emit: T }) {
    const d = T, { layout: O } = ee(), V = N(), w = b(null), p = z(), a = b(null), A = b(!1), y = S(() => {
      if (!n.security?.length)
        return null;
      const e = !n.security.some(
        (f) => Object.keys(f).length > 1
      ) && n.security.some((f) => Object.keys(f).length === 0);
      return { icon: e ? "Unlock" : "Lock", text: e ? "Optional" : "Required" };
    }), s = S(() => n.selectedSecurity?.length ? n.selectedSecurity.map((t) => {
      const e = Object.keys(t);
      if (e.length > 1)
        return W(t);
      const l = e[0];
      if (!l)
        return;
      const c = K(n.securitySchemes?.[l]);
      if (c)
        return X({ name: l, type: c.type, value: { [l]: [] } });
    }).filter(B) : []);
    function D({
      label: t,
      value: e
    }) {
      a.value = { label: t, payload: e }, p.show();
    }
    const R = S(() => Y(
      n.security ?? [],
      n.securitySchemes ?? {},
      O === "modal" || n.layout === "reference"
    )), $ = (t) => {
      A.value && t.stopPropagation(), w.value?.$el.click();
    }, q = (t) => {
      d("update:selectedSecurity", {
        value: t.map((e) => e.value),
        create: t.map((e) => e.payload).filter(B)
      });
    }, L = () => {
      a.value && (d("deleteOperationAuth", Object.keys(a.value.payload)), a.value = null, p.hide());
    };
    return I({
      authIndicator: y,
      selectedSchemeOptions: s,
      schemeOptions: R
    }), (t, e) => (r(), x(o(_), {
      class: "group/params relative",
      itemCount: s.value.length,
      layout: t.layout,
      "onUpdate:modelValue": e[3] || (e[3] = (l) => A.value = l)
    }, {
      title: u(() => [
        i("div", {
          id: o(V),
          class: "inline-flex items-center gap-0.5 leading-[20px]"
        }, [
          i("span", null, h(t.title), 1),
          y.value ? (r(), v("span", {
            key: 0,
            class: j(["text-c-3 hover:bg-b-3 hover:text-c-1 -mr-1 cursor-pointer rounded px-1 py-0.5 text-xs leading-[normal]", { "text-c-1": y.value.text === "Required" }]),
            "data-testid": "auth-indicator",
            onClick: $
          }, h(y.value.text), 3)) : g("", !0)
        ], 8, te)
      ]),
      actions: u(() => [
        i("div", le, [
          m(o(E), {
            class: "w-72 text-xs",
            modelValue: s.value,
            multiple: "",
            options: R.value,
            placement: "bottom-end",
            teleport: "",
            onDelete: D,
            "onUpdate:modelValue": q
          }, {
            option: u(({ option: l, selected: c }) => [
              m(o(P), {
                multiselect: "",
                selected: c
              }, null, 8, ["selected"]),
              i("div", oe, h(l.label), 1),
              l.isDeletable ?? (o(O) !== "modal" && t.layout !== "reference") ? (r(), x(o(G), {
                key: 0,
                class: "-m-0.5 shrink-0 p-0.5 opacity-0 group-hover/item:opacity-100",
                icon: o(J),
                label: `Delete ${l.label}`,
                size: "xs",
                onClick: U((f) => D(l), ["stop"])
              }, null, 8, ["icon", "label", "onClick"])) : g("", !0)
            ]),
            default: u(() => [
              m(o(F), {
                ref_key: "comboboxButtonRef",
                ref: w,
                "aria-describedby": o(V),
                class: "group/combobox-button hover:text-c-1 text-c-2 flex h-fit w-full items-center gap-1 px-0.75 py-0.25 text-base font-normal transition-transform",
                variant: "ghost"
              }, {
                default: u(() => [
                  s.value.length === 1 ? (r(), v(k, { key: 0 }, [
                    e[4] || (e[4] = i("span", { class: "sr-only" }, "Selected Auth Type:", -1)),
                    C(" " + h(s.value[0]?.label), 1)
                  ], 64)) : s.value.length > 1 ? (r(), v(k, { key: 1 }, [
                    e[5] || (e[5] = C(" Multiple ")),
                    e[6] || (e[6] = i("span", { class: "sr-only" }, "Auth Types Selected", -1))
                  ], 64)) : (r(), v(k, { key: 2 }, [
                    e[7] || (e[7] = i("span", { class: "sr-only" }, "Select", -1)),
                    e[8] || (e[8] = C(" Auth Type "))
                  ], 64)),
                  m(o(H), {
                    class: "size-3 shrink-0 transition-transform duration-100 group-aria-expanded/combobox-button:rotate-180",
                    weight: "bold"
                  })
                ]),
                _: 1
              }, 8, ["aria-describedby"])
            ]),
            _: 1
          }, 8, ["modelValue", "options"])
        ])
      ]),
      default: u(() => [
        m(Z, {
          envVariables: t.envVariables,
          environment: t.environment,
          layout: t.layout,
          securitySchemes: t.securitySchemes,
          selectedSchemeOptions: s.value,
          server: t.server,
          "onUpdate:securityScheme": e[0] || (e[0] = (l) => d("update:securityScheme", l)),
          "onUpdate:selectedScopes": e[1] || (e[1] = (l) => d("update:selectedScopes", l))
        }, null, 8, ["envVariables", "environment", "layout", "securitySchemes", "selectedSchemeOptions", "server"]),
        a.value ? (r(), x(Q, {
          key: 0,
          label: a.value.label,
          scheme: a.value,
          state: o(p),
          onClose: e[2] || (e[2] = (l) => o(p).hide()),
          onDelete: L
        }, null, 8, ["label", "scheme", "state"])) : g("", !0)
      ]),
      _: 1
    }, 8, ["itemCount", "layout"]));
  }
});
export {
  ye as default
};
