// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/components/DeleteRequestAuthModal.vue.js
import { defineComponent as m, createBlock as p, openBlock as f, unref as o, withCtx as n, createElementVNode as s, toDisplayString as a, createVNode as r, createTextVNode as i } from "vue";
import { ScalarModal as h, ScalarButton as c } from "@scalar/components";
const x = { class: "text-c-2 mb-4 text-sm leading-normal" }, _ = { class: "flex justify-between gap-2" }, C = /* @__PURE__ */ m({
  __name: "DeleteRequestAuthModal",
  props: {
    state: {},
    label: {}
  },
  emits: ["close", "delete"],
  setup(b, { emit: u }) {
    const l = u, d = () => {
      l("delete");
    };
    return (t, e) => (f(), p(o(h), {
      size: "xxs",
      state: t.state,
      title: "Delete Security Scheme"
    }, {
      default: n(() => [
        s("p", x, " This cannot be undone. You're about to delete the " + a(t.label) + " security scheme from the collection. ", 1),
        s("div", _, [
          r(o(c), {
            class: "flex h-8 cursor-pointer items-center gap-1.5 px-3 shadow-none focus:outline-none",
            type: "button",
            variant: "outlined",
            onClick: e[0] || (e[0] = (y) => l("close"))
          }, {
            default: n(() => e[1] || (e[1] = [
              i(" Cancel ")
            ])),
            _: 1,
            __: [1]
          }),
          r(o(c), {
            class: "flex h-8 cursor-pointer items-center gap-1.5 px-3 shadow-none focus:outline-none",
            type: "submit",
            onClick: d
          }, {
            default: n(() => [
              i(" Delete " + a(t.label), 1)
            ]),
            _: 1
          })
        ])
      ]),
      _: 1
    }, 8, ["state"]));
  }
});
export {
  C as default
};
