// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/components/OAuth2.vue.js
import { defineComponent as C, computed as T, createElementBlock as v, openBlock as u, Fragment as x, createBlock as y, createCommentVNode as f, createVNode as r, unref as o, withCtx as l, mergeProps as p, createTextVNode as d, createElementVNode as S } from "vue";
import { useLoadingState as z, ScalarButton as g } from "@scalar/components";
import { pkceOptions as P } from "@scalar/oas-utils/entities/spec";
import { useToasts as I } from "@scalar/use-toasts";
import L from "./OAuthScopesInput.vue.js";
import { authorizeOauth2 as R } from "../helpers/oauth.js";
import c from "./RequestAuthDataTableInput.vue.js";
import n from "../../../../components/DataTable/DataTableRow.vue.js";
const B = { class: "flex h-8 items-center justify-end gap-2 border-t" }, F = { class: "flex h-8 w-full items-center justify-end border-t" }, M = /* @__PURE__ */ C({
  __name: "OAuth2",
  props: {
    environment: {},
    envVariables: {},
    flows: {},
    type: {},
    selectedScopes: {},
    server: {},
    proxyUrl: {}
  },
  emits: ["update:selectedScopes", "update:securityScheme"],
  setup(i, { emit: w }) {
    const s = w, V = z(), { toast: U } = I(), b = async () => {
      if (V.isLoading)
        return;
      if (!i.server) {
        U("No server selected", "error");
        return;
      }
      V.startLoading();
      const [k, e] = await R(
        i.flows,
        i.type,
        i.selectedScopes,
        i.server,
        i.proxyUrl
      ).finally(() => V.stopLoading());
      e ? s("update:securityScheme", { token: e }) : (console.error(k), U(k?.message ?? "Failed to authorize", "error"));
    }, m = {
      environment: i.environment,
      envVariables: i.envVariables
    }, t = T(() => i.flows[i.type]);
    return (k, e) => (u(), v(x, null, [
      t.value["x-scalar-secret-token"] ? (u(), v(x, { key: 0 }, [
        r(o(n), null, {
          default: l(() => [
            r(c, p(m, {
              class: "border-r-transparent",
              modelValue: t.value["x-scalar-secret-token"],
              placeholder: "QUxMIFlPVVIgQkFTRSBBUkUgQkVMT05HIFRPIFVT",
              type: "password",
              "onUpdate:modelValue": e[0] || (e[0] = (a) => s("update:securityScheme", { token: a }))
            }), {
              default: l(() => e[11] || (e[11] = [
                d(" Access Token ")
              ])),
              _: 1,
              __: [11]
            }, 16, ["modelValue"])
          ]),
          _: 1
        }),
        r(o(n), { class: "min-w-full" }, {
          default: l(() => [
            S("div", B, [
              r(o(g), {
                class: "mr-1 p-0 px-2 py-0.5",
                loading: o(V),
                size: "sm",
                variant: "outlined",
                onClick: e[1] || (e[1] = (a) => s("update:securityScheme", { token: "" }))
              }, {
                default: l(() => e[12] || (e[12] = [
                  d(" Clear ")
                ])),
                _: 1,
                __: [12]
              }, 8, ["loading"])
            ])
          ]),
          _: 1
        })
      ], 64)) : (u(), v(x, { key: 1 }, [
        r(o(n), null, {
          default: l(() => [
            "authorizationUrl" in t.value ? (u(), y(c, p({ key: 0 }, m, {
              containerClass: "border-r-0",
              modelValue: t.value.authorizationUrl,
              placeholder: "https://galaxy.scalar.com/authorize",
              "onUpdate:modelValue": e[2] || (e[2] = (a) => s("update:securityScheme", { authUrl: a }))
            }), {
              default: l(() => e[13] || (e[13] = [
                d(" Auth URL ")
              ])),
              _: 1,
              __: [13]
            }, 16, ["modelValue"])) : f("", !0),
            "tokenUrl" in t.value ? (u(), y(c, p({ key: 1 }, m, {
              modelValue: t.value.tokenUrl,
              placeholder: "https://galaxy.scalar.com/token",
              "onUpdate:modelValue": e[3] || (e[3] = (a) => s("update:securityScheme", { tokenUrl: a }))
            }), {
              default: l(() => e[14] || (e[14] = [
                d(" Token URL ")
              ])),
              _: 1,
              __: [14]
            }, 16, ["modelValue"])) : f("", !0)
          ]),
          _: 1
        }),
        "x-scalar-secret-redirect-uri" in t.value ? (u(), y(o(n), { key: 0 }, {
          default: l(() => [
            r(c, p(m, {
              modelValue: t.value["x-scalar-secret-redirect-uri"],
              placeholder: "https://galaxy.scalar.com/callback",
              "onUpdate:modelValue": e[4] || (e[4] = (a) => s("update:securityScheme", { redirectUrl: a }))
            }), {
              default: l(() => e[15] || (e[15] = [
                d(" Redirect URL ")
              ])),
              _: 1,
              __: [15]
            }, 16, ["modelValue"])
          ]),
          _: 1
        })) : f("", !0),
        "x-scalar-secret-username" in t.value && "x-scalar-secret-password" in t.value ? (u(), v(x, { key: 1 }, [
          r(o(n), null, {
            default: l(() => [
              r(c, p(m, {
                class: "text-c-2",
                modelValue: t.value["x-scalar-secret-username"],
                placeholder: "janedoe",
                "onUpdate:modelValue": e[5] || (e[5] = (a) => s("update:securityScheme", { username: a }))
              }), {
                default: l(() => e[16] || (e[16] = [
                  d(" Username ")
                ])),
                _: 1,
                __: [16]
              }, 16, ["modelValue"])
            ]),
            _: 1
          }),
          r(o(n), null, {
            default: l(() => [
              r(c, p(m, {
                modelValue: t.value["x-scalar-secret-password"],
                placeholder: "********",
                type: "password",
                "onUpdate:modelValue": e[6] || (e[6] = (a) => s("update:securityScheme", { password: a }))
              }), {
                default: l(() => e[17] || (e[17] = [
                  d(" Password ")
                ])),
                _: 1,
                __: [17]
              }, 16, ["modelValue"])
            ]),
            _: 1
          })
        ], 64)) : f("", !0),
        r(o(n), null, {
          default: l(() => [
            r(c, p(m, {
              modelValue: t.value["x-scalar-secret-client-id"],
              placeholder: "12345",
              "onUpdate:modelValue": e[7] || (e[7] = (a) => s("update:securityScheme", { clientId: a }))
            }), {
              default: l(() => e[18] || (e[18] = [
                d(" Client ID ")
              ])),
              _: 1,
              __: [18]
            }, 16, ["modelValue"])
          ]),
          _: 1
        }),
        "x-scalar-secret-client-secret" in t.value ? (u(), y(o(n), { key: 2 }, {
          default: l(() => [
            r(c, p(m, {
              modelValue: t.value["x-scalar-secret-client-secret"],
              placeholder: "XYZ123",
              type: "password",
              "onUpdate:modelValue": e[8] || (e[8] = (a) => s("update:securityScheme", { clientSecret: a }))
            }), {
              default: l(() => e[19] || (e[19] = [
                d(" Client Secret ")
              ])),
              _: 1,
              __: [19]
            }, 16, ["modelValue"])
          ]),
          _: 1
        })) : f("", !0),
        "x-usePkce" in t.value ? (u(), y(o(n), { key: 3 }, {
          default: l(() => [
            r(c, p(m, {
              enum: o(P),
              modelValue: t.value["x-usePkce"],
              readOnly: "",
              "onUpdate:modelValue": e[9] || (e[9] = (a) => s("update:securityScheme", {
                usePkce: a
              }))
            }), {
              default: l(() => e[20] || (e[20] = [
                d(" Use PKCE ")
              ])),
              _: 1,
              __: [20]
            }, 16, ["enum", "modelValue"])
          ]),
          _: 1
        })) : f("", !0),
        Object.keys(t.value.scopes ?? {}).length ? (u(), y(o(n), { key: 4 }, {
          default: l(() => [
            r(L, {
              flow: t.value,
              selectedScopes: k.selectedScopes,
              "onUpdate:selectedScopes": e[10] || (e[10] = (a) => s("update:selectedScopes", a))
            }, null, 8, ["flow", "selectedScopes"])
          ]),
          _: 1
        })) : f("", !0)
      ], 64)),
      t.value["x-scalar-secret-token"] ? f("", !0) : (u(), y(o(n), {
        key: 2,
        class: "min-w-full"
      }, {
        default: l(() => [
          S("div", F, [
            r(o(g), {
              class: "mr-0.75 p-0 px-2 py-0.5",
              loading: o(V),
              size: "sm",
              variant: "outlined",
              onClick: b
            }, {
              default: l(() => e[21] || (e[21] = [
                d(" Authorize ")
              ])),
              _: 1,
              __: [21]
            }, 8, ["loading"])
          ])
        ]),
        _: 1
      }))
    ], 64));
  }
});
export {
  M as default
};
