// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/components/RequestAuthDataTableInput.vue.js
import { defineComponent as p, mergeModels as u, useModel as V, useId as f, createBlock as v, openBlock as y, unref as t, mergeProps as b, withCtx as s, renderSlot as i, createElementVNode as C } from "vue";
import B from "../../../../components/DataTable/DataTableInput.vue.js";
const q = ["for"], $ = /* @__PURE__ */ p({
  __name: "RequestAuthDataTableInput",
  props: /* @__PURE__ */ u({
    containerClass: {},
    environment: {},
    envVariables: {},
    readOnly: { type: Boolean, default: !1 },
    required: { type: Boolean, default: !1 },
    type: {}
  }, {
    modelValue: { default: "", required: !0 },
    modelModifiers: {}
  }),
  emits: /* @__PURE__ */ u(["inputFocus", "inputBlur", "selectVariable"], ["update:modelValue"]),
  setup(d, { emit: m }) {
    const a = m, o = V(d, "modelValue"), r = f();
    return (e, n) => (y(), v(t(B), b({ id: t(r) }, e.$attrs, {
      modelValue: o.value,
      "onUpdate:modelValue": n[0] || (n[0] = (l) => o.value = l),
      canAddCustomEnumValue: !e.readOnly,
      containerClass: e.containerClass,
      envVariables: e.envVariables,
      environment: e.environment,
      lineWrapping: "",
      readOnly: e.readOnly,
      required: e.required,
      type: e.type,
      onInputBlur: n[1] || (n[1] = (l) => a("inputBlur")),
      onInputFocus: n[2] || (n[2] = (l) => a("inputFocus")),
      onSelectVariable: n[3] || (n[3] = (l) => a("selectVariable", l))
    }), {
      default: s(() => [
        C("label", { for: t(r) }, [
          i(e.$slots, "default")
        ], 8, q)
      ]),
      icon: s(() => [
        i(e.$slots, "icon")
      ]),
      _: 3
    }, 16, ["id", "modelValue", "canAddCustomEnumValue", "containerClass", "envVariables", "environment", "readOnly", "required", "type"]));
  }
});
export {
  $ as default
};
