// @scalar/api-client/dist/v2/blocks/scalar-auth-selector-block/components/RequestAuthTab.vue.js
import { defineComponent as R, computed as N, ref as P, createElementBlock as u, openBlock as r, Fragment as i, renderList as U, createBlock as k, createCommentVNode as b, unref as s, withCtx as o, createVNode as a, createElementVNode as C, toDisplayString as F, mergeProps as v, createTextVNode as x, normalizeClass as j, capitalize as _ } from "vue";
import { ScalarMarkdownSummary as z } from "@scalar/components";
import { getResolvedRef as K } from "@scalar/workspace-store/helpers/get-resolved-ref";
import O from "./OAuth2.vue.js";
import w from "./RequestAuthDataTableInput.vue.js";
import f from "../../../../components/DataTable/DataTableRow.vue.js";
import B from "../../../../components/DataTable/DataTableCell.vue.js";
const L = { class: "bg-b-1 text-c-2 outline-b-3 top-0 z-1 h-full w-full overflow-hidden px-3 py-1.25 text-ellipsis group-hover/auth:absolute group-hover/auth:h-auto group-hover/auth:border-b *:first:line-clamp-1 *:first:text-ellipsis group-hover/auth:*:first:line-clamp-none" }, M = {
  key: 0,
  class: "flex min-h-8 border-t text-base"
}, Q = { class: "flex h-8 max-w-full gap-2.5 overflow-x-auto px-3" }, q = ["onClick"], D = { class: "relative z-10" }, Y = /* @__PURE__ */ R({
  __name: "RequestAuthTab",
  props: {
    environment: {},
    envVariables: {},
    layout: {},
    selectedSecuritySchema: {},
    securitySchemes: {},
    server: {}
  },
  emits: ["update:securityScheme", "update:selectedScopes"],
  setup(S, { emit: I }) {
    const c = I, $ = N(
      () => Object.entries(S.selectedSecuritySchema).map(([l, e]) => ({
        scheme: K(S.securitySchemes[l]),
        name: l,
        scopes: e
      }))
    ), p = P(""), T = (l, e) => {
      const t = e.description ? `: ${e.description}` : "", m = `${_(l)}${t || `: ${e.type}`}`;
      if (e.type === "apiKey")
        return `${_(l)}${t || `: ${e.in}`}`;
      if (e.type === "oauth2") {
        const h = Object.keys(e.flows ?? {})[0];
        return `${_(l)}: ${p.value ? p.value : h}${t}`;
      }
      return e.type === "http" ? `${_(l)}: ${e.scheme}${t}` : `${m}${t}`;
    }, y = {
      environment: S.environment,
      envVariables: S.envVariables
    };
    return (l, e) => (r(!0), u(i, null, U($.value, ({ scheme: t, name: m, scopes: h }) => (r(), u(i, { key: m }, [
      $.value.length > 1 && t ? (r(), k(s(f), { key: 0 }, {
        default: o(() => [
          a(s(B), {
            "aria-label": T(m, t),
            class: "text-c-2 group/auth flex items-center leading-[22px] whitespace-nowrap outline-none hover:whitespace-normal"
          }, {
            default: o(() => [
              C("p", L, F(T(m, t)), 1)
            ]),
            _: 2
          }, 1032, ["aria-label"])
        ]),
        _: 2
      }, 1024)) : b("", !0),
      t?.description && $.value.length <= 1 ? (r(), k(s(f), { key: 1 }, {
        default: o(() => [
          a(s(B), {
            "aria-label": t.description,
            class: "max-h-[auto]"
          }, {
            default: o(() => [
              a(s(z), {
                class: "auth-description bg-b-1 text-c-2 min-w-0 flex-1 px-3 py-1.25",
                value: t.description
              }, null, 8, ["value"])
            ]),
            _: 2
          }, 1032, ["aria-label"])
        ]),
        _: 2
      }, 1024)) : b("", !0),
      t?.type === "http" ? (r(), u(i, { key: 2 }, [
        t.scheme === "bearer" ? (r(), k(s(f), { key: 0 }, {
          default: o(() => [
            a(w, v({ ref_for: !0 }, y, {
              containerClass: l.layout === "reference" && "border-t",
              modelValue: t["x-scalar-secret-token"],
              placeholder: "Token",
              type: "password",
              "onUpdate:modelValue": e[0] || (e[0] = (n) => c("update:securityScheme", {
                type: "http",
                payload: { token: n }
              }))
            }), {
              default: o(() => e[5] || (e[5] = [
                x(" Bearer Token ")
              ])),
              _: 2,
              __: [5]
            }, 1040, ["containerClass", "modelValue"])
          ]),
          _: 2
        }, 1024)) : t?.scheme === "basic" ? (r(), u(i, { key: 1 }, [
          a(s(f), null, {
            default: o(() => [
              a(w, v({ ref_for: !0 }, y, {
                class: "text-c-2",
                modelValue: t["x-scalar-secret-username"],
                placeholder: "janedoe",
                required: "",
                "onUpdate:modelValue": e[1] || (e[1] = (n) => c("update:securityScheme", {
                  type: "http",
                  payload: { username: n }
                }))
              }), {
                default: o(() => e[6] || (e[6] = [
                  x(" Username ")
                ])),
                _: 2,
                __: [6]
              }, 1040, ["modelValue"])
            ]),
            _: 2
          }, 1024),
          a(s(f), null, {
            default: o(() => [
              a(w, v({ ref_for: !0 }, y, {
                modelValue: t["x-scalar-secret-password"],
                placeholder: "********",
                type: "password",
                "onUpdate:modelValue": e[2] || (e[2] = (n) => c("update:securityScheme", {
                  type: "http",
                  payload: { password: n }
                }))
              }), {
                default: o(() => e[7] || (e[7] = [
                  x(" Password ")
                ])),
                _: 2,
                __: [7]
              }, 1040, ["modelValue"])
            ]),
            _: 2
          }, 1024)
        ], 64)) : b("", !0)
      ], 64)) : t?.type === "apiKey" ? (r(), u(i, { key: 3 }, [
        a(s(f), null, {
          default: o(() => [
            a(w, v({ ref_for: !0 }, y, {
              containerClass: l.layout === "reference" && "border-t",
              modelValue: t.name,
              placeholder: "api-key",
              "onUpdate:modelValue": e[3] || (e[3] = (n) => c("update:securityScheme", {
                type: "apiKey",
                payload: { name: n }
              }))
            }), {
              default: o(() => e[8] || (e[8] = [
                x(" Name ")
              ])),
              _: 2,
              __: [8]
            }, 1040, ["containerClass", "modelValue"])
          ]),
          _: 2
        }, 1024),
        a(s(f), null, {
          default: o(() => [
            a(w, v({ ref_for: !0 }, y, {
              modelValue: t["x-scalar-secret-token"],
              placeholder: "QUxMIFlPVVIgQkFTRSBBUkUgQkVMT05HIFRPIFVT",
              type: "password",
              "onUpdate:modelValue": e[4] || (e[4] = (n) => c("update:securityScheme", {
                type: "apiKey",
                payload: { value: n }
              }))
            }), {
              default: o(() => e[9] || (e[9] = [
                x(" Value ")
              ])),
              _: 2,
              __: [9]
            }, 1040, ["modelValue"])
          ]),
          _: 2
        }, 1024)
      ], 64)) : t?.type === "oauth2" ? (r(), u(i, { key: 4 }, [
        a(s(f), null, {
          default: o(() => [
            Object.keys(t.flows).length > 1 ? (r(), u("div", M, [
              C("div", Q, [
                (r(!0), u(i, null, U(t?.flows, (n, d, V) => (r(), u("button", {
                  key: d,
                  class: j(["floating-bg text-c-3 relative cursor-pointer border-b-[1px] border-transparent py-1 text-base font-medium", {
                    "!text-c-1 !rounded-none border-b-[1px] !border-current": l.layout !== "reference" && (p.value === d || V === 0 && !p.value),
                    "!text-c-1 !rounded-none border-b-[1px] !border-current opacity-100": l.layout === "reference" && (p.value === d || V === 0 && !p.value)
                  }]),
                  type: "button",
                  onClick: (g) => p.value = d
                }, [
                  C("span", D, F(d), 1)
                ], 10, q))), 128))
              ])
            ])) : b("", !0)
          ]),
          _: 2
        }, 1024),
        (r(!0), u(i, null, U(t.flows, (n, d, V) => (r(), u(i, { key: d }, [
          p.value === d || V === 0 && !p.value ? (r(), k(O, v({
            key: 0,
            ref_for: !0
          }, y, {
            flows: t.flows,
            proxyUrl: "",
            scheme: t,
            selectedScopes: h,
            server: l.server,
            type: d,
            "onUpdate:securityScheme": (g) => c("update:securityScheme", {
              type: "oauth2",
              flow: d,
              payload: g
            }),
            "onUpdate:selectedScopes": (g) => c("update:selectedScopes", {
              id: Object.keys(l.selectedSecuritySchema),
              name: m,
              ...g
            })
          }), null, 16, ["flows", "scheme", "selectedScopes", "server", "type", "onUpdate:securityScheme", "onUpdate:selectedScopes"])) : b("", !0)
        ], 64))), 128))
      ], 64)) : t?.type === "openIdConnect" ? (r(), u("div", {
        key: 5,
        class: j(["text-c-3 bg-b-1 flex min-h-[calc(4rem+1px)] items-center justify-center border-t border-b-0 px-4 text-base", { "rounded-b-lg": l.layout === "reference" }])
      }, " Coming soon ", 2)) : b("", !0)
    ], 64))), 128));
  }
});
export {
  Y as default
};
