// @scalar/components/dist/components/ScalarTextArea/ScalarTextArea.vue.js
import { defineComponent as u, useModel as m, onMounted as f, createBlock as d, openBlock as c, unref as t, mergeProps as l, withCtx as p, withDirectives as x, createElementVNode as _, vModelText as v } from "vue";
import { useBindCx as h } from "@scalar/use-hooks/useBindCx";
import { useTextareaAutosize as k } from "@vueuse/core";
import C from "../ScalarForm/ScalarFormInput.vue.js";
const V = {}, T = /* @__PURE__ */ u({
  ...V,
  inheritAttrs: !1,
  __name: "ScalarTextArea",
  props: {
    modelValue: {},
    modelModifiers: {}
  },
  emits: ["update:modelValue"],
  setup(a) {
    const o = m(a, "modelValue"), { textarea: r } = k({
      // @ts-expect-error - unexpected type mismatch
      input: o,
      styleProp: "minHeight"
    }), { classCx: i, otherAttrs: n } = h();
    return f(() => {
      "autofocus" in n.value && r.value?.focus();
    }), (w, e) => (c(), d(t(C), l({ is: "div" }, t(i)("custom-scroll flex text-c-1 min-h-0 shrink cursor-text"), {
      onClick: e[1] || (e[1] = (s) => t(r)?.focus())
    }), {
      default: p(() => [
        x(_("textarea", l({
          ref_key: "textarea",
          ref: r,
          "onUpdate:modelValue": e[0] || (e[0] = (s) => o.value = s)
        }, t(n), { class: "w-full resize-none border-none bg-transparent text-sm focus-within:outline-none" }), null, 16), [
          [v, o.value]
        ])
      ]),
      _: 1
    }, 16));
  }
});
export {
  T as default
};
