// @scalar/api-reference/dist/v2/blocks/scalar-info-block/components/InfoBlock.vue.js
import { defineComponent as s, computed as r, createBlock as a, openBlock as d, withCtx as n, createVNode as e, renderSlot as l } from "vue";
import u from "./DownloadLink.vue.js";
import m from "./IntroductionCard.vue.js";
import p from "./IntroductionLayout.vue.js";
const E = /* @__PURE__ */ s({
  __name: "InfoBlock",
  props: {
    id: {},
    oasVersion: {},
    info: {},
    externalDocs: {},
    documentExtensions: {},
    infoExtensions: {},
    options: {}
  },
  setup(i) {
    const t = r(
      () => i.options.layout === "classic" ? "after" : "aside"
    );
    return (o, c) => (d(), a(p, {
      id: o.id,
      documentExtensions: o.documentExtensions,
      externalDocs: o.externalDocs,
      info: o.info,
      infoExtensions: o.infoExtensions,
      isLoading: o.options.isLoading,
      oasVersion: o.oasVersion,
      onLoaded: o.options.onLoaded
    }, {
      [t.value]: n(() => [
        e(m, {
          row: o.options.layout === "classic"
        }, {
          default: n(() => [
            l(o.$slots, "selectors")
          ]),
          _: 3
        }, 8, ["row"])
      ]),
      "download-link": n(() => [
        e(u, {
          documentDownloadType: o.options.documentDownloadType ?? "both",
          getOriginalDocument: o.options.getOriginalDocument,
          title: o.info?.title,
          url: o.options.url
        }, null, 8, ["documentDownloadType", "getOriginalDocument", "title", "url"])
      ]),
      _: 2
    }, 1032, ["id", "documentExtensions", "externalDocs", "info", "infoExtensions", "isLoading", "oasVersion", "onLoaded"]));
  }
});
export {
  E as default
};
