// @scalar/components/dist/components/ScalarMarkdown/ScalarMarkdown.vue.js
import { defineComponent as s, useTemplateRef as f, computed as u, createElementBlock as h, openBlock as d, mergeProps as g, unref as w } from "vue";
import { htmlFromMarkdown as T } from "@scalar/code-highlight";
import { useBindCx as p } from "@scalar/use-hooks/useBindCx";
const v = ["innerHTML"], B = /* @__PURE__ */ s({
  inheritAttrs: !1,
  __name: "ScalarMarkdown",
  props: {
    value: {},
    withImages: { type: Boolean, default: !1 },
    withAnchors: { type: Boolean, default: !1 },
    transform: { type: Function },
    transformType: {},
    clamp: {},
    anchorPrefix: {}
  },
  setup(e, { expose: n }) {
    const { cx: a } = p(), o = f("div");
    n({ el: o });
    const m = (r) => {
      if (!e.withAnchors)
        return e.transform?.(r) || r;
      const t = (r.children?.[0]?.value || "").toLowerCase().replace(/\s+/g, "-"), l = e.anchorPrefix ? `${e.anchorPrefix}/description/${t}` : t;
      return r.data = {
        hProperties: {
          id: l
        }
      }, e.transform ? e.transform(r) : r;
    }, i = u(() => T(e.value ?? "", {
      removeTags: e.withImages ? [] : ["img", "picture"],
      transform: e.withAnchors && e.transformType === "heading" ? m : e.transform,
      transformType: e.transformType
    }));
    return (r, c) => (d(), h("div", g({ ref: "div" }, w(a)("markdown", { "line-clamp-(--markdown-clamp)": !!r.clamp }), {
      style: { "--markdown-clamp": r.clamp },
      innerHTML: i.value
    }), null, 16, v));
  }
});
export {
  B as default
};
