// @scalar/api-reference/dist/features/document-source/hooks/useDocumentSource.js
import { createWorkspaceStore as A, createActiveEntitiesStore as v } from "@scalar/api-client/store";
import { measureAsync as P } from "@scalar/helpers/testing/measure";
import { normalize as b, dereference as O } from "@scalar/openapi-parser";
import { upgrade as V } from "@scalar/openapi-upgrader";
import { apiReferenceConfigurationSchema as x } from "@scalar/types/api-reference";
import { computed as f, toValue as t, ref as d, watch as C } from "vue";
import { useDocumentFetcher as D } from "./useDocumentFetcher.js";
function j({
  originalDocument: c,
  dereferencedDocument: o,
  configuration: s
}) {
  const { originalDocument: h } = D({
    configuration: s
  }), u = f(() => c ? t(c) : t(h)), n = d(""), g = f(() => o ? t(o) : t(m)), m = d({
    openapi: "3.1.0",
    info: {
      title: "",
      version: ""
    },
    paths: {}
  });
  C(
    () => t(u),
    async (a) => {
      if (!a)
        return;
      const e = b(a);
      if (e === void 0) {
        console.error("Failed to parse the OpenAPI document:", a);
        return;
      }
      if (n.value = typeof e == "object" && (e.openapi || e.swagger) || "", o)
        return;
      const y = !n.value.startsWith("3.1") ? (
        // Upgrade needed
        V(e, "3.1")
      ) : (
        // Skip the upgrade
        e
      ), w = await P("dereference", async () => {
        const { schema: p, errors: i } = await O(y);
        return i?.length && console.warn(
          `Please open an issue on https://github.com/scalar/scalar
`,
          `Scalar OpenAPI Parser Warning:
`,
          i
        ), p === void 0 && console.error("Failed to dereference the OpenAPI document", i), p;
      });
      m.value = w;
    },
    { immediate: !0 }
  );
  const r = t(s) ?? x.parse({}), l = A({
    useLocalStorage: !1,
    proxyUrl: r.proxyUrl,
    theme: r.theme,
    showSidebar: r.showSidebar,
    hideClientButton: r.hideClientButton,
    _integration: r._integration
  }), S = v(l);
  return {
    originalDocument: u,
    originalOpenApiVersion: n,
    dereferencedDocument: g,
    workspaceStore: l,
    activeEntitiesStore: S
  };
}
export {
  j as useDocumentSource
};
