// @scalar/api-client/dist/v2/blocks/scalar-operation-block/components/OperationBlock.vue2.js
import { defineComponent as g, ref as D, computed as f, useId as q, watch as O, createBlock as U, openBlock as y, withCtx as k, withDirectives as p, createVNode as s, unref as o, vShow as m, createElementVNode as T, createElementBlock as b, createCommentVNode as N, toDisplayString as H } from "vue";
import { canMethodHaveBody as V, REGEX as K } from "@scalar/oas-utils/helpers";
import { getResolvedRef as R } from "@scalar/workspace-store/helpers/get-resolved-ref";
import I from "../../../../components/SectionFilter.vue.js";
import E from "../../../../components/ViewLayout/ViewLayoutSection.vue.js";
import P from "./OperationBody.vue.js";
import v from "./OperationParams.vue.js";
import { groupBy as Q } from "../helpers/group-by.js";
import $ from "../../scalar-auth-selector-block/components/AuthSelector.vue.js";
const j = { class: "group pointer-events-none flex flex-1 items-center gap-1 lg:pr-24" }, z = ["for"], F = ["id", "placeholder", "value"], L = {
  key: 2,
  class: "text-c-1 flex h-8 items-center"
}, _ = /* @__PURE__ */ g({
  __name: "OperationBlock",
  props: {
    method: {},
    path: {},
    operation: {},
    exampleKey: {},
    selectedContentType: {},
    securitySchemes: {},
    selectedSecurity: {},
    security: {},
    server: {},
    layout: {},
    environment: {},
    envVariables: {}
  },
  emits: ["operation:update:requestName", "auth:delete", "auth:update:securityScheme", "auth:update:selectedScopes", "auth:update:selectedSecurity", "parameters:add", "parameters:update", "parameters:delete", "parameters:deleteAll", "requestBody:update:contentType", "requestBody:update:value", "requestBody:add:formRow", "requestBody:update:formRow"],
  setup(l, { emit: h }) {
    const a = h, i = Q(
      l.operation.parameters?.map((r) => R(r)) ?? [],
      "in"
    ), w = [
      "Auth",
      "Variables",
      "Cookies",
      "Headers",
      "Query",
      "Body"
    ], d = D("All"), A = f(() => {
      const r = /* @__PURE__ */ new Set(["All", ...w]);
      return i.path?.length || r.delete("Variables"), V(l.method) || r.delete("Body"), S.value && r.delete("Auth"), [...r];
    }), n = f(
      () => Object.fromEntries(
        A.value.map((r) => [r, q()])
      )
    );
    O(
      () => l.method,
      (r) => {
        d.value === "Body" && !V(r) && (d.value = "All");
      }
    );
    const S = f(
      () => l.layout === "modal" && !l.operation.security && !Object.keys(l.securitySchemes ?? {}).length
    ), C = () => l.operation.summary ? l.operation.summary : l.path.replace(K.PROTOCOL, "") ? l.path.replace(K.PROTOCOL, "") : "Request Name", u = (r) => d.value === "All" || d.value === r, B = q();
    return (r, e) => (y(), U(E, {
      "aria-label": `Request: ${r.operation.summary}`
    }, {
      title: k(() => [
        T("div", j, [
          r.layout !== "modal" ? (y(), b("label", {
            key: 0,
            class: "pointer-events-auto absolute top-0 left-0 h-full w-full cursor-text opacity-0",
            for: o(B)
          }, null, 8, z)) : N("", !0),
          r.layout !== "modal" ? (y(), b("input", {
            key: 1,
            id: o(B),
            class: "text-c-1 group-hover-input pointer-events-auto relative z-10 -ml-0.5 h-8 w-full rounded pl-1.25 has-[:focus-visible]:outline md:-ml-1.25",
            placeholder: C(),
            value: r.operation.summary,
            onInput: e[0] || (e[0] = (t) => a("operation:update:requestName", {
              name: t.target.value
            }))
          }, null, 40, F)) : (y(), b("span", L, H(r.operation.summary), 1))
        ]),
        s(I, {
          modelValue: d.value,
          "onUpdate:modelValue": e[1] || (e[1] = (t) => d.value = t),
          filterIds: n.value,
          filters: A.value
        }, null, 8, ["modelValue", "filterIds", "filters"])
      ]),
      default: k(() => [
        p(s(o($), {
          id: n.value.Auth,
          envVariables: r.envVariables,
          environment: r.environment,
          layout: "client",
          security: r.security,
          securitySchemes: r.securitySchemes,
          selectedSecurity: r.selectedSecurity,
          server: r.server,
          title: "Authorization",
          onDeleteOperationAuth: e[2] || (e[2] = (t) => a("auth:delete", t)),
          "onUpdate:securityScheme": e[3] || (e[3] = (t) => a("auth:update:securityScheme", t)),
          "onUpdate:selectedScopes": e[4] || (e[4] = (t) => a("auth:update:selectedScopes", t)),
          "onUpdate:selectedSecurity": e[5] || (e[5] = (t) => a("auth:update:selectedSecurity", t))
        }, null, 8, ["id", "envVariables", "environment", "security", "securitySchemes", "selectedSecurity", "server"]), [
          [m, u("Auth") && !S.value]
        ]),
        p(s(v, {
          id: n.value.Variables,
          envVariables: r.envVariables,
          environment: r.environment,
          exampleKey: r.exampleKey,
          parameters: o(i).path ?? [],
          title: "Variables",
          onAdd: e[6] || (e[6] = (t) => a("parameters:add", {
            type: "path",
            payload: t
          })),
          onDelete: e[7] || (e[7] = (t) => a("parameters:delete", {
            type: "path",
            ...t
          })),
          onDeleteAll: e[8] || (e[8] = () => a("parameters:deleteAll", {
            type: "path"
          })),
          onUpdate: e[9] || (e[9] = (t) => a("parameters:update", {
            type: "path",
            ...t
          }))
        }, null, 8, ["id", "envVariables", "environment", "exampleKey", "parameters"]), [
          [m, u("Variables") && o(i).path?.length]
        ]),
        p(s(v, {
          id: n.value.Cookies,
          envVariables: r.envVariables,
          environment: r.environment,
          exampleKey: r.exampleKey,
          parameters: o(i).cookie ?? [],
          title: "Cookies",
          onAdd: e[10] || (e[10] = (t) => a("parameters:add", {
            type: "cookie",
            payload: t
          })),
          onDelete: e[11] || (e[11] = (t) => a("parameters:delete", {
            type: "cookie",
            ...t
          })),
          onDeleteAll: e[12] || (e[12] = () => a("parameters:deleteAll", {
            type: "cookie"
          })),
          onUpdate: e[13] || (e[13] = (t) => a("parameters:update", {
            type: "cookie",
            ...t
          }))
        }, null, 8, ["id", "envVariables", "environment", "exampleKey", "parameters"]), [
          [m, u("Cookies")]
        ]),
        p(s(v, {
          id: n.value.Headers,
          envVariables: r.envVariables,
          environment: r.environment,
          exampleKey: r.exampleKey,
          parameters: o(i).header ?? [],
          title: "Headers",
          onAdd: e[14] || (e[14] = (t) => a("parameters:add", {
            type: "header",
            payload: t
          })),
          onDelete: e[15] || (e[15] = (t) => a("parameters:delete", {
            type: "header",
            ...t
          })),
          onDeleteAll: e[16] || (e[16] = () => a("parameters:deleteAll", {
            type: "header"
          })),
          onUpdate: e[17] || (e[17] = (t) => a("parameters:update", {
            type: "header",
            ...t
          }))
        }, null, 8, ["id", "envVariables", "environment", "exampleKey", "parameters"]), [
          [m, u("Headers")]
        ]),
        p(s(v, {
          id: n.value.Query,
          envVariables: r.envVariables,
          environment: r.environment,
          exampleKey: r.exampleKey,
          parameters: o(i).query ?? [],
          title: "Query Parameters",
          onAdd: e[18] || (e[18] = (t) => a("parameters:add", {
            type: "query",
            payload: t
          })),
          onDelete: e[19] || (e[19] = (t) => a("parameters:delete", {
            type: "query",
            ...t
          })),
          onDeleteAll: e[20] || (e[20] = () => a("parameters:deleteAll", {
            type: "query"
          })),
          onUpdate: e[21] || (e[21] = (t) => a("parameters:update", {
            type: "query",
            ...t
          }))
        }, null, 8, ["id", "envVariables", "environment", "exampleKey", "parameters"]), [
          [m, u("Query")]
        ]),
        p(s(P, {
          id: n.value.Body,
          envVariables: r.envVariables,
          environment: r.environment,
          exampleKey: r.exampleKey,
          requestBody: o(R)(r.operation.requestBody),
          selectedContentType: r.selectedContentType ?? "other",
          title: "Request Body",
          "onAdd:formRow": e[22] || (e[22] = (t) => a("requestBody:add:formRow", t)),
          "onUpdate:contentType": e[23] || (e[23] = (t) => a("requestBody:update:contentType", t)),
          "onUpdate:formRow": e[24] || (e[24] = (t) => a("requestBody:update:formRow", t)),
          "onUpdate:value": e[25] || (e[25] = (t) => a("requestBody:update:value", t))
        }, null, 8, ["id", "envVariables", "environment", "exampleKey", "requestBody", "selectedContentType"]), [
          [m, u("Body") && o(V)(r.method)]
        ])
      ]),
      _: 1
    }, 8, ["aria-label"]));
  }
});
export {
  _ as default
};
