// @scalar/api-client/dist/v2/blocks/scalar-operation-block/components/OperationBody.vue2.js
import { defineComponent as j, computed as g, createBlock as u, openBlock as i, withCtx as a, createVNode as r, unref as l, createElementVNode as m, toDisplayString as f, createElementBlock as c, Fragment as F, createTextVNode as T } from "vue";
import { ScalarListbox as h, ScalarButton as y, ScalarIcon as x } from "@scalar/components";
import N from "../../../../components/CodeInput/CodeInput.vue.js";
import O from "../../../../components/DataTable/DataTable.vue.js";
import U from "../../../../components/DataTable/DataTableHeader.vue.js";
import D from "../../../../components/DataTable/DataTableRow.vue.js";
import R from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import V from "./OperationTable.vue.js";
import { getFileName as C } from "../helpers/files.js";
import { getExampleFromBody as S } from "../helpers/get-request-body-example.js";
import { useFileDialog as E } from "../../../../hooks/useFileDialog.js";
const L = {
  key: 0,
  class: "text-c-3 flex min-h-10 w-full items-center justify-center border-t p-2 text-sm"
}, M = {
  key: 1,
  class: "flex items-center justify-center overflow-hidden border-t p-1.5"
}, z = { class: "text-c-2 w-full max-w-full overflow-hidden rounded border px-1.5 py-1 text-xs whitespace-nowrap" }, Q = /* @__PURE__ */ j({
  __name: "OperationBody",
  props: {
    requestBody: {},
    selectedContentType: {},
    exampleKey: {},
    title: {},
    environment: {},
    envVariables: {}
  },
  emits: ["update:contentType", "update:value", "add:formRow", "update:formRow"],
  setup(d, { emit: k }) {
    const p = k, B = {
      "application/json": "json",
      "application/xml": "xml",
      "application/yaml": "yaml"
    }, v = Object.entries({
      "multipart/form-data": "Multipart Form",
      "application/x-www-form-urlencoded": "Form URL Encoded",
      "application/octet-stream": "Binary File",
      "application/json": "JSON",
      "application/xml": "XML",
      "application/yaml": "YAML",
      "application/edn": "EDN",
      other: "Other",
      none: "None"
    }).map(([t, e]) => ({
      id: t,
      label: e
    })), b = g({
      get: () => v.find((e) => e.id === d.selectedContentType) ?? v.at(-1),
      set: (t) => {
        p("update:contentType", { value: t.id });
      }
    });
    function w(t) {
      const { open: e } = E({
        onChange: async (n) => {
          const o = n?.[0];
          o && t(o);
        },
        multiple: !1,
        accept: "*/*"
      });
      e();
    }
    const s = g(
      () => d.requestBody && S(d.requestBody, d.selectedContentType, d.exampleKey)
    );
    return (t, e) => (i(), u(R, null, {
      title: a(() => [
        T(f(t.title), 1)
      ]),
      default: a(() => [
        r(O, {
          columns: [""],
          presentational: ""
        }, {
          default: a(() => [
            r(U, { class: "relative col-span-full flex h-8 cursor-pointer items-center justify-between !p-0" }, {
              default: a(() => [
                r(l(h), {
                  modelValue: b.value,
                  "onUpdate:modelValue": e[0] || (e[0] = (n) => b.value = n),
                  options: l(v),
                  teleport: ""
                }, {
                  default: a(() => [
                    r(l(y), {
                      class: "text-c-2 hover:text-c-1 flex h-full w-fit gap-1.5 px-3 font-normal",
                      fullWidth: "",
                      variant: "ghost"
                    }, {
                      default: a(() => [
                        m("span", null, f(t.selectedContentType), 1),
                        r(l(x), {
                          icon: "ChevronDown",
                          size: "md"
                        })
                      ]),
                      _: 1
                    })
                  ]),
                  _: 1
                }, 8, ["modelValue", "options"])
              ]),
              _: 1
            }),
            r(D, null, {
              default: a(() => [
                t.selectedContentType === "none" ? (i(), c("div", L, e[5] || (e[5] = [
                  m("span", null, "No Body", -1)
                ]))) : t.selectedContentType === "application/octet-stream" ? (i(), c("div", M, [
                  l(C)(s.value?.value) !== void 0 ? (i(), c(F, { key: 0 }, [
                    m("span", z, f(l(C)(s.value?.value)), 1),
                    r(l(y), {
                      class: "bg-b-2 hover:bg-b-3 text-c-2 ml-1 border-0 shadow-none",
                      size: "sm",
                      variant: "outlined",
                      onClick: e[1] || (e[1] = (n) => p("update:value", { value: void 0 }))
                    }, {
                      default: a(() => e[6] || (e[6] = [
                        T(" Delete ")
                      ])),
                      _: 1,
                      __: [6]
                    })
                  ], 64)) : (i(), u(l(y), {
                    key: 1,
                    class: "bg-b-2 hover:bg-b-3 text-c-2 border-0 shadow-none",
                    size: "sm",
                    variant: "outlined",
                    onClick: e[2] || (e[2] = () => w(
                      (n) => p("update:value", { value: n })
                    ))
                  }, {
                    default: a(() => [
                      e[7] || (e[7] = m("span", null, "Upload File", -1)),
                      r(l(x), {
                        class: "ml-1",
                        icon: "Upload",
                        size: "xs",
                        thickness: "2.5"
                      })
                    ]),
                    _: 1,
                    __: [7]
                  }))
                ])) : t.selectedContentType === "multipart/form-data" ? (i(), u(V, {
                  key: 2,
                  data: Object.entries(
                    typeof s.value?.value == "object" ? s.value.value ?? {} : {}
                  ).map(([n, o]) => ({
                    name: n,
                    value: o
                  })),
                  envVariables: t.envVariables,
                  environment: t.environment,
                  showUploadButton: "",
                  onUploadFile: e[3] || (e[3] = (n) => w((o) => {
                    if (n !== void 0)
                      return p("update:formRow", {
                        index: n,
                        payload: { value: o ?? void 0 }
                      });
                    p("add:formRow", { value: o ?? void 0 });
                  }))
                }, null, 8, ["data", "envVariables", "environment"])) : t.selectedContentType === "application/x-www-form-urlencoded" ? (i(), u(V, {
                  key: 3,
                  data: Object.entries(
                    typeof s.value?.value == "object" ? s.value.value : {}
                  ).map(([n, o]) => ({
                    name: n,
                    value: String(o)
                  })),
                  envVariables: t.envVariables,
                  environment: t.environment
                }, null, 8, ["data", "envVariables", "environment"])) : (i(), u(N, {
                  key: 4,
                  class: "border-t px-3",
                  content: "",
                  envVariables: t.envVariables,
                  environment: t.environment,
                  language: B[t.selectedContentType] ?? "plaintext",
                  lineNumbers: "",
                  lint: "",
                  modelValue: s.value?.value ?? "",
                  "onUpdate:modelValue": e[4] || (e[4] = (n) => p("update:value", { value: n }))
                }, null, 8, ["envVariables", "environment", "language", "modelValue"]))
              ]),
              _: 1
            })
          ]),
          _: 1
        })
      ]),
      _: 1
    }));
  }
});
export {
  Q as default
};
