// @scalar/api-client/dist/v2/blocks/scalar-operation-block/components/OperationTable.vue2.js
import { defineComponent as h, computed as y, createBlock as p, openBlock as r, withCtx as o, createVNode as t, createElementBlock as B, createTextVNode as s, toDisplayString as u, Fragment as D, renderList as F } from "vue";
import k from "../../../../components/DataTable/DataTable.vue.js";
import m from "../../../../components/DataTable/DataTableHeader.vue.js";
import U from "../../../../components/DataTable/DataTableRow.vue.js";
import V from "./OperationTableRow.vue.js";
const T = /* @__PURE__ */ h({
  __name: "OperationTable",
  props: {
    data: {},
    isReadOnly: { type: Boolean },
    hasCheckboxDisabled: { type: Boolean },
    invalidParams: {},
    label: {},
    showUploadButton: { type: Boolean },
    environment: {},
    envVariables: {}
  },
  emits: ["addRow", "updateRow", "deleteRow", "uploadFile", "removeFile"],
  setup(n, { emit: b }) {
    const a = b, v = (e) => {
      a("uploadFile", e);
    }, c = ["36px", "", "", "auto"], f = y(() => {
      const e = n.data.at(-1);
      return !e || e.name !== "" && e.value !== "" ? [...n.data, { name: "", value: "", isDisabled: !0 }] : n.data;
    }), R = ({
      index: e,
      payload: i
    }) => {
      if (e >= n.data.length) {
        a("addRow", i);
        return;
      }
      a("updateRow", e, i);
    };
    return (e, i) => (r(), p(k, {
      class: "group/table flex-1",
      columns: c
    }, {
      default: o(() => [
        t(U, { class: "sr-only !block" }, {
          default: o(() => [
            t(m, null, {
              default: o(() => [
                s(u(e.label) + " Enabled", 1)
              ]),
              _: 1
            }),
            t(m, null, {
              default: o(() => [
                s(u(e.label) + " Key", 1)
              ]),
              _: 1
            }),
            t(m, null, {
              default: o(() => [
                s(u(e.label) + " Value", 1)
              ]),
              _: 1
            })
          ]),
          _: 1
        }),
        (r(!0), B(D, null, F(f.value, (w, l) => (r(), p(V, {
          key: l,
          data: w,
          envVariables: e.envVariables,
          environment: e.environment,
          hasCheckboxDisabled: e.hasCheckboxDisabled,
          invalidParams: e.invalidParams,
          isReadOnly: e.isReadOnly,
          label: e.label,
          showUploadButton: e.showUploadButton,
          onDeleteRow: (d) => a("deleteRow", l),
          onRemoveFile: (d) => a("removeFile", l),
          onUpdateRow: (d) => R({ index: l, payload: d }),
          onUploadFile: () => v(l)
        }, null, 8, ["data", "envVariables", "environment", "hasCheckboxDisabled", "invalidParams", "isReadOnly", "label", "showUploadButton", "onDeleteRow", "onRemoveFile", "onUpdateRow", "onUploadFile"]))), 128))
      ]),
      _: 1
    }));
  }
});
export {
  T as default
};
