// @scalar/api-client/dist/v2/blocks/scalar-operation-block/components/OperationParams.vue.js
import { defineComponent as y, computed as r, createBlock as i, openBlock as s, withCtx as l, createVNode as m, createElementVNode as p, createCommentVNode as R, unref as d, withModifiers as w, createTextVNode as u, toDisplayString as v } from "vue";
import { ScalarTooltip as C, ScalarButton as V } from "@scalar/components";
import x from "../../../../components/ViewLayout/ViewLayoutCollapse.vue.js";
import P from "./OperationTable.vue.js";
import { getParameterExample as K } from "../helpers/get-parameter-example.js";
import { getParameterSchema as h } from "../helpers/get-parameter-schema.js";
const k = { class: "text-c-2 request-meta-buttons flex whitespace-nowrap opacity-0 group-hover/params:opacity-100 has-[:focus-visible]:opacity-100" }, A = { class: "sr-only" }, E = /* @__PURE__ */ y({
  __name: "OperationParams",
  props: {
    parameters: {},
    exampleKey: {},
    title: {},
    label: {},
    invalidParams: {},
    globalRoute: {},
    environment: {},
    envVariables: {}
  },
  emits: ["add", "update", "delete", "deleteAll"],
  setup(o, { emit: b }) {
    const n = b, f = r(
      () => o.parameters.map((e) => {
        const t = K(e, o.exampleKey);
        return {
          name: e.name,
          value: t?.value ?? "",
          globalRoute: o.globalRoute,
          schema: h(e),
          isRequired: e.required,
          isDisabled: t?.["x-disabled"] ?? !1
        };
      })
    ), g = r(() => o.parameters.length > 1);
    return (e, t) => (s(), i(x, {
      class: "group/params",
      itemCount: e.parameters.length
    }, {
      title: l(() => [
        u(v(e.title), 1)
      ]),
      actions: l(() => [
        p("div", k, [
          g.value ? (s(), i(d(C), {
            key: 0,
            content: "Clear optional parameters",
            placement: "left"
          }, {
            default: l(() => [
              m(d(V), {
                class: "pr-0.75 pl-1 transition-none",
                size: "sm",
                variant: "ghost",
                onClick: t[0] || (t[0] = w((a) => n("deleteAll"), ["stop"]))
              }, {
                default: l(() => [
                  t[4] || (t[4] = u(" Clear ")),
                  p("span", A, "All " + v(e.title), 1)
                ]),
                _: 1,
                __: [4]
              })
            ]),
            _: 1
          })) : R("", !0)
        ])
      ]),
      default: l(() => [
        m(P, {
          class: "flex-1",
          columns: ["32px", "", ""],
          data: f.value,
          envVariables: e.envVariables,
          environment: e.environment,
          exampleKey: e.exampleKey,
          globalRoute: e.globalRoute,
          invalidParams: e.invalidParams,
          label: e.label,
          onAddRow: t[1] || (t[1] = (a) => n("add", a)),
          onDeleteRow: t[2] || (t[2] = (a) => n("delete", { index: a })),
          onUpdateRow: t[3] || (t[3] = (a, c) => n("update", { index: a, payload: c }))
        }, null, 8, ["data", "envVariables", "environment", "exampleKey", "globalRoute", "invalidParams", "label"])
      ]),
      _: 1
    }, 8, ["itemCount"]));
  }
});
export {
  E as default
};
