// @scalar/api-client/dist/v2/blocks/scalar-operation-block/components/OperationTableRow.vue.js
import { defineComponent as z, computed as r, resolveComponent as S, createBlock as u, openBlock as s, unref as t, normalizeClass as U, withCtx as i, createVNode as d, createCommentVNode as b, createElementVNode as m, createElementBlock as p, Fragment as $, toDisplayString as x } from "vue";
import { ScalarTooltip as D, ScalarIcon as c, ScalarButton as y } from "@scalar/components";
import { ScalarIconTrash as I } from "@scalar/icons";
import E from "./OperationTableTooltip.vue.js";
import { validateParameter as O } from "../helpers/validate-parameter.js";
import q from "../../../../components/DataTable/DataTableRow.vue.js";
import j from "../../../../components/DataTable/DataTableCheckbox.vue.js";
import v from "../../../../components/DataTable/DataTableCell.vue.js";
import h from "../../../../components/CodeInput/CodeInput.vue.js";
const G = { class: "text-c-2 filemask flex w-full max-w-[100%] items-center justify-center overflow-hidden p-1" }, N = {
  key: 1,
  class: "p-0.5"
}, X = /* @__PURE__ */ z({
  __name: "OperationTableRow",
  props: {
    data: {},
    isReadOnly: { type: Boolean },
    hasCheckboxDisabled: { type: Boolean },
    invalidParams: {},
    label: {},
    environment: {},
    envVariables: {},
    showUploadButton: { type: Boolean }
  },
  emits: ["updateRow", "deleteRow", "uploadFile", "removeFile"],
  setup(l, { emit: k }) {
    const n = k, V = r(() => l.data.schema?.default), w = r(() => l.data.schema?.enum), R = r(
      () => l.data.schema && "minimum" in l.data.schema ? l.data.schema.minimum : void 0
    ), g = r(
      () => l.data.schema && "maximum" in l.data.schema ? l.data.schema.maximum : void 0
    ), B = r(
      () => l.data.schema && "type" in l.data.schema ? l.data.schema.type : void 0
    ), f = r(
      () => O(l.data.schema, l.data.value)
    ), C = (e) => e instanceof File, F = r({
      get: () => l.data.value instanceof File ? l.data.value.name : l.data.value === null ? "" : l.data.value,
      set: (e) => {
        typeof e == "string" && n("updateRow", { value: e });
      }
    });
    return (e, a) => {
      const T = S("RouterLink");
      return s(), u(t(q), {
        id: e.data.name,
        class: U({
          alert: f.value.ok === !1,
          error: f.value.ok === !1 && e.invalidParams?.has(e.data.name)
        })
      }, {
        default: i(() => [
          e.data.globalRoute !== void 0 ? (s(), u(T, {
            key: 0,
            class: "text-c-2 flex items-center justify-center border-t !border-r",
            to: e.data.globalRoute ?? {}
          }, {
            default: i(() => [
              a[8] || (a[8] = m("span", { class: "sr-only" }, "Global", -1)),
              d(t(D), {
                content: "Global cookies are shared across the whole workspace.",
                placement: "top"
              }, {
                default: i(() => [
                  d(t(c), {
                    class: "text-c-1",
                    icon: "Globe",
                    size: "xs",
                    tabindex: "0"
                  })
                ]),
                _: 1
              })
            ]),
            _: 1,
            __: [8]
          }, 8, ["to"])) : (s(), u(t(j), {
            key: 1,
            class: "!border-r",
            disabled: e.hasCheckboxDisabled ?? !1,
            modelValue: !e.data.isDisabled,
            "onUpdate:modelValue": a[0] || (a[0] = (o) => n("updateRow", { isEnabled: o }))
          }, null, 8, ["disabled", "modelValue"])),
          d(t(v), null, {
            default: i(() => [
              d(t(h), {
                "aria-label": `${e.label} Key`,
                disableCloseBrackets: "",
                disabled: e.isReadOnly,
                disableEnter: "",
                disableTabIndent: "",
                envVariables: e.envVariables,
                environment: e.environment,
                lineWrapping: "",
                modelValue: e.data.name,
                placeholder: "Key",
                required: !!e.data.isRequired,
                onSelectVariable: a[1] || (a[1] = (o) => n("updateRow", { key: o })),
                "onUpdate:modelValue": a[2] || (a[2] = (o) => n("updateRow", { key: o }))
              }, null, 8, ["aria-label", "disabled", "envVariables", "environment", "modelValue", "required"])
            ]),
            _: 1
          }),
          d(t(v), null, {
            default: i(() => [
              d(t(h), {
                "aria-label": `${e.label} Value`,
                class: "pr-6 group-hover:pr-10 group-has-[.cm-focused]:pr-10",
                default: V.value,
                disableCloseBrackets: "",
                disabled: e.isReadOnly,
                disableEnter: "",
                disableTabIndent: "",
                enum: w.value ?? [],
                envVariables: e.envVariables,
                environment: e.environment,
                lineWrapping: "",
                max: g.value,
                min: R.value,
                modelValue: F.value,
                placeholder: "Value",
                type: B.value,
                onSelectVariable: a[4] || (a[4] = (o) => n("updateRow", { value: o })),
                "onUpdate:modelValue": a[5] || (a[5] = (o) => n("updateRow", { value: o }))
              }, {
                icon: i(() => [
                  (e.data.name || e.data.value) && !e.data.isRequired ? (s(), u(t(y), {
                    key: 0,
                    class: "text-c-2 hover:text-c-1 hover:bg-b-2 z-context -mr-0.5 hidden h-fit rounded p-1 group-hover:flex group-has-[.cm-focused]:flex",
                    size: "sm",
                    variant: "ghost",
                    onClick: a[3] || (a[3] = (o) => n("deleteRow"))
                  }, {
                    default: i(() => [
                      d(t(I), { class: "size-3.5" })
                    ]),
                    _: 1
                  })) : b("", !0),
                  e.data.schema ? (s(), u(E, {
                    key: 1,
                    schema: e.data.schema,
                    value: e.data.value
                  }, null, 8, ["schema", "value"])) : b("", !0)
                ]),
                _: 1
              }, 8, ["aria-label", "default", "disabled", "enum", "envVariables", "environment", "max", "min", "modelValue", "type"])
            ]),
            _: 1
          }),
          e.showUploadButton ? (s(), u(t(v), {
            key: 2,
            class: "group/upload flex items-center justify-center whitespace-nowrap"
          }, {
            default: i(() => [
              C(e.data.value) ? (s(), p($, { key: 0 }, [
                m("div", G, [
                  m("span", null, x(e.data.value.name), 1)
                ]),
                m("button", {
                  class: "bg-b-2 centered-x centered-y absolute hidden w-[calc(100%_-_8px)] rounded p-0.5 text-center text-xs font-medium group-hover/upload:block",
                  type: "button",
                  onClick: a[6] || (a[6] = (o) => n("removeFile"))
                }, " Delete ")
              ], 64)) : (s(), p("div", N, [
                d(t(y), {
                  class: "bg-b-2 hover:bg-b-3 text-c-2 h-fit border-0 py-px shadow-none",
                  size: "sm",
                  variant: "outlined",
                  onClick: a[7] || (a[7] = (o) => n("uploadFile"))
                }, {
                  default: i(() => [
                    a[9] || (a[9] = m("span", null, "Upload File", -1)),
                    d(t(c), {
                      class: "ml-1",
                      icon: "Upload",
                      size: "xs",
                      thickness: "2.5"
                    })
                  ]),
                  _: 1,
                  __: [9]
                })
              ]))
            ]),
            _: 1
          })) : b("", !0)
        ]),
        _: 1
      }, 8, ["id", "class"]);
    };
  }
});
export {
  X as default
};
