// @scalar/workspace-store/dist/schemas/v3.1/strict/oauth-flow.js
import { Type } from "@scalar/typebox";
import { compose } from "../../../schemas/compose.js";
import {
  XScalarCredentialsLocationSchema
} from "../../../schemas/extensions/security/x-scalar-credentials-location.js";
import {
  XScalarSecurityBodySchema
} from "../../../schemas/extensions/security/x-scalar-security-body.js";
import {
  XScalarSecurityQuerySchema
} from "../../../schemas/extensions/security/x-scalar-security-query.js";
import {
  XScalarSecretClientIdSchema,
  XScalarSecretClientSecretSchema,
  XScalarSecretHTTPSchema,
  XScalarSecretRedirectUriSchema,
  XScalarSecretTokenSchema
} from "../../../schemas/extensions/security/x-scalar-security-secrets.js";
import { XTokenNameSchema } from "../../../schemas/extensions/security/x-tokenName.js";
import { XusePkceSchema } from "../../../schemas/extensions/security/x-use-pkce.js";
const OAuthFlowCommonSchema = compose(
  Type.Object({
    /** The URL to be used for obtaining refresh tokens. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    refreshUrl: Type.String(),
    /** REQUIRED. The available scopes for the OAuth2 security scheme. A map between the scope name and a short description for it. The map MAY be empty. */
    scopes: Type.Record(Type.String(), Type.String())
  }),
  XScalarSecretClientIdSchema,
  XScalarSecretTokenSchema,
  XScalarSecurityQuerySchema,
  XScalarSecurityBodySchema,
  XTokenNameSchema,
  XScalarCredentialsLocationSchema
);
const OAuthFlowImplicitSchema = compose(
  OAuthFlowCommonSchema,
  Type.Object({
    /** REQUIRED. The authorization URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    authorizationUrl: Type.String()
  }),
  XScalarSecretRedirectUriSchema
);
const OAuthFlowPasswordSchema = compose(
  OAuthFlowCommonSchema,
  Type.Object({
    /** REQUIRED. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    tokenUrl: Type.String()
  }),
  XScalarSecretHTTPSchema,
  XScalarSecretClientSecretSchema
);
const OAuthFlowClientCredentialsSchema = compose(
  OAuthFlowCommonSchema,
  Type.Object({
    /** REQUIRED. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    tokenUrl: Type.String()
  }),
  XScalarSecretClientSecretSchema
);
const OAuthFlowAuthorizationCodeSchema = compose(
  OAuthFlowCommonSchema,
  Type.Object({
    /** REQUIRED. The authorization URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    authorizationUrl: Type.String(),
    /** REQUIRED. The token URL to be used for this flow. This MUST be in the form of a URL. The OAuth2 standard requires the use of TLS. */
    tokenUrl: Type.String()
  }),
  XScalarSecretClientSecretSchema,
  XScalarSecretRedirectUriSchema,
  XusePkceSchema
);
const OAuthFlowObjectSchemaDefinition = Type.Union([
  OAuthFlowImplicitSchema,
  OAuthFlowPasswordSchema,
  OAuthFlowClientCredentialsSchema,
  OAuthFlowAuthorizationCodeSchema
]);
export {
  OAuthFlowAuthorizationCodeSchema,
  OAuthFlowClientCredentialsSchema,
  OAuthFlowImplicitSchema,
  OAuthFlowObjectSchemaDefinition,
  OAuthFlowPasswordSchema
};
