// @scalar/api-client/dist/components/ImportCollection/ImportCollectionListener.vue.js
import { defineComponent as I, ref as u, createElementBlock as k, openBlock as y, Fragment as w, createVNode as r, renderSlot as g, nextTick as h } from "vue";
import { useToasts as C } from "@scalar/use-toasts";
import { useRouter as T } from "vue-router";
import { workspaceStoreIsEmpty as $ } from "./utils/workspace-store-is-empty.js";
import { useActiveEntities as q } from "../../store/active-entities.js";
import E from "./DropEventListener.vue.js";
import F from "./ImportCollectionModal.vue.js";
/* empty css                           */
import W from "./PasteEventListener.vue.js";
import x from "./UrlQueryParameterChecker.vue.js";
import { importCollection as B } from "./utils/import-collection.js";
import { useWorkspace as M } from "../../store/store.js";
const J = /* @__PURE__ */ I({
  __name: "ImportCollectionListener",
  setup(N) {
    const c = M(), n = u(null), s = u(null), i = u(null);
    async function m() {
      n.value = null, s.value = null, i.value = null, await h();
    }
    async function a(e, t = null, o) {
      if (await m(), $(c)) {
        if (console.info("Workspace store is empty, directly importing:", e), await d(e))
          return;
        console.warn("Failed to import the collection from:", e);
      }
      n.value = e, s.value = t, i.value = o;
    }
    const f = T(), { activeWorkspace: p } = q(), { toast: l } = C();
    async function d(e) {
      return e ? new Promise((t) => {
        B({
          store: c,
          workspace: p.value,
          source: e,
          // Use watch mode by default.
          watchMode: !0,
          onSuccess(o) {
            if (o) {
              _(o), l("Import successful", "info"), t(!0);
              return;
            }
            l("Import failed: No collection was created", "error"), t(!1);
          },
          onError(o) {
            console.error("[importCollection]", o);
            const v = o?.message || "Unknown error";
            l(`Import failed: ${v}`, "error"), t(!1);
          }
        });
      }) : !1;
    }
    function _(e) {
      e && f.push({
        name: "request",
        params: {
          workspace: p.value?.uid,
          request: e?.requests[0]
        }
      });
    }
    return (e, t) => (y(), k(w, null, [
      r(F, {
        eventType: i.value,
        integration: s.value,
        source: n.value,
        onImportFinished: m
      }, null, 8, ["eventType", "integration", "source"]),
      r(W, { onInput: a }),
      r(E, { onInput: a }),
      r(x, { onInput: a }),
      g(e.$slots, "default")
    ], 64));
  }
});
export {
  J as default
};
