// happy-dom/lib/console/enums/VirtualConsoleLogTypeEnum.js
/**
 * Virtual console log type.
 */
var VirtualConsoleLogTypeEnum;
(function (VirtualConsoleLogTypeEnum) {
    // Log
    VirtualConsoleLogTypeEnum["log"] = "log";
    VirtualConsoleLogTypeEnum["table"] = "table";
    VirtualConsoleLogTypeEnum["trace"] = "trace";
    VirtualConsoleLogTypeEnum["dir"] = "dir";
    VirtualConsoleLogTypeEnum["dirxml"] = "dirxml";
    VirtualConsoleLogTypeEnum["group"] = "group";
    VirtualConsoleLogTypeEnum["groupCollapsed"] = "groupCollapsed";
    VirtualConsoleLogTypeEnum["debug"] = "debug";
    VirtualConsoleLogTypeEnum["timeLog"] = "timeLog";
    // Info
    VirtualConsoleLogTypeEnum["info"] = "info";
    VirtualConsoleLogTypeEnum["count"] = "count";
    VirtualConsoleLogTypeEnum["timeEnd"] = "timeEnd";
    // Warning
    VirtualConsoleLogTypeEnum["warn"] = "warn";
    VirtualConsoleLogTypeEnum["countReset"] = "countReset";
    // Error
    VirtualConsoleLogTypeEnum["error"] = "error";
    VirtualConsoleLogTypeEnum["assert"] = "assert";
})(VirtualConsoleLogTypeEnum || (VirtualConsoleLogTypeEnum = {}));
export default VirtualConsoleLogTypeEnum;
