// @scalar/api-reference/dist/features/Operation/components/callbacks/Callbacks.vue.js
import { defineComponent as k, createElementBlock as t, openBlock as e, createElementVNode as f, Fragment as o, renderList as n, unref as m, createBlock as h, createCommentVNode as b } from "vue";
import { isHttpMethod as C } from "@scalar/helpers/http/is-http-method";
import { getResolvedRef as g } from "@scalar/workspace-store/helpers/get-resolved-ref";
import y from "./Callback.vue.js";
const v = { class: "callbacks-list gap-3" }, R = /* @__PURE__ */ k({
  __name: "Callbacks",
  props: {
    path: {},
    method: {},
    callbacks: {},
    options: {}
  },
  setup(p) {
    return (l, a) => (e(), t("div", v, [
      a[0] || (a[0] = f("div", { class: "callbacks-title text-c-1 my-3 text-lg font-medium" }, " Callbacks ", -1)),
      (e(!0), t(o, null, n(l.callbacks, (i, s) => (e(), t(o, { key: s }, [
        (e(!0), t(o, null, n(m(g)(i), (u, c) => (e(), t(o, null, [
          (e(!0), t(o, null, n(u, (d, r) => (e(), t(o, { key: r }, [
            m(C)(r) ? (e(), h(y, {
              key: 0,
              callback: d,
              method: r,
              name: s,
              operationMethod: p.method,
              options: l.options,
              path: l.path,
              url: c
            }, null, 8, ["callback", "method", "name", "operationMethod", "options", "path", "url"])) : b("", !0)
          ], 64))), 128))
        ], 64))), 256))
      ], 64))), 128))
    ]));
  }
});
export {
  R as default
};
