// @scalar/api-client/dist/store/store.js
import { useModal as L } from "@scalar/components";
import { inject as ee, reactive as te, ref as re, toRaw as r } from "vue";
import { createStoreCollections as oe, extendedCollectionDataFactory as R } from "./collections.js";
import { createStoreCookies as se } from "./cookies.js";
import { createStoreEnvironments as de, extendedEnvironmentDataFactory as ae } from "./environment.js";
import { createStoreEvents as ne } from "./events.js";
import { importSpecFileFactory as ce } from "./import-spec.js";
import { createStoreRequestExamples as ie, extendedExampleDataFactory as me } from "./request-example.js";
import { createStoreRequests as le, extendedRequestDataFactory as pe } from "./requests.js";
import { createStoreSecuritySchemes as ue, extendedSecurityDataFactory as ve } from "./security-schemes.js";
import { createStoreServers as Se, extendedServerDataFactory as fe } from "./servers.js";
import { createStoreTags as Ee, extendedTagDataFactory as Me } from "./tags.js";
import { createStoreWorkspaces as xe, extendedWorkspaceDataFactory as ye } from "./workspace.js";
const Pe = ({
  useLocalStorage: e = !0,
  showSidebar: C = !0,
  proxyUrl: a,
  theme: E,
  hideClientButton: A = !1,
  _integration: D
}) => {
  const { collections: n, collectionMutators: c } = oe(e), { tags: i, tagMutators: m } = Ee(e), { requests: l, requestMutators: p } = le(e), { requestExamples: u, requestExampleMutators: v } = ie(e), { cookies: M, cookieMutators: b } = se(e), { environments: x, environmentMutators: y } = de(e), { servers: q, serverMutators: S } = Se(e), { securitySchemes: w, securitySchemeMutators: f } = ue(e), { workspaces: d, workspaceMutators: o } = xe(e), t = {
    collections: n,
    collectionMutators: c,
    tags: i,
    tagMutators: m,
    requests: l,
    requestMutators: p,
    requestExamples: u,
    requestExampleMutators: v,
    environmentMutators: y,
    serverMutators: S,
    securitySchemeMutators: f,
    workspaces: d,
    workspaceMutators: o
  }, { addTag: k, deleteTag: O } = Me(t), { addRequest: T, deleteRequest: I, findRequestParents: P } = pe(t, k), { deleteEnvironment: g } = ae(t), { addServer: j, deleteServer: B } = fe(t), { addCollection: K, deleteCollection: Y } = R(t), { addRequestExample: _, deleteRequestExample: H } = me(t), { addWorkspace: z, deleteWorkspace: G } = ye(t), { addSecurityScheme: J, deleteSecurityScheme: N } = ve(t), { addCollectionEnvironment: W, removeCollectionEnvironment: h } = R(t), Q = te([]), { importSpecFile: U, importSpecFromUrl: V } = ce(t), F = re((e ? localStorage?.getItem("sidebarWidth") : void 0) || "280px"), X = (s) => {
    F.value = s, e && localStorage?.setItem("sidebarWidth", s);
  }, Z = L();
  Object.values(d).forEach(({ uid: s }) => {
    a && o.edit(s, "proxyUrl", a), E && o.edit(s, "themeId", E);
  }), typeof window < "u" && (window.dataDump = () => ({
    collections: r(n),
    cookies: r(M),
    environments: r(x),
    requestExamples: r(u),
    requests: r(l),
    securitySchemes: r(w),
    servers: r(q),
    tags: r(i),
    workspaces: r(d)
  }));
  const $ = ne();
  return {
    // ---------------------------------------------------------------------------
    // STATE
    workspaces: d,
    collections: n,
    tags: i,
    cookies: M,
    environments: x,
    requestExamples: u,
    requests: l,
    servers: q,
    securitySchemes: w,
    modalState: Z,
    events: $,
    sidebarWidth: F,
    setSidebarWidth: X,
    proxyUrl: a,
    // ---------------------------------------------------------------------------
    // CONFIGURATION "PROPS"
    // TODO: move these to their own store
    hideClientButton: A,
    showSidebar: C,
    integration: D,
    // ---------------------------------------------------------------------------
    // METHODS
    importSpecFile: U,
    importSpecFromUrl: V,
    cookieMutators: b,
    collectionMutators: {
      ...c,
      rawAdd: c.add,
      add: K,
      delete: Y,
      addEnvironment: W,
      removeEnvironment: h
    },
    environmentMutators: {
      ...y,
      delete: g
    },
    requestMutators: {
      ...p,
      rawAdd: p.add,
      add: T,
      delete: I
    },
    findRequestParents: P,
    requestExampleMutators: {
      ...v,
      rawAdd: v.add,
      add: _,
      delete: H
    },
    requestHistory: Q,
    securitySchemeMutators: {
      ...f,
      rawAdd: f.add,
      add: J,
      delete: N
    },
    serverMutators: {
      ...S,
      rawAdd: S.add,
      add: j,
      delete: B
    },
    tagMutators: {
      ...m,
      rawAdd: m.add,
      add: k,
      delete: O
    },
    workspaceMutators: {
      ...o,
      rawAdd: o.add,
      add: z,
      delete: G
    },
    addCollectionEnvironment: W,
    removeCollectionEnvironment: h
  };
}, qe = "WORKSPACE_SYMBOL", ge = () => {
  const e = ee(qe);
  if (!e)
    throw new Error("Workspace store not provided");
  return e;
};
export {
  qe as WORKSPACE_SYMBOL,
  Pe as createWorkspaceStore,
  ge as useWorkspace
};
