// @scalar/api-client/dist/store/tags.js
import { tagSchema as h } from "@scalar/oas-utils/entities/spec";
import { schemaModel as o } from "@scalar/oas-utils/helpers";
import { LS_KEYS as T } from "@scalar/helpers/object/local-storage";
import { mutationFactory as p } from "@scalar/object-utils/mutator-record";
import { reactive as g } from "vue";
function I(d) {
  const a = g({}), i = p(a, g({}), d && T.TAG);
  return {
    tags: a,
    tagMutators: i
  };
}
function L({
  collectionMutators: d,
  collections: a,
  requests: i,
  requestMutators: u,
  tagMutators: f
}) {
  return { addTag: (t, n) => {
    const r = a[n], e = o(t, h, !1);
    return !e || !r ? console.error("INVALID TAG DATA", t) : (d.edit(n, "tags", [...r.tags, e.uid]), d.edit(n, "children", [...r.children, e.uid]), f.add(e), e);
  }, deleteTag: (t, n) => {
    const r = a[n];
    r && (d.edit(
      n,
      "tags",
      r.tags.filter((e) => e !== t.uid)
    ), d.edit(
      n,
      "children",
      r.children.filter((e) => e !== t.uid)
    ), t.children.forEach((e) => {
      const s = i[e];
      if (!s)
        return;
      const m = s.tags?.filter((c) => c !== t.name);
      u.edit(s.uid, "tags", m), !m?.length && !r.children.includes(e) && d.edit(n, "children", [...r.children, e]);
    }), f.delete(t.uid));
  } };
}
export {
  I as createStoreTags,
  L as extendedTagDataFactory
};
