// @scalar/api-reference/dist/components/Content/Schema/SchemaPropertyHeading.vue2.js
import { defineComponent as P, toRef as A, computed as s, createElementBlock as u, openBlock as a, createCommentVNode as l, createBlock as m, normalizeClass as I, renderSlot as L, Fragment as p, withCtx as n, createVNode as c, createTextVNode as i, toDisplayString as v, renderList as R, createSlots as T, unref as k } from "vue";
import { isDefined as f } from "@scalar/helpers/array/is-defined";
import { getResolvedRef as S } from "@scalar/workspace-store/helpers/get-resolved-ref";
import { isArraySchema as h, isStringSchema as N, isNumberSchema as B } from "@scalar/workspace-store/schemas/v3.1/strict/type-guards";
import _ from "../../ScreenReader.vue.js";
import { getSchemaType as V } from "./helpers/get-schema-type.js";
import { getModelName as $ } from "./helpers/schema-name.js";
import b from "./RenderString.vue.js";
import d from "./SchemaPropertyDetail.vue.js";
import C from "./SchemaPropertyExamples.vue.js";
import E from "../../Badge/Badge.vue.js";
const J = { class: "property-heading" }, F = {
  key: 1,
  class: "property-discriminator"
}, j = {
  key: 3,
  class: "property-additional"
}, z = {
  key: 4,
  class: "property-deprecated"
}, H = {
  key: 5,
  class: "property-const"
}, G = {
  key: 7,
  class: "property-write-only"
}, K = {
  key: 8,
  class: "property-read-only"
}, Q = {
  key: 9,
  class: "property-required"
}, le = /* @__PURE__ */ P({
  __name: "SchemaPropertyHeading",
  props: {
    value: {},
    enum: { type: Boolean },
    isDiscriminator: { type: Boolean, default: !1 },
    required: { type: Boolean, default: !1 },
    additional: { type: Boolean },
    withExamples: { type: Boolean, default: !0 },
    hideModelNames: { type: Boolean, default: !1 }
  },
  setup(q) {
    const r = q, y = A(r, "value"), x = s(() => {
      if (!y.value)
        return;
      const e = y.value;
      if (e.const !== void 0)
        return e.const;
      if (e.enum?.length === 1)
        return e.enum[0];
      if (h(e) && e.items) {
        const t = S(e.items);
        if (f(t.const))
          return t.const;
        if (t.enum?.length === 1)
          return t.enum[0];
      }
    }), w = s(() => {
      if (!y.value)
        return [];
      const e = y.value, t = [];
      return h(e) && ((e.minItems || e.maxItems) && t.push({
        key: "array-range",
        value: `${e.minItems || ""}…${e.maxItems || ""}`
      }), e.uniqueItems && t.push({
        key: "unique-items",
        value: "unique!"
      })), N(e) && (e.minLength && t.push({
        key: "min-length",
        prefix: "min length: ",
        value: e.minLength
      }), e.maxLength && t.push({
        key: "max-length",
        prefix: "max length: ",
        value: e.maxLength
      }), e.pattern && t.push({
        key: "pattern",
        value: e.pattern,
        code: !0,
        truncate: !0
      })), (N(e) || B(e)) && e.format && t.push({
        key: "format",
        value: e.format,
        truncate: !0
      }), B(e) && (f(e.exclusiveMinimum) && t.push({
        key: "exclusive-minimum",
        prefix: "greater than: ",
        value: e.exclusiveMinimum
      }), f(e.minimum) && t.push({
        key: "minimum",
        prefix: "min: ",
        value: e.minimum
      }), f(e.exclusiveMaximum) && t.push({
        key: "exclusive-maximum",
        prefix: "less than: ",
        value: e.exclusiveMaximum
      }), f(e.maximum) && t.push({
        key: "maximum",
        prefix: "max: ",
        value: e.maximum
      }), f(e.multipleOf) && t.push({
        key: "multiple-of",
        prefix: "multiple of: ",
        value: e.multipleOf
      })), t;
    }), D = s(() => r.value ? $(r.value, r.hideModelNames) : null), M = s(() => !r.value || !("type" in r.value) ? !1 : r.value.type === "array" ? !0 : !x.value), O = s(() => r.value ? D.value || V(r.value) : ""), g = s(() => {
      const e = r.value;
      return e?.default === null ? "null" : Array.isArray(e?.default) && e?.default.length === 1 ? String(e?.default[0]) : typeof e?.default == "string" ? JSON.stringify(e.default) : Array.isArray(e?.default) || typeof e?.default == "object" ? JSON.stringify(e?.default) : e?.default;
    });
    return (e, t) => (a(), u("div", J, [
      e.$slots.name ? (a(), u("div", {
        key: 0,
        class: I(["property-name", { deprecated: r.value?.deprecated }])
      }, [
        L(e.$slots, "name", {}, void 0, !0)
      ], 2)) : l("", !0),
      r.isDiscriminator ? (a(), u("div", F, " Discriminator ")) : l("", !0),
      r.value ? (a(), u(p, { key: 2 }, [
        M.value ? (a(), m(d, {
          key: 0,
          truncate: ""
        }, {
          default: n(() => [
            c(_, null, {
              default: n(() => t[0] || (t[0] = [
                i("Type: ")
              ])),
              _: 1,
              __: [0]
            }),
            i(v(O.value), 1)
          ]),
          _: 1
        })) : l("", !0),
        (a(!0), u(p, null, R(w.value, (o) => (a(), m(d, {
          key: o.key,
          code: o.code,
          truncate: o.truncate
        }, T({
          default: n(() => [
            o.key === "format" ? (a(), m(_, { key: 0 }, {
              default: n(() => t[1] || (t[1] = [
                i("Format:")
              ])),
              _: 1,
              __: [1]
            })) : o.key === "pattern" ? (a(), m(_, { key: 1 }, {
              default: n(() => t[2] || (t[2] = [
                i(" Pattern: ")
              ])),
              _: 1,
              __: [2]
            })) : l("", !0),
            i(" " + v(o.value), 1)
          ]),
          _: 2
        }, [
          o.prefix ? {
            name: "prefix",
            fn: n(() => [
              i(v(o.prefix), 1)
            ]),
            key: "0"
          } : void 0
        ]), 1032, ["code", "truncate"]))), 128)),
        r.enum ? (a(), m(d, { key: 1 }, {
          default: n(() => t[3] || (t[3] = [
            i("enum")
          ])),
          _: 1,
          __: [3]
        })) : l("", !0),
        g.value !== void 0 ? (a(), m(d, {
          key: 2,
          truncate: ""
        }, {
          prefix: n(() => t[4] || (t[4] = [
            i("default:")
          ])),
          default: n(() => [
            i(v(g.value), 1)
          ]),
          _: 1
        })) : l("", !0)
      ], 64)) : l("", !0),
      r.additional ? (a(), u("div", j, [
        r.value?.["x-additionalPropertiesName"] ? (a(), u(p, { key: 0 }, [
          i(v(r.value["x-additionalPropertiesName"]), 1)
        ], 64)) : (a(), u(p, { key: 1 }, [
          i("additional properties")
        ], 64))
      ])) : l("", !0),
      r.value?.deprecated ? (a(), u("div", z, [
        c(k(E), null, {
          default: n(() => t[5] || (t[5] = [
            i("deprecated")
          ])),
          _: 1,
          __: [5]
        })
      ])) : l("", !0),
      x.value !== void 0 ? (a(), u("div", H, [
        c(d, { truncate: "" }, {
          prefix: n(() => t[6] || (t[6] = [
            i("const: ")
          ])),
          default: n(() => [
            c(b, { value: x.value }, null, 8, ["value"])
          ]),
          _: 1
        })
      ])) : (a(), u(p, { key: 6 }, [
        r.value?.nullable === !0 ? (a(), m(d, { key: 0 }, {
          default: n(() => t[7] || (t[7] = [
            i(" nullable ")
          ])),
          _: 1,
          __: [7]
        })) : l("", !0)
      ], 64)),
      r.value?.writeOnly ? (a(), u("div", G, " write-only ")) : r.value?.readOnly ? (a(), u("div", K, " read-only ")) : l("", !0),
      r.required ? (a(), u("div", Q, " required ")) : l("", !0),
      r.withExamples ? (a(), m(C, {
        key: 10,
        example: r.value?.example || r.value && k(h)(r.value) && k(S)(r.value?.items)?.example,
        examples: r.value?.examples
      }, null, 8, ["example", "examples"])) : l("", !0)
    ]));
  }
});
export {
  le as default
};
