// @azure/data-tables/dist/commonjs/index.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestError = exports.AzureSASCredential = exports.AzureNamedKeyCredential = exports.odata = exports.TableClient = exports.TableTransaction = exports.TableServiceClient = void 0;
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./generatedModels.js"), exports);
tslib_1.__exportStar(require("./models.js"), exports);
tslib_1.__exportStar(require("./sas/index.js"), exports);
var TableServiceClient_js_1 = require("./TableServiceClient.js");
Object.defineProperty(exports, "TableServiceClient", { enumerable: true, get: function () { return TableServiceClient_js_1.TableServiceClient; } });
var TableTransaction_js_1 = require("./TableTransaction.js");
Object.defineProperty(exports, "TableTransaction", { enumerable: true, get: function () { return TableTransaction_js_1.TableTransaction; } });
var TableClient_js_1 = require("./TableClient.js");
Object.defineProperty(exports, "TableClient", { enumerable: true, get: function () { return TableClient_js_1.TableClient; } });
var odata_js_1 = require("./odata.js");
Object.defineProperty(exports, "odata", { enumerable: true, get: function () { return odata_js_1.odata; } });
var core_auth_1 = require("@azure/core-auth");
Object.defineProperty(exports, "AzureNamedKeyCredential", { enumerable: true, get: function () { return core_auth_1.AzureNamedKeyCredential; } });
Object.defineProperty(exports, "AzureSASCredential", { enumerable: true, get: function () { return core_auth_1.AzureSASCredential; } });
var core_rest_pipeline_1 = require("@azure/core-rest-pipeline");
Object.defineProperty(exports, "RestError", { enumerable: true, get: function () { return core_rest_pipeline_1.RestError; } });
