// @scalar/workspace-store/dist/schemas/navigation.js
import { HTTP_METHODS } from "@scalar/helpers/http/http-methods";
import { Type } from "@scalar/typebox";
import { compose } from "../schemas/compose.js";
import { TraversedEntryObjectRef } from "../schemas/v3.1/strict/ref-definitions.js";
const NavigationBaseSchemaDefinition = Type.Object({
  id: Type.String(),
  title: Type.String()
});
const TraversedDescriptionSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("text"),
    children: Type.Optional(Type.Array(TraversedEntryObjectRef))
  })
);
const TraversedExampleSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("example"),
    name: Type.String()
  })
);
const TraversedOperationSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("operation"),
    ref: Type.String(),
    method: Type.Union(HTTP_METHODS.map((method) => Type.Literal(method))),
    path: Type.String(),
    isDeprecated: Type.Optional(Type.Boolean()),
    children: Type.Optional(Type.Array(TraversedEntryObjectRef))
  })
);
const TraversedSchemaSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("model"),
    ref: Type.String(),
    name: Type.String()
  })
);
const TraversedWebhookSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("webhook"),
    ref: Type.String(),
    method: Type.Union(HTTP_METHODS.map((method) => Type.Literal(method))),
    name: Type.String(),
    isDeprecated: Type.Optional(Type.Boolean())
  })
);
const TraversedTagSchemaDefinition = compose(
  NavigationBaseSchemaDefinition,
  Type.Object({
    type: Type.Literal("tag"),
    name: Type.String(),
    description: Type.Optional(Type.String()),
    children: Type.Optional(Type.Array(TraversedEntryObjectRef)),
    isGroup: Type.Boolean(),
    isWebhooks: Type.Optional(Type.Boolean()),
    xKeys: Type.Optional(Type.Record(Type.String(), Type.Unknown()))
  })
);
const TraversedEntrySchemaDefinition = Type.Union([
  TraversedDescriptionSchemaDefinition,
  TraversedOperationSchemaDefinition,
  TraversedSchemaSchemaDefinition,
  TraversedTagSchemaDefinition,
  TraversedWebhookSchemaDefinition,
  TraversedExampleSchemaDefinition
]);
export {
  NavigationBaseSchemaDefinition,
  TraversedDescriptionSchemaDefinition,
  TraversedEntrySchemaDefinition,
  TraversedExampleSchemaDefinition,
  TraversedOperationSchemaDefinition,
  TraversedSchemaSchemaDefinition,
  TraversedTagSchemaDefinition,
  TraversedWebhookSchemaDefinition
};
