// @scalar/components/dist/components/ScalarToggle/ScalarToggleGroup.vue.js
import { defineComponent as n, mergeModels as s, useModel as f, createBlock as o, openBlock as t, unref as c, withCtx as u, createElementBlock as _, Fragment as i, renderList as p, createTextVNode as V, toDisplayString as v } from "vue";
import g from "./ScalarToggleInput.vue.js";
import k from "../ScalarForm/ScalarFormInputGroup.vue.js";
const x = {}, C = /* @__PURE__ */ n({
  ...x,
  __name: "ScalarToggleGroup",
  props: /* @__PURE__ */ s({
    options: { default: () => [] }
  }, {
    modelValue: { default: [] },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(r) {
    const l = f(r, "modelValue");
    return (m, B) => (t(), o(c(k), null, {
      default: u(() => [
        (t(!0), _(i, null, p(m.options, (e) => (t(), o(g, {
          modelValue: l.value?.some(({ value: a }) => a === e.value),
          "onUpdate:modelValue": (a) => l.value = a ? [...l.value, e] : l.value.filter(({ value: d }) => d !== e.value),
          key: e.value
        }, {
          default: u(() => [
            V(v(e.label), 1)
          ]),
          _: 2
        }, 1032, ["modelValue", "onUpdate:modelValue"]))), 128))
      ]),
      _: 1
    }));
  }
});
export {
  C as default
};
