// @scalar/components/dist/components/ScalarToggle/ScalarToggleInput.vue.js
import { defineComponent as u, useModel as d, createBlock as i, openBlock as c, unref as t, mergeProps as l, withCtx as f, createElementVNode as p, createVNode as _, renderSlot as V } from "vue";
import { useBindCx as x } from "@scalar/use-hooks/useBindCx";
import g from "./ScalarToggle.vue.js";
import h from "../ScalarForm/ScalarFormInput.vue.js";
const v = { class: "flex-1 text-left min-w-0 truncate" }, B = {}, w = /* @__PURE__ */ u({
  ...B,
  inheritAttrs: !1,
  __name: "ScalarToggleInput",
  props: {
    modelValue: { type: Boolean },
    modelModifiers: {}
  },
  emits: ["update:modelValue"],
  setup(r) {
    const e = d(r, "modelValue"), { classCx: a, otherAttrs: s } = x();
    return (n, o) => (c(), i(t(h), l(
      { is: "label" },
      t(a)("cursor-pointer gap-2 hover:bg-b-2", { "text-c-1": e.value })
    ), {
      default: f(() => [
        p("div", v, [
          V(n.$slots, "default")
        ]),
        _(g, l({
          modelValue: e.value,
          "onUpdate:modelValue": o[0] || (o[0] = (m) => e.value = m),
          class: "shrink-0"
        }, t(s)), null, 16, ["modelValue"])
      ]),
      _: 3
    }, 16));
  }
});
export {
  w as default
};
