// @scalar/components/dist/components/ScalarToggle/ScalarToggle.vue.js
import { defineComponent as c, mergeModels as u, useModel as m, createElementBlock as t, openBlock as o, normalizeClass as n, unref as r, createElementVNode as b, createCommentVNode as p, toDisplayString as f } from "vue";
import { cva as h, cx as v } from "@scalar/use-hooks/useBindCx";
const g = ["aria-checked", "aria-disabled"], k = {
  key: 0,
  class: "sr-only"
}, x = /* @__PURE__ */ c({
  __name: "ScalarToggle",
  props: /* @__PURE__ */ u({
    disabled: { type: Boolean },
    label: {}
  }, {
    modelValue: { type: Boolean, default: !1 },
    modelModifiers: {}
  }),
  emits: ["update:modelValue"],
  setup(l) {
    const s = l, e = m(l, "modelValue");
    function d() {
      s.disabled || (e.value = !e.value);
    }
    const i = h({
      base: "relative h-3.5 min-w-6 w-6 cursor-pointer rounded-full bg-b-3 transition-colors duration-300",
      variants: {
        checked: { true: "bg-c-accent" },
        disabled: { true: "cursor-not-allowed opacity-40" }
      }
    });
    return (a, y) => (o(), t("button", {
      "aria-checked": e.value,
      "aria-disabled": a.disabled,
      class: n(r(v)(r(i)({ checked: e.value, disabled: a.disabled }))),
      role: "switch",
      type: "button",
      onClick: d
    }, [
      b("div", {
        class: n(["absolute left-px top-px flex h-3 w-3 items-center justify-center rounded-full bg-b-1 text-c-accent transition-transform duration-300", { "translate-x-2.5": e.value }])
      }, null, 2),
      a.label ? (o(), t("span", k, f(a.label), 1)) : p("", !0)
    ], 10, g));
  }
});
export {
  x as default
};
