// @scalar/components/dist/components/ScalarFloating/ScalarFloating.vue.js
import { defineComponent as w, ref as d, computed as a, createElementBlock as b, openBlock as s, Fragment as z, createElementVNode as m, createBlock as S, createCommentVNode as $, normalizeClass as k, renderSlot as u, unref as n, withCtx as B, normalizeStyle as R } from "vue";
import { getSideAxis as c } from "@floating-ui/utils";
import { useFloating as x, autoUpdate as E, offset as F, flip as C, shift as M, size as H } from "@floating-ui/vue";
import { useResizeWithTarget as N } from "./useResizeWithTarget.js";
import T from "../ScalarTeleport/ScalarTeleport.vue.js";
const D = /* @__PURE__ */ w({
  inheritAttrs: !1,
  __name: "ScalarFloating",
  props: {
    placement: {},
    offset: { type: [Number, Object, Function], default: 5 },
    resize: { type: Boolean, default: !1 },
    target: {},
    middleware: { default: () => [] },
    teleport: { type: [Boolean, String] }
  },
  setup(t) {
    const o = d(null), l = d(null), i = a(() => {
      if (typeof window < "u" && l.value) {
        if (typeof t.target == "string") {
          const e = document.getElementById(t.target);
          if (e)
            return e;
          console.warn(`ScalarFloating: Target with id="${t.target}" not found`);
        } else if (t.target instanceof HTMLElement)
          return t.target;
        return l.value.children?.[0] || l.value;
      }
    }), r = N(i, {
      enabled: a(() => t.resize)
    }), g = a(
      () => c(t.placement ?? "bottom") === "y" ? r.width.value : void 0
    ), h = a(
      () => c(t.placement ?? "bottom") === "x" ? r.height.value : void 0
    ), { floatingStyles: p, middlewareData: y } = x(i, o, {
      placement: a(() => t.placement ?? "bottom"),
      whileElementsMounted: E,
      middleware: a(() => [
        F(t.offset),
        C(),
        M({ padding: 10 }),
        H({
          apply({ availableWidth: e, availableHeight: f, elements: v }) {
            Object.assign(v.floating.style, {
              maxWidth: `${Math.max(0, e) - 20}px`,
              maxHeight: `${Math.max(0, f) - 20}px`
            });
          }
        }),
        ...t.middleware
      ])
    });
    return (e, f) => (s(), b(z, null, [
      m("div", {
        ref_key: "wrapperRef",
        ref: l,
        class: k({ contents: !!e.$slots.default })
      }, [
        u(e.$slots, "default")
      ], 2),
      e.$slots.floating ? (s(), S(n(T), {
        key: 0,
        disabled: !e.teleport,
        to: typeof e.teleport == "string" ? e.teleport : void 0
      }, {
        default: B(() => [
          m("div", {
            ref_key: "floatingRef",
            ref: o,
            class: "relative z-context",
            style: R(n(p))
          }, [
            u(e.$slots, "floating", {
              data: n(y),
              height: h.value,
              width: g.value
            })
          ], 4)
        ]),
        _: 3
      }, 8, ["disabled", "to"])) : $("", !0)
    ], 64));
  }
});
export {
  D as default
};
