// @scalar/workspace-store/dist/navigation/helpers/traverse-document.js
import { getNavigationOptions } from "../../navigation/get-navigation-options.js";
import { traverseDescription } from "./traverse-description.js";
import { traversePaths } from "./traverse-paths.js";
import { traverseSchemas } from "./traverse-schemas.js";
import { traverseTags } from "./traverse-tags.js";
import { traverseWebhooks } from "./traverse-webhooks.js";
const traverseDocument = (document, config) => {
  const { hideModels, tagsSorter, operationsSorter, getHeadingId, getOperationId, getWebhookId, getModelId, getTagId } = getNavigationOptions(config);
  const tagsMap = new Map(
    document.tags?.map((tag) => [tag.name ?? "Untitled Tag", { tag, entries: [] }]) ?? []
  );
  const entries = traverseDescription(document.info?.description, getHeadingId);
  traversePaths(document, tagsMap, getOperationId);
  const untaggedWebhooks = traverseWebhooks(document, tagsMap, getWebhookId);
  const tagsEntries = traverseTags(document, tagsMap, {
    getTagId,
    tagsSorter,
    operationsSorter
  });
  entries.push(...tagsEntries);
  if (untaggedWebhooks.length) {
    entries.push({
      type: "tag",
      id: getWebhookId({ name: "" }),
      title: "Webhooks",
      name: "Webhooks",
      children: untaggedWebhooks,
      isGroup: false,
      isWebhooks: true
    });
  }
  if (!hideModels && document.components?.schemas) {
    const untaggedModels = traverseSchemas(document, tagsMap, getModelId);
    if (untaggedModels.length) {
      entries.push({
        type: "text",
        id: "models",
        title: "Models",
        children: untaggedModels
      });
    }
  }
  return { entries };
};
export {
  traverseDocument
};
