// @azure/arm-appservice/dist-esm/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link AppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem} that the service accepts. */
export var KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem;
(function (KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem) {
    KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem["RegistrationStatusNotSupportedForRenewal"] = "RegistrationStatusNotSupportedForRenewal";
    KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem["ExpirationNotInRenewalTimeRange"] = "ExpirationNotInRenewalTimeRange";
    KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem["SubscriptionNotActive"] = "SubscriptionNotActive";
})(KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem || (KnownAppServiceCertificateOrderPropertiesAppServiceCertificateNotRenewableReasonsItem = {}));
/** Known values of {@link AppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem} that the service accepts. */
export var KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem;
(function (KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem) {
    KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem["RegistrationStatusNotSupportedForRenewal"] = "RegistrationStatusNotSupportedForRenewal";
    KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem["ExpirationNotInRenewalTimeRange"] = "ExpirationNotInRenewalTimeRange";
    KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem["SubscriptionNotActive"] = "SubscriptionNotActive";
})(KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem || (KnownAppServiceCertificateOrderPatchResourcePropertiesAppServiceCertificateNotRenewableReasonsItem = {}));
/** Known values of {@link DomainPropertiesDomainNotRenewableReasonsItem} that the service accepts. */
export var KnownDomainPropertiesDomainNotRenewableReasonsItem;
(function (KnownDomainPropertiesDomainNotRenewableReasonsItem) {
    KnownDomainPropertiesDomainNotRenewableReasonsItem["RegistrationStatusNotSupportedForRenewal"] = "RegistrationStatusNotSupportedForRenewal";
    KnownDomainPropertiesDomainNotRenewableReasonsItem["ExpirationNotInRenewalTimeRange"] = "ExpirationNotInRenewalTimeRange";
    KnownDomainPropertiesDomainNotRenewableReasonsItem["SubscriptionNotActive"] = "SubscriptionNotActive";
})(KnownDomainPropertiesDomainNotRenewableReasonsItem || (KnownDomainPropertiesDomainNotRenewableReasonsItem = {}));
/** Known values of {@link DomainPatchResourcePropertiesDomainNotRenewableReasonsItem} that the service accepts. */
export var KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem;
(function (KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem) {
    KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem["RegistrationStatusNotSupportedForRenewal"] = "RegistrationStatusNotSupportedForRenewal";
    KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem["ExpirationNotInRenewalTimeRange"] = "ExpirationNotInRenewalTimeRange";
    KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem["SubscriptionNotActive"] = "SubscriptionNotActive";
})(KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem || (KnownDomainPatchResourcePropertiesDomainNotRenewableReasonsItem = {}));
/** Known values of {@link LoadBalancingMode} that the service accepts. */
export var KnownLoadBalancingMode;
(function (KnownLoadBalancingMode) {
    KnownLoadBalancingMode["None"] = "None";
    KnownLoadBalancingMode["Web"] = "Web";
    KnownLoadBalancingMode["Publishing"] = "Publishing";
    KnownLoadBalancingMode["WebPublishing"] = "Web, Publishing";
})(KnownLoadBalancingMode || (KnownLoadBalancingMode = {}));
/** Known values of {@link ScmType} that the service accepts. */
export var KnownScmType;
(function (KnownScmType) {
    KnownScmType["None"] = "None";
    KnownScmType["Dropbox"] = "Dropbox";
    KnownScmType["Tfs"] = "Tfs";
    KnownScmType["LocalGit"] = "LocalGit";
    KnownScmType["GitHub"] = "GitHub";
    KnownScmType["CodePlexGit"] = "CodePlexGit";
    KnownScmType["CodePlexHg"] = "CodePlexHg";
    KnownScmType["BitbucketGit"] = "BitbucketGit";
    KnownScmType["BitbucketHg"] = "BitbucketHg";
    KnownScmType["ExternalGit"] = "ExternalGit";
    KnownScmType["ExternalHg"] = "ExternalHg";
    KnownScmType["OneDrive"] = "OneDrive";
    KnownScmType["VSO"] = "VSO";
    KnownScmType["Vstsrm"] = "VSTSRM";
})(KnownScmType || (KnownScmType = {}));
/** Known values of {@link IpFilterTag} that the service accepts. */
export var KnownIpFilterTag;
(function (KnownIpFilterTag) {
    KnownIpFilterTag["Default"] = "Default";
    KnownIpFilterTag["XffProxy"] = "XffProxy";
    KnownIpFilterTag["ServiceTag"] = "ServiceTag";
})(KnownIpFilterTag || (KnownIpFilterTag = {}));
/** Known values of {@link SupportedTlsVersions} that the service accepts. */
export var KnownSupportedTlsVersions;
(function (KnownSupportedTlsVersions) {
    KnownSupportedTlsVersions["One0"] = "1.0";
    KnownSupportedTlsVersions["One1"] = "1.1";
    KnownSupportedTlsVersions["One2"] = "1.2";
})(KnownSupportedTlsVersions || (KnownSupportedTlsVersions = {}));
/** Known values of {@link FtpsState} that the service accepts. */
export var KnownFtpsState;
(function (KnownFtpsState) {
    KnownFtpsState["AllAllowed"] = "AllAllowed";
    KnownFtpsState["FtpsOnly"] = "FtpsOnly";
    KnownFtpsState["Disabled"] = "Disabled";
})(KnownFtpsState || (KnownFtpsState = {}));
/** Known values of {@link RouteType} that the service accepts. */
export var KnownRouteType;
(function (KnownRouteType) {
    KnownRouteType["Default"] = "DEFAULT";
    KnownRouteType["Inherited"] = "INHERITED";
    KnownRouteType["Static"] = "STATIC";
})(KnownRouteType || (KnownRouteType = {}));
/** Known values of {@link ContainerAppProvisioningState} that the service accepts. */
export var KnownContainerAppProvisioningState;
(function (KnownContainerAppProvisioningState) {
    KnownContainerAppProvisioningState["InProgress"] = "InProgress";
    KnownContainerAppProvisioningState["Succeeded"] = "Succeeded";
    KnownContainerAppProvisioningState["Failed"] = "Failed";
    KnownContainerAppProvisioningState["Canceled"] = "Canceled";
})(KnownContainerAppProvisioningState || (KnownContainerAppProvisioningState = {}));
/** Known values of {@link ActiveRevisionsMode} that the service accepts. */
export var KnownActiveRevisionsMode;
(function (KnownActiveRevisionsMode) {
    KnownActiveRevisionsMode["Multiple"] = "multiple";
    KnownActiveRevisionsMode["Single"] = "single";
})(KnownActiveRevisionsMode || (KnownActiveRevisionsMode = {}));
/** Known values of {@link IngressTransportMethod} that the service accepts. */
export var KnownIngressTransportMethod;
(function (KnownIngressTransportMethod) {
    KnownIngressTransportMethod["Auto"] = "auto";
    KnownIngressTransportMethod["Http"] = "http";
    KnownIngressTransportMethod["Http2"] = "http2";
})(KnownIngressTransportMethod || (KnownIngressTransportMethod = {}));
/** Known values of {@link RevisionHealthState} that the service accepts. */
export var KnownRevisionHealthState;
(function (KnownRevisionHealthState) {
    KnownRevisionHealthState["Healthy"] = "Healthy";
    KnownRevisionHealthState["Unhealthy"] = "Unhealthy";
    KnownRevisionHealthState["None"] = "None";
})(KnownRevisionHealthState || (KnownRevisionHealthState = {}));
/** Known values of {@link RevisionProvisioningState} that the service accepts. */
export var KnownRevisionProvisioningState;
(function (KnownRevisionProvisioningState) {
    KnownRevisionProvisioningState["Provisioning"] = "Provisioning";
    KnownRevisionProvisioningState["Provisioned"] = "Provisioned";
    KnownRevisionProvisioningState["Failed"] = "Failed";
    KnownRevisionProvisioningState["Deprovisioning"] = "Deprovisioning";
    KnownRevisionProvisioningState["Deprovisioned"] = "Deprovisioned";
})(KnownRevisionProvisioningState || (KnownRevisionProvisioningState = {}));
/** Known values of {@link Enum15} that the service accepts. */
export var KnownEnum15;
(function (KnownEnum15) {
    KnownEnum15["Windows"] = "Windows";
    KnownEnum15["Linux"] = "Linux";
    KnownEnum15["WindowsFunctions"] = "WindowsFunctions";
    KnownEnum15["LinuxFunctions"] = "LinuxFunctions";
    KnownEnum15["All"] = "All";
})(KnownEnum15 || (KnownEnum15 = {}));
/** Known values of {@link Enum16} that the service accepts. */
export var KnownEnum16;
(function (KnownEnum16) {
    KnownEnum16["Windows"] = "Windows";
    KnownEnum16["Linux"] = "Linux";
    KnownEnum16["All"] = "All";
})(KnownEnum16 || (KnownEnum16 = {}));
/** Known values of {@link Enum17} that the service accepts. */
export var KnownEnum17;
(function (KnownEnum17) {
    KnownEnum17["Windows"] = "Windows";
    KnownEnum17["Linux"] = "Linux";
    KnownEnum17["All"] = "All";
})(KnownEnum17 || (KnownEnum17 = {}));
/** Known values of {@link Enum18} that the service accepts. */
export var KnownEnum18;
(function (KnownEnum18) {
    KnownEnum18["Windows"] = "Windows";
    KnownEnum18["Linux"] = "Linux";
    KnownEnum18["All"] = "All";
})(KnownEnum18 || (KnownEnum18 = {}));
/** Known values of {@link Enum19} that the service accepts. */
export var KnownEnum19;
(function (KnownEnum19) {
    KnownEnum19["Windows"] = "Windows";
    KnownEnum19["Linux"] = "Linux";
    KnownEnum19["All"] = "All";
})(KnownEnum19 || (KnownEnum19 = {}));
/** Known values of {@link Enum20} that the service accepts. */
export var KnownEnum20;
(function (KnownEnum20) {
    KnownEnum20["Windows"] = "Windows";
    KnownEnum20["Linux"] = "Linux";
    KnownEnum20["WindowsFunctions"] = "WindowsFunctions";
    KnownEnum20["LinuxFunctions"] = "LinuxFunctions";
    KnownEnum20["All"] = "All";
})(KnownEnum20 || (KnownEnum20 = {}));
/** Known values of {@link ResourceScopeType} that the service accepts. */
export var KnownResourceScopeType;
(function (KnownResourceScopeType) {
    KnownResourceScopeType["ServerFarm"] = "ServerFarm";
    KnownResourceScopeType["Subscription"] = "Subscription";
    KnownResourceScopeType["WebSite"] = "WebSite";
})(KnownResourceScopeType || (KnownResourceScopeType = {}));
/** Known values of {@link CheckNameResourceTypes} that the service accepts. */
export var KnownCheckNameResourceTypes;
(function (KnownCheckNameResourceTypes) {
    KnownCheckNameResourceTypes["Site"] = "Site";
    KnownCheckNameResourceTypes["Slot"] = "Slot";
    KnownCheckNameResourceTypes["HostingEnvironment"] = "HostingEnvironment";
    KnownCheckNameResourceTypes["PublishingUser"] = "PublishingUser";
    KnownCheckNameResourceTypes["MicrosoftWebSites"] = "Microsoft.Web/sites";
    KnownCheckNameResourceTypes["MicrosoftWebSitesSlots"] = "Microsoft.Web/sites/slots";
    KnownCheckNameResourceTypes["MicrosoftWebHostingEnvironments"] = "Microsoft.Web/hostingEnvironments";
    KnownCheckNameResourceTypes["MicrosoftWebPublishingUsers"] = "Microsoft.Web/publishingUsers";
})(KnownCheckNameResourceTypes || (KnownCheckNameResourceTypes = {}));
/** Known values of {@link InAvailabilityReasonType} that the service accepts. */
export var KnownInAvailabilityReasonType;
(function (KnownInAvailabilityReasonType) {
    KnownInAvailabilityReasonType["Invalid"] = "Invalid";
    KnownInAvailabilityReasonType["AlreadyExists"] = "AlreadyExists";
})(KnownInAvailabilityReasonType || (KnownInAvailabilityReasonType = {}));
/** Known values of {@link SkuName} that the service accepts. */
export var KnownSkuName;
(function (KnownSkuName) {
    KnownSkuName["Free"] = "Free";
    KnownSkuName["Shared"] = "Shared";
    KnownSkuName["Basic"] = "Basic";
    KnownSkuName["Standard"] = "Standard";
    KnownSkuName["Premium"] = "Premium";
    KnownSkuName["Dynamic"] = "Dynamic";
    KnownSkuName["Isolated"] = "Isolated";
    KnownSkuName["IsolatedV2"] = "IsolatedV2";
    KnownSkuName["PremiumV2"] = "PremiumV2";
    KnownSkuName["PremiumV3"] = "PremiumV3";
    KnownSkuName["PremiumContainer"] = "PremiumContainer";
    KnownSkuName["ElasticPremium"] = "ElasticPremium";
    KnownSkuName["ElasticIsolated"] = "ElasticIsolated";
})(KnownSkuName || (KnownSkuName = {}));
/** Known values of {@link ValidateResourceTypes} that the service accepts. */
export var KnownValidateResourceTypes;
(function (KnownValidateResourceTypes) {
    KnownValidateResourceTypes["ServerFarm"] = "ServerFarm";
    KnownValidateResourceTypes["Site"] = "Site";
    KnownValidateResourceTypes["MicrosoftWebHostingEnvironments"] = "Microsoft.Web/hostingEnvironments";
})(KnownValidateResourceTypes || (KnownValidateResourceTypes = {}));
/** Known values of {@link EnterpriseGradeCdnStatus} that the service accepts. */
export var KnownEnterpriseGradeCdnStatus;
(function (KnownEnterpriseGradeCdnStatus) {
    KnownEnterpriseGradeCdnStatus["Enabled"] = "Enabled";
    KnownEnterpriseGradeCdnStatus["Enabling"] = "Enabling";
    KnownEnterpriseGradeCdnStatus["Disabled"] = "Disabled";
    KnownEnterpriseGradeCdnStatus["Disabling"] = "Disabling";
})(KnownEnterpriseGradeCdnStatus || (KnownEnterpriseGradeCdnStatus = {}));
/** Known values of {@link BuildStatus} that the service accepts. */
export var KnownBuildStatus;
(function (KnownBuildStatus) {
    KnownBuildStatus["WaitingForDeployment"] = "WaitingForDeployment";
    KnownBuildStatus["Uploading"] = "Uploading";
    KnownBuildStatus["Deploying"] = "Deploying";
    KnownBuildStatus["Ready"] = "Ready";
    KnownBuildStatus["Failed"] = "Failed";
    KnownBuildStatus["Deleting"] = "Deleting";
    KnownBuildStatus["Detached"] = "Detached";
})(KnownBuildStatus || (KnownBuildStatus = {}));
/** Known values of {@link TriggerTypes} that the service accepts. */
export var KnownTriggerTypes;
(function (KnownTriggerTypes) {
    KnownTriggerTypes["HttpTrigger"] = "HttpTrigger";
    KnownTriggerTypes["Unknown"] = "Unknown";
})(KnownTriggerTypes || (KnownTriggerTypes = {}));
/** Known values of {@link CustomDomainStatus} that the service accepts. */
export var KnownCustomDomainStatus;
(function (KnownCustomDomainStatus) {
    KnownCustomDomainStatus["RetrievingValidationToken"] = "RetrievingValidationToken";
    KnownCustomDomainStatus["Validating"] = "Validating";
    KnownCustomDomainStatus["Adding"] = "Adding";
    KnownCustomDomainStatus["Ready"] = "Ready";
    KnownCustomDomainStatus["Failed"] = "Failed";
    KnownCustomDomainStatus["Deleting"] = "Deleting";
})(KnownCustomDomainStatus || (KnownCustomDomainStatus = {}));
/** Known values of {@link DatabaseType} that the service accepts. */
export var KnownDatabaseType;
(function (KnownDatabaseType) {
    KnownDatabaseType["SqlAzure"] = "SqlAzure";
    KnownDatabaseType["MySql"] = "MySql";
    KnownDatabaseType["LocalMySql"] = "LocalMySql";
    KnownDatabaseType["PostgreSql"] = "PostgreSql";
})(KnownDatabaseType || (KnownDatabaseType = {}));
/** Known values of {@link PublishingProfileFormat} that the service accepts. */
export var KnownPublishingProfileFormat;
(function (KnownPublishingProfileFormat) {
    KnownPublishingProfileFormat["FileZilla3"] = "FileZilla3";
    KnownPublishingProfileFormat["WebDeploy"] = "WebDeploy";
    KnownPublishingProfileFormat["Ftp"] = "Ftp";
})(KnownPublishingProfileFormat || (KnownPublishingProfileFormat = {}));
