// @scalar/components/dist/components/ScalarSearchResults/ScalarSearchResultItem.vue.js
import { defineComponent as c, createElementBlock as o, openBlock as t, mergeProps as d, unref as l, createCommentVNode as s, createElementVNode as n, renderSlot as r, createBlock as m } from "vue";
import { useBindCx as p } from "@scalar/use-hooks/useBindCx";
import u from "../ScalarIcon/ScalarIconLegacyAdapter.vue.js";
const f = ["aria-selected"], h = {
  key: 0,
  class: "flex h-fit items-center text-sm font-medium text-c-3 group-hover:text-c-1"
}, _ = { class: "flex min-w-0 flex-1 flex-col gap-0.75" }, b = { class: "flex items-center gap-1" }, x = { class: "flex-1 truncate zoomed:!whitespace-normal break-words text-base font-medium" }, k = {
  key: 0,
  class: "text-base text-c-2"
}, g = {
  key: 0,
  class: "truncate zoomed:!whitespace-normal break-words text-base text-c-2"
}, B = /* @__PURE__ */ c({
  inheritAttrs: !1,
  __name: "ScalarSearchResultItem",
  props: {
    icon: {},
    selected: { type: Boolean }
  },
  setup(v) {
    const { cx: a } = p();
    return (e, i) => (t(), o("a", d(
      {
        "aria-selected": e.selected,
        role: "option",
        tabindex: "-1"
      },
      l(a)(
        "group flex cursor-pointer gap-2 rounded px-2 py-1.5 no-underline hover:bg-b-2",
        { "bg-b-2": e.selected }
      )
    ), [
      e.icon ? (t(), o("div", h, [
        r(e.$slots, "icon", {}, () => [
          e.icon ? (t(), m(l(u), {
            key: 0,
            icon: e.icon,
            class: "size-4"
          }, null, 8, ["icon"])) : s("", !0)
        ]),
        i[0] || (i[0] = n("span", null, " ", -1))
      ])) : s("", !0),
      n("div", _, [
        n("div", b, [
          n("div", x, [
            r(e.$slots, "default")
          ]),
          e.$slots.addon ? (t(), o("div", k, [
            r(e.$slots, "addon")
          ])) : s("", !0)
        ]),
        e.$slots.description ? (t(), o("div", g, [
          r(e.$slots, "description")
        ])) : s("", !0)
      ])
    ], 16, f));
  }
});
export {
  B as default
};
