// @scalar/components/dist/components/ScalarTooltip/ScalarHotkeyTooltip.vue.js
import { defineComponent as c, ref as i, computed as o, render as f, h as l, createElementBlock as m, openBlock as s, normalizeClass as u, renderSlot as h } from "vue";
import { DEFAULT_OFFSET as y, DEFAULT_DELAY as k } from "./constants.js";
import { useTooltip as p } from "./useTooltip.js";
import a from "../ScalarHotkey/ScalarHotkey.vue.js";
const T = {}, w = /* @__PURE__ */ c({
  ...T,
  __name: "ScalarHotkeyTooltip",
  props: {
    content: { default: "" },
    hotkey: {},
    modifier: {},
    delay: { default: () => k },
    placement: { default: "top" },
    offset: { default: () => y }
  },
  setup(e) {
    const n = i(null), d = o(() => {
      if (!document) return "";
      const t = document.createElement("div"), r = "-m-0.5 border-none p-0 grid grid-flow-col *:border *:h-5 *:min-w-5 *:border-border-tooltip *:rounded *:px-1 *:flex *:items-center *:justify-center";
      return e.content ? f(
        l("div", { class: "flex items-center gap-2" }, [
          e.content,
          l(a, {
            class: r,
            hotkey: e.hotkey,
            modifier: e.modifier
          })
        ]),
        t
      ) : f(
        l(a, {
          class: r,
          hotkey: e.hotkey,
          modifier: e.modifier
        }),
        t
      ), t.innerHTML;
    });
    return p({
      content: d,
      contentTarget: "innerHTML",
      delay: o(() => e.delay),
      placement: o(() => e.placement),
      offset: o(() => e.offset),
      targetRef: o(
        () => n.value?.children?.[0] || n.value || void 0
      )
    }), (t, r) => (s(), m("div", {
      ref_key: "wrapperRef",
      ref: n,
      class: u({ contents: !!t.$slots.default })
    }, [
      h(t.$slots, "default")
    ], 2));
  }
});
export {
  w as default
};
