// @scalar/components/dist/components/ScalarTooltip/useTooltip.js
import { useFloating as E, autoUpdate as y, flip as T, shift as g } from "@floating-ui/vue";
import { watch as m, unref as o, ref as p, computed as c } from "vue";
import { ELEMENT_ID as v, ELEMENT_CLASS as L, DEFAULT_DELAY as h, DEFAULT_OFFSET as b } from "./constants.js";
const d = p(), t = p(), s = p(), { floatingStyles: i } = E(
  c(() => o(s.value?.targetRef)),
  t,
  {
    placement: c(() => o(s.value?.placement)),
    whileElementsMounted: y,
    middleware: c(() => [T(), g()])
  }
);
m(i, () => {
  t.value && (t.value.style.position = i.value.position, t.value.style.top = i.value.top, t.value.style.left = i.value.left, t.value.style.transform = i.value.transform ?? "", t.value.style.willChange = i.value.willChange ?? "");
});
m(
  s,
  (e) => {
    if (t.value)
      if (e) {
        const a = o(e?.contentTarget) ?? "textContent";
        t.value[a] = o(e?.content) ?? "";
        const n = o(e?.offset) ?? b;
        t.value.style.setProperty("--scalar-tooltip-offset", `${n}px`), t.value.style.setProperty("display", "block");
      } else
        t.value.innerHTML = "", t.value.style.removeProperty("--scalar-tooltip-offset"), t.value.style.setProperty("display", "none");
  },
  { deep: !0 }
);
function A() {
  if (typeof document > "u" || typeof window > "u" || t.value)
    return;
  const e = document.getElementById(v);
  e ? t.value = e : (t.value = document.createElement("div"), t.value.role = "tooltip", t.value.id = v, t.value.classList.add(L), t.value.classList.add("scalar-app"), t.value.style.setProperty("display", "none"), t.value.addEventListener("mouseleave", u), document.body.appendChild(t.value));
}
function u(e) {
  M(e) && (f(), s.value = void 0);
}
function P(e) {
  e.key === "Escape" && (e.stopPropagation(), u(e));
}
function f() {
  d.value && (clearTimeout(d.value), d.value = void 0);
}
function w(e) {
  const a = [];
  let n = e.parentElement;
  for (; n; )
    a.push(n), n = n.parentElement;
  return a;
}
function M(e) {
  const a = o(s.value?.targetRef);
  if (e instanceof MouseEvent && e.relatedTarget instanceof Element && a) {
    const n = w(e.relatedTarget);
    return e.relatedTarget.id !== v && !n.some((l) => l.id === v) && e.relatedTarget !== a;
  }
  return !0;
}
function F(e) {
  A();
  function a(l) {
    const r = o(e.delay) ?? h;
    f(), r > 0 ? d.value = setTimeout(() => n(), r) : n();
  }
  function n(l) {
    f(), document.addEventListener("keydown", P, { once: !0, capture: !0 }), s.value = e;
  }
  m(
    () => o(e.targetRef),
    (l, r) => {
      r && (r.removeEventListener("mouseenter", a), r.removeEventListener("mouseleave", u), r.removeEventListener("focus", n), r.removeEventListener("blur", u), r.removeAttribute("aria-describedby")), l && (l.addEventListener("mouseenter", a), l.addEventListener("mouseleave", u), l.addEventListener("focus", n), l.addEventListener("blur", u), l.setAttribute("aria-describedby", v));
    },
    { immediate: !0 }
  );
}
export {
  A as initializeTooltipElement,
  F as useTooltip
};
