// @scalar/components/dist/components/ScalarFileUpload/ScalarFileUploadInputCompact.vue.js
import { defineComponent as c, createElementBlock as o, openBlock as s, mergeProps as m, unref as n, createElementVNode as l, renderSlot as i, createVNode as d, toDisplayString as r, createCommentVNode as u } from "vue";
import { ScalarIconUploadSimple as f } from "@scalar/icons";
import { useBindCx as x } from "@scalar/use-hooks/useBindCx";
const _ = { class: "flex items-center gap-1 whitespace-nowrap text-c-2" }, b = {
  key: 0,
  class: "text-c-3 text-xs"
}, k = {}, h = /* @__PURE__ */ c({
  ...k,
  inheritAttrs: !1,
  __name: "ScalarFileUploadInputCompact",
  props: {
    multiple: { type: Boolean },
    extensions: {}
  },
  emits: ["click"],
  setup(y) {
    const { cx: p } = x();
    return (e, t) => (s(), o("button", m(
      { type: "button" },
      n(p)(
        "flex items-center justify-center gap-1 rounded flex-col px-4 py-3 hover:bg-b-2"
      ),
      {
        onClick: t[0] || (t[0] = (a) => e.$emit("click", a))
      }
    ), [
      l("div", _, [
        i(e.$slots, "default", {}, () => [
          d(n(f), { class: "size-3.5" }),
          l("span", null, " Upload " + r(e.multiple ? "files" : "file"), 1)
        ])
      ]),
      i(e.$slots, "sublabel", {}, () => [
        e.extensions ? (s(), o("div", b, r(e.extensions.join(", ")), 1)) : u("", !0)
      ])
    ], 16));
  }
});
export {
  h as default
};
