// @scalar/components/dist/components/ScalarFileUpload/ScalarFileUploadFileList.vue.js
import { defineComponent as x, mergeModels as _, useModel as v, createElementBlock as o, openBlock as i, normalizeProps as g, guardReactiveProps as C, unref as l, createElementVNode as t, Fragment as V, renderList as k, toDisplayString as n, createVNode as d, withCtx as a, createTextVNode as m } from "vue";
import { useBindCx as y } from "@scalar/use-hooks/useBindCx";
import z from "pretty-bytes";
import u from "../ScalarButton/ScalarButton.vue.js";
const B = { class: "border rounded p-1 text-xs" }, $ = { class: "flex items-center justify-between" }, w = { class: "flex-1 min-w-0 truncate" }, F = { class: "text-c-3" }, M = { class: "flex items-center gap-1.5" }, N = {}, P = /* @__PURE__ */ x({
  ...N,
  inheritAttrs: !1,
  __name: "ScalarFileUploadFileList",
  props: {
    modelValue: {},
    modelModifiers: {}
  },
  emits: /* @__PURE__ */ _(["upload"], ["update:modelValue"]),
  setup(p) {
    const r = v(p, "modelValue"), { cx: f } = y();
    return (c, e) => (i(), o("div", g(C(l(f)("flex flex-col gap-2 p-4"))), [
      e[4] || (e[4] = t("div", { class: "text-c-2" }, "Selected files:", -1)),
      (i(!0), o(V, null, k(r.value, (s) => (i(), o("ul", {
        key: s.name,
        class: "flex flex-col gap-0.5"
      }, [
        t("li", B, [
          t("div", $, [
            t("div", w, n(s.name), 1),
            t("div", F, n(l(z)(s.size)), 1)
          ])
        ])
      ]))), 128)),
      t("div", M, [
        d(l(u), {
          size: "xs",
          onClick: e[0] || (e[0] = (s) => c.$emit("upload"))
        }, {
          default: a(() => e[2] || (e[2] = [
            m(" Upload ")
          ])),
          _: 1,
          __: [2]
        }),
        d(l(u), {
          size: "xs",
          variant: "outlined",
          onClick: e[1] || (e[1] = (s) => r.value = [])
        }, {
          default: a(() => e[3] || (e[3] = [
            m(" Clear ")
          ])),
          _: 1,
          __: [3]
        })
      ])
    ], 16));
  }
});
export {
  P as default
};
