// @scalar/components/dist/components/ScalarCombobox/ScalarComboboxMultiselect.vue.js
import { defineComponent as b, mergeModels as m, useModel as y, ref as c, createBlock as i, openBlock as f, withCtx as t, createCommentVNode as k, createSlots as V, renderSlot as r, normalizeProps as d, guardReactiveProps as s } from "vue";
import w from "./ScalarComboboxOptions.vue.js";
import z from "./ScalarComboboxPopover.vue.js";
const P = /* @__PURE__ */ b({
  __name: "ScalarComboboxMultiselect",
  props: /* @__PURE__ */ m({
    options: {},
    placeholder: {},
    placement: {},
    offset: { type: [Number, Object, Function] },
    resize: { type: Boolean },
    target: {},
    middleware: {},
    teleport: { type: [Boolean, String] }
  }, {
    modelValue: { default: [] },
    modelModifiers: {}
  }),
  emits: /* @__PURE__ */ m(["add"], ["update:modelValue"]),
  setup(u, { expose: g, emit: v }) {
    const $ = v, n = y(u, "modelValue"), p = c(null);
    return g({ comboboxPopoverRef: p }), (e, l) => (f(), i(z, {
      ref_key: "comboboxPopoverRef",
      ref: p,
      middleware: e.middleware,
      offset: e.offset,
      placement: e.placement ?? "bottom-start",
      resize: e.resize,
      target: e.target,
      teleport: e.teleport
    }, {
      default: t(({ open: a }) => [
        r(e.$slots, "default", { open: a })
      ]),
      popover: t(({ open: a }) => [
        e.options?.length ? (f(), i(w, {
          key: 0,
          modelValue: n.value,
          "onUpdate:modelValue": l[0] || (l[0] = (o) => n.value = o),
          multiselect: "",
          open: a,
          options: e.options,
          placeholder: e.placeholder,
          onAdd: l[1] || (l[1] = (o) => $("add"))
        }, V({ _: 2 }, [
          e.$slots.option ? {
            name: "option",
            fn: t((o) => [
              r(e.$slots, "option", d(s(o)))
            ]),
            key: "0"
          } : void 0,
          e.$slots.group ? {
            name: "group",
            fn: t((o) => [
              r(e.$slots, "group", d(s(o)))
            ]),
            key: "1"
          } : void 0,
          e.$slots.add ? {
            name: "add",
            fn: t((o) => [
              r(e.$slots, "add", d(s(o)))
            ]),
            key: "2"
          } : void 0
        ]), 1032, ["modelValue", "open", "options", "placeholder"])) : k("", !0)
      ]),
      _: 3
    }, 8, ["middleware", "offset", "placement", "resize", "target", "teleport"]));
  }
});
export {
  P as default
};
