// @scalar/components/dist/components/ScalarCombobox/ScalarComboboxPopover.vue.js
import { defineComponent as d, ref as u, createBlock as c, openBlock as _, unref as o, withCtx as t, createVNode as r, normalizeProps as v, guardReactiveProps as y, createSlots as w, renderSlot as p, mergeProps as x } from "vue";
import { Popover as B, PopoverButton as P, PopoverPanel as h } from "@headlessui/vue";
import { useBindCx as k } from "@scalar/use-hooks/useBindCx";
import g from "../ScalarFloating/ScalarFloating.vue.js";
import b from "../ScalarFloating/ScalarFloatingBackdrop.vue.js";
const z = /* @__PURE__ */ d({
  inheritAttrs: !1,
  __name: "ScalarComboboxPopover",
  props: {
    placement: {},
    offset: { type: [Number, Object, Function] },
    resize: { type: Boolean },
    target: {},
    middleware: {},
    teleport: { type: [Boolean, String] }
  },
  setup($, { expose: l }) {
    const { cx: s } = k(), a = u(null), f = (e) => {
      ["ArrowUp", "ArrowDown"].includes(e.key) && (e.preventDefault(), e.target?.dispatchEvent(new KeyboardEvent("keydown", { key: "Enter" })));
    };
    return l({ popoverButtonRef: a }), (e, C) => (_(), c(o(B), { as: "template" }, {
      default: t(({ open: n }) => [
        r(o(g), v(y(e.$props)), w({
          default: t(() => [
            r(o(P), {
              ref_key: "popoverButtonRef",
              ref: a,
              as: "template",
              onKeydown: f
            }, {
              default: t(() => [
                p(e.$slots, "default", { open: n })
              ]),
              _: 2
            }, 1536)
          ]),
          _: 2
        }, [
          n ? {
            name: "floating",
            fn: t(({ width: i }) => [
              r(o(h), x(
                {
                  style: { width: i }
                },
                o(s)("relative flex flex-col max-h-[inherit] w-40 rounded text-sm")
              ), {
                default: t(({ close: m }) => [
                  p(e.$slots, "popover", {
                    close: m,
                    open: n
                  }),
                  r(o(b))
                ]),
                _: 2
              }, 1040, ["style"])
            ]),
            key: "0"
          } : void 0
        ]), 1040)
      ]),
      _: 3
    }));
  }
});
export {
  z as default
};
