// @scalar/components/dist/components/ScalarCombobox/ScalarCombobox.vue.js
import { defineComponent as f, mergeModels as s, useModel as c, createBlock as g, openBlock as $, withCtx as t, createVNode as v, createSlots as V, renderSlot as l, normalizeProps as r, guardReactiveProps as n } from "vue";
import h from "./ScalarComboboxOptions.vue.js";
import y from "./ScalarComboboxPopover.vue.js";
const B = /* @__PURE__ */ f({
  __name: "ScalarCombobox",
  props: /* @__PURE__ */ s({
    options: {},
    placeholder: {},
    placement: {},
    offset: { type: [Number, Object, Function] },
    resize: { type: Boolean },
    target: {},
    middleware: {},
    teleport: { type: [Boolean, String] }
  }, {
    modelValue: {},
    modelModifiers: {}
  }),
  emits: /* @__PURE__ */ s(["add"], ["update:modelValue"]),
  setup(m, { emit: i }) {
    const a = c(m, "modelValue"), u = i;
    return (e, b) => ($(), g(y, {
      middleware: e.middleware,
      offset: e.offset,
      placement: e.placement ?? "bottom-start",
      resize: e.resize,
      target: e.target,
      teleport: e.teleport
    }, {
      default: t(({ open: d }) => [
        l(e.$slots, "default", { open: d })
      ]),
      popover: t(({ open: d, close: p }) => [
        v(h, {
          modelValue: a.value ? [a.value] : [],
          open: d,
          options: e.options,
          placeholder: e.placeholder,
          "onUpdate:modelValue": (o) => (p(), a.value = o[0]),
          onAdd: () => (p(), u("add"))
        }, V({ _: 2 }, [
          e.$slots.option ? {
            name: "option",
            fn: t((o) => [
              l(e.$slots, "option", r(n(o)))
            ]),
            key: "0"
          } : void 0,
          e.$slots.group ? {
            name: "group",
            fn: t((o) => [
              l(e.$slots, "group", r(n(o)))
            ]),
            key: "1"
          } : void 0,
          e.$slots.add ? {
            name: "add",
            fn: t((o) => [
              l(e.$slots, "add", r(n(o)))
            ]),
            key: "2"
          } : void 0
        ]), 1032, ["modelValue", "open", "options", "placeholder", "onUpdate:modelValue", "onAdd"])
      ]),
      _: 3
    }, 8, ["middleware", "offset", "placement", "resize", "target", "teleport"]));
  }
});
export {
  B as default
};
